{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Dataframe join examples\n",
    "\n",
    "In this use case, we provide several examples of how to use the built-in <code>cptac</code> functions for joining different dataframes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import cptac\n",
    "en = cptac.Endometrial()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## General format\n",
    "\n",
    "In all of the join functions, you specify the dataframes you want to join by passing their names to the appropriate parameters in the function call. The function will automatically check that the dataframes whose names you provided are valid for the join function, and print an error message if they aren't.\n",
    "\n",
    "Whenever a column from an -omics dataframe is included in a joined table, the name of the -omics dataframe it came from is appended to the column header, to avoid confusion.\n",
    "\n",
    "If you wish to only include particular columns in the join, pass them to the appropriate parameters in the join function. All such parameters will accept either a single column name as a string, or a list of column name strings.\n",
    "\n",
    "All join functions use logic analogous to an SQL <code>INNER JOIN</code>."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## join_omics_to_omics\n",
    "\n",
    "The <code>join_omics_to_omics</code> function joins two -omics dataframes to each other:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>A1BG_proteomics</th>\n",
       "      <th>A2M_proteomics</th>\n",
       "      <th>A2ML1_proteomics</th>\n",
       "      <th>A4GALT_proteomics</th>\n",
       "      <th>AAAS_proteomics</th>\n",
       "      <th>AACS_proteomics</th>\n",
       "      <th>AADAT_proteomics</th>\n",
       "      <th>AAED1_proteomics</th>\n",
       "      <th>AAGAB_proteomics</th>\n",
       "      <th>AAK1_proteomics</th>\n",
       "      <th>...</th>\n",
       "      <th>ZZZ3-S397_phosphoproteomics</th>\n",
       "      <th>ZZZ3-S411_phosphoproteomics</th>\n",
       "      <th>ZZZ3-S420_phosphoproteomics</th>\n",
       "      <th>ZZZ3-S424_phosphoproteomics</th>\n",
       "      <th>ZZZ3-S426_phosphoproteomics</th>\n",
       "      <th>ZZZ3-S468_phosphoproteomics</th>\n",
       "      <th>ZZZ3-S89_phosphoproteomics</th>\n",
       "      <th>ZZZ3-T415_phosphoproteomics</th>\n",
       "      <th>ZZZ3-T418_phosphoproteomics</th>\n",
       "      <th>ZZZ3-Y399_phosphoproteomics</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sample_ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>S001</th>\n",
       "      <td>-1.180</td>\n",
       "      <td>-0.8630</td>\n",
       "      <td>-0.802</td>\n",
       "      <td>0.222</td>\n",
       "      <td>0.2560</td>\n",
       "      <td>0.6650</td>\n",
       "      <td>1.2800</td>\n",
       "      <td>-0.3390</td>\n",
       "      <td>0.412</td>\n",
       "      <td>-0.664</td>\n",
       "      <td>...</td>\n",
       "      <td>0.18400</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.20500</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S002</th>\n",
       "      <td>-0.685</td>\n",
       "      <td>-1.0700</td>\n",
       "      <td>-0.684</td>\n",
       "      <td>0.984</td>\n",
       "      <td>0.1350</td>\n",
       "      <td>0.3340</td>\n",
       "      <td>1.3000</td>\n",
       "      <td>0.1390</td>\n",
       "      <td>1.330</td>\n",
       "      <td>-0.367</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.17100</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.393</td>\n",
       "      <td>-0.17100</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.29</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.1605</td>\n",
       "      <td>-0.0635</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S003</th>\n",
       "      <td>-0.528</td>\n",
       "      <td>-1.3200</td>\n",
       "      <td>0.435</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.2400</td>\n",
       "      <td>1.0400</td>\n",
       "      <td>-0.0213</td>\n",
       "      <td>-0.0479</td>\n",
       "      <td>0.419</td>\n",
       "      <td>-0.500</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S005</th>\n",
       "      <td>-1.670</td>\n",
       "      <td>-1.1900</td>\n",
       "      <td>-0.443</td>\n",
       "      <td>0.243</td>\n",
       "      <td>-0.0993</td>\n",
       "      <td>0.7570</td>\n",
       "      <td>0.7400</td>\n",
       "      <td>-0.9290</td>\n",
       "      <td>0.229</td>\n",
       "      <td>-0.223</td>\n",
       "      <td>...</td>\n",
       "      <td>0.13970</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.55900</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.2980</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S006</th>\n",
       "      <td>-0.374</td>\n",
       "      <td>-0.0206</td>\n",
       "      <td>-0.537</td>\n",
       "      <td>0.311</td>\n",
       "      <td>0.3750</td>\n",
       "      <td>0.0131</td>\n",
       "      <td>-1.1000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.565</td>\n",
       "      <td>-0.101</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.15875</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.196</td>\n",
       "      <td>0.06175</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.2900</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 84211 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           A1BG_proteomics  A2M_proteomics  A2ML1_proteomics  \\\n",
       "Sample_ID                                                      \n",
       "S001                -1.180         -0.8630            -0.802   \n",
       "S002                -0.685         -1.0700            -0.684   \n",
       "S003                -0.528         -1.3200             0.435   \n",
       "S005                -1.670         -1.1900            -0.443   \n",
       "S006                -0.374         -0.0206            -0.537   \n",
       "\n",
       "           A4GALT_proteomics  AAAS_proteomics  AACS_proteomics  \\\n",
       "Sample_ID                                                        \n",
       "S001                   0.222           0.2560           0.6650   \n",
       "S002                   0.984           0.1350           0.3340   \n",
       "S003                     NaN          -0.2400           1.0400   \n",
       "S005                   0.243          -0.0993           0.7570   \n",
       "S006                   0.311           0.3750           0.0131   \n",
       "\n",
       "           AADAT_proteomics  AAED1_proteomics  AAGAB_proteomics  \\\n",
       "Sample_ID                                                         \n",
       "S001                 1.2800           -0.3390             0.412   \n",
       "S002                 1.3000            0.1390             1.330   \n",
       "S003                -0.0213           -0.0479             0.419   \n",
       "S005                 0.7400           -0.9290             0.229   \n",
       "S006                -1.1000               NaN             0.565   \n",
       "\n",
       "           AAK1_proteomics  ...  ZZZ3-S397_phosphoproteomics  \\\n",
       "Sample_ID                   ...                                \n",
       "S001                -0.664  ...                      0.18400   \n",
       "S002                -0.367  ...                     -0.17100   \n",
       "S003                -0.500  ...                          NaN   \n",
       "S005                -0.223  ...                      0.13970   \n",
       "S006                -0.101  ...                     -0.15875   \n",
       "\n",
       "           ZZZ3-S411_phosphoproteomics  ZZZ3-S420_phosphoproteomics  \\\n",
       "Sample_ID                                                             \n",
       "S001                               NaN                          NaN   \n",
       "S002                               NaN                          NaN   \n",
       "S003                               NaN                          NaN   \n",
       "S005                               NaN                          NaN   \n",
       "S006                               NaN                          NaN   \n",
       "\n",
       "           ZZZ3-S424_phosphoproteomics  ZZZ3-S426_phosphoproteomics  \\\n",
       "Sample_ID                                                             \n",
       "S001                               NaN                     -0.20500   \n",
       "S002                            -0.393                     -0.17100   \n",
       "S003                               NaN                          NaN   \n",
       "S005                               NaN                     -0.55900   \n",
       "S006                             0.196                      0.06175   \n",
       "\n",
       "           ZZZ3-S468_phosphoproteomics  ZZZ3-S89_phosphoproteomics  \\\n",
       "Sample_ID                                                            \n",
       "S001                               NaN                         NaN   \n",
       "S002                               NaN                        0.29   \n",
       "S003                               NaN                         NaN   \n",
       "S005                               NaN                         NaN   \n",
       "S006                               NaN                         NaN   \n",
       "\n",
       "           ZZZ3-T415_phosphoproteomics  ZZZ3-T418_phosphoproteomics  \\\n",
       "Sample_ID                                                             \n",
       "S001                               NaN                          NaN   \n",
       "S002                               NaN                       0.1605   \n",
       "S003                               NaN                          NaN   \n",
       "S005                               NaN                          NaN   \n",
       "S006                               NaN                          NaN   \n",
       "\n",
       "           ZZZ3-Y399_phosphoproteomics  \n",
       "Sample_ID                               \n",
       "S001                               NaN  \n",
       "S002                           -0.0635  \n",
       "S003                               NaN  \n",
       "S005                            0.2980  \n",
       "S006                           -0.2900  \n",
       "\n",
       "[5 rows x 84211 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "prot_and_phos = en.join_omics_to_omics(omics_df1_name=\"proteomics\", omics_df2_name=\"phosphoproteomics\")\n",
    "prot_and_phos.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Joining only specific columns:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>A1BG_proteomics</th>\n",
       "      <th>PIK3CA-S312_phosphoproteomics</th>\n",
       "      <th>PIK3CA-T313_phosphoproteomics</th>\n",
       "      <th>TP53-S315_phosphoproteomics</th>\n",
       "      <th>TP53-T150_phosphoproteomics</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sample_ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>S001</th>\n",
       "      <td>-1.180</td>\n",
       "      <td>-0.00615</td>\n",
       "      <td>0.0731</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S002</th>\n",
       "      <td>-0.685</td>\n",
       "      <td>-0.02220</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.646</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S003</th>\n",
       "      <td>-0.528</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0830</td>\n",
       "      <td>-0.800</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S005</th>\n",
       "      <td>-1.670</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.8460</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S006</th>\n",
       "      <td>-0.374</td>\n",
       "      <td>0.43600</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3.760</td>\n",
       "      <td>0.253</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           A1BG_proteomics  PIK3CA-S312_phosphoproteomics  \\\n",
       "Sample_ID                                                   \n",
       "S001                -1.180                       -0.00615   \n",
       "S002                -0.685                       -0.02220   \n",
       "S003                -0.528                            NaN   \n",
       "S005                -1.670                            NaN   \n",
       "S006                -0.374                        0.43600   \n",
       "\n",
       "           PIK3CA-T313_phosphoproteomics  TP53-S315_phosphoproteomics  \\\n",
       "Sample_ID                                                               \n",
       "S001                              0.0731                          NaN   \n",
       "S002                                 NaN                        0.646   \n",
       "S003                              0.0830                       -0.800   \n",
       "S005                             -0.8460                          NaN   \n",
       "S006                                 NaN                        3.760   \n",
       "\n",
       "           TP53-T150_phosphoproteomics  \n",
       "Sample_ID                               \n",
       "S001                               NaN  \n",
       "S002                               NaN  \n",
       "S003                               NaN  \n",
       "S005                               NaN  \n",
       "S006                             0.253  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "prot_and_phos_selected = en.join_omics_to_omics(omics_df1_name=\"proteomics\", omics_df2_name=\"phosphoproteomics\", genes1=\"A1BG\", genes2=[\"PIK3CA\", \"TP53\"])\n",
    "prot_and_phos_selected.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## join_metadata_to_omics\n",
    "\n",
    "The <code>join_metadata_to_omics</code> function joins a metadata dataframe (e.g. clinical, derived_molecular, or treatment) with an -omics dataframe:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Patient_ID</th>\n",
       "      <th>Proteomics_Tumor_Normal</th>\n",
       "      <th>Country</th>\n",
       "      <th>Histologic_Grade_FIGO</th>\n",
       "      <th>Myometrial_invasion_Specify</th>\n",
       "      <th>Histologic_type</th>\n",
       "      <th>Treatment_naive</th>\n",
       "      <th>Tumor_purity</th>\n",
       "      <th>Path_Stage_Primary_Tumor-pT</th>\n",
       "      <th>Path_Stage_Reg_Lymph_Nodes-pN</th>\n",
       "      <th>...</th>\n",
       "      <th>ZWILCH_transcriptomics</th>\n",
       "      <th>ZWINT_transcriptomics</th>\n",
       "      <th>ZXDA_transcriptomics</th>\n",
       "      <th>ZXDB_transcriptomics</th>\n",
       "      <th>ZXDC_transcriptomics</th>\n",
       "      <th>ZYG11A_transcriptomics</th>\n",
       "      <th>ZYG11B_transcriptomics</th>\n",
       "      <th>ZYX_transcriptomics</th>\n",
       "      <th>ZZEF1_transcriptomics</th>\n",
       "      <th>ZZZ3_transcriptomics</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sample_ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>S001</th>\n",
       "      <td>C3L-00006</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>United States</td>\n",
       "      <td>FIGO grade 1</td>\n",
       "      <td>under 50 %</td>\n",
       "      <td>Endometrioid</td>\n",
       "      <td>YES</td>\n",
       "      <td>Normal</td>\n",
       "      <td>pT1a (FIGO IA)</td>\n",
       "      <td>pN0</td>\n",
       "      <td>...</td>\n",
       "      <td>11.06</td>\n",
       "      <td>10.73</td>\n",
       "      <td>8.40</td>\n",
       "      <td>9.78</td>\n",
       "      <td>10.88</td>\n",
       "      <td>5.93</td>\n",
       "      <td>11.52</td>\n",
       "      <td>10.23</td>\n",
       "      <td>11.50</td>\n",
       "      <td>11.47</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S002</th>\n",
       "      <td>C3L-00008</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>United States</td>\n",
       "      <td>FIGO grade 1</td>\n",
       "      <td>under 50 %</td>\n",
       "      <td>Endometrioid</td>\n",
       "      <td>YES</td>\n",
       "      <td>Normal</td>\n",
       "      <td>pT1a (FIGO IA)</td>\n",
       "      <td>pNX</td>\n",
       "      <td>...</td>\n",
       "      <td>10.87</td>\n",
       "      <td>11.43</td>\n",
       "      <td>8.39</td>\n",
       "      <td>9.14</td>\n",
       "      <td>10.38</td>\n",
       "      <td>7.25</td>\n",
       "      <td>11.64</td>\n",
       "      <td>10.64</td>\n",
       "      <td>11.26</td>\n",
       "      <td>11.57</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S003</th>\n",
       "      <td>C3L-00032</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>United States</td>\n",
       "      <td>FIGO grade 2</td>\n",
       "      <td>under 50 %</td>\n",
       "      <td>Endometrioid</td>\n",
       "      <td>YES</td>\n",
       "      <td>Normal</td>\n",
       "      <td>pT1a (FIGO IA)</td>\n",
       "      <td>pN0</td>\n",
       "      <td>...</td>\n",
       "      <td>10.06</td>\n",
       "      <td>10.13</td>\n",
       "      <td>8.35</td>\n",
       "      <td>9.27</td>\n",
       "      <td>10.46</td>\n",
       "      <td>6.85</td>\n",
       "      <td>11.60</td>\n",
       "      <td>10.21</td>\n",
       "      <td>11.51</td>\n",
       "      <td>11.09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S005</th>\n",
       "      <td>C3L-00090</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>United States</td>\n",
       "      <td>FIGO grade 2</td>\n",
       "      <td>under 50 %</td>\n",
       "      <td>Endometrioid</td>\n",
       "      <td>YES</td>\n",
       "      <td>Normal</td>\n",
       "      <td>pT1a (FIGO IA)</td>\n",
       "      <td>pNX</td>\n",
       "      <td>...</td>\n",
       "      <td>10.29</td>\n",
       "      <td>10.41</td>\n",
       "      <td>9.10</td>\n",
       "      <td>9.59</td>\n",
       "      <td>10.15</td>\n",
       "      <td>7.89</td>\n",
       "      <td>11.90</td>\n",
       "      <td>10.21</td>\n",
       "      <td>11.34</td>\n",
       "      <td>11.51</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S006</th>\n",
       "      <td>C3L-00098</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>United States</td>\n",
       "      <td>NaN</td>\n",
       "      <td>under 50 %</td>\n",
       "      <td>Serous</td>\n",
       "      <td>YES</td>\n",
       "      <td>Normal</td>\n",
       "      <td>pT1a (FIGO IA)</td>\n",
       "      <td>pNX</td>\n",
       "      <td>...</td>\n",
       "      <td>10.36</td>\n",
       "      <td>11.24</td>\n",
       "      <td>8.60</td>\n",
       "      <td>9.44</td>\n",
       "      <td>11.80</td>\n",
       "      <td>9.32</td>\n",
       "      <td>11.97</td>\n",
       "      <td>9.77</td>\n",
       "      <td>11.37</td>\n",
       "      <td>12.35</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 28083 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          Patient_ID Proteomics_Tumor_Normal        Country  \\\n",
       "Sample_ID                                                     \n",
       "S001       C3L-00006                   Tumor  United States   \n",
       "S002       C3L-00008                   Tumor  United States   \n",
       "S003       C3L-00032                   Tumor  United States   \n",
       "S005       C3L-00090                   Tumor  United States   \n",
       "S006       C3L-00098                   Tumor  United States   \n",
       "\n",
       "          Histologic_Grade_FIGO Myometrial_invasion_Specify Histologic_type  \\\n",
       "Sample_ID                                                                     \n",
       "S001               FIGO grade 1                  under 50 %    Endometrioid   \n",
       "S002               FIGO grade 1                  under 50 %    Endometrioid   \n",
       "S003               FIGO grade 2                  under 50 %    Endometrioid   \n",
       "S005               FIGO grade 2                  under 50 %    Endometrioid   \n",
       "S006                        NaN                  under 50 %          Serous   \n",
       "\n",
       "          Treatment_naive Tumor_purity Path_Stage_Primary_Tumor-pT  \\\n",
       "Sample_ID                                                            \n",
       "S001                  YES       Normal              pT1a (FIGO IA)   \n",
       "S002                  YES       Normal              pT1a (FIGO IA)   \n",
       "S003                  YES       Normal              pT1a (FIGO IA)   \n",
       "S005                  YES       Normal              pT1a (FIGO IA)   \n",
       "S006                  YES       Normal              pT1a (FIGO IA)   \n",
       "\n",
       "          Path_Stage_Reg_Lymph_Nodes-pN  ... ZWILCH_transcriptomics  \\\n",
       "Sample_ID                                ...                          \n",
       "S001                                pN0  ...                  11.06   \n",
       "S002                                pNX  ...                  10.87   \n",
       "S003                                pN0  ...                  10.06   \n",
       "S005                                pNX  ...                  10.29   \n",
       "S006                                pNX  ...                  10.36   \n",
       "\n",
       "          ZWINT_transcriptomics ZXDA_transcriptomics ZXDB_transcriptomics  \\\n",
       "Sample_ID                                                                   \n",
       "S001                      10.73                 8.40                 9.78   \n",
       "S002                      11.43                 8.39                 9.14   \n",
       "S003                      10.13                 8.35                 9.27   \n",
       "S005                      10.41                 9.10                 9.59   \n",
       "S006                      11.24                 8.60                 9.44   \n",
       "\n",
       "           ZXDC_transcriptomics  ZYG11A_transcriptomics  \\\n",
       "Sample_ID                                                 \n",
       "S001                      10.88                    5.93   \n",
       "S002                      10.38                    7.25   \n",
       "S003                      10.46                    6.85   \n",
       "S005                      10.15                    7.89   \n",
       "S006                      11.80                    9.32   \n",
       "\n",
       "           ZYG11B_transcriptomics ZYX_transcriptomics ZZEF1_transcriptomics  \\\n",
       "Sample_ID                                                                     \n",
       "S001                        11.52               10.23                 11.50   \n",
       "S002                        11.64               10.64                 11.26   \n",
       "S003                        11.60               10.21                 11.51   \n",
       "S005                        11.90               10.21                 11.34   \n",
       "S006                        11.97                9.77                 11.37   \n",
       "\n",
       "          ZZZ3_transcriptomics  \n",
       "Sample_ID                       \n",
       "S001                     11.47  \n",
       "S002                     11.57  \n",
       "S003                     11.09  \n",
       "S005                     11.51  \n",
       "S006                     12.35  \n",
       "\n",
       "[5 rows x 28083 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clin_and_tran = en.join_metadata_to_omics(metadata_df_name=\"clinical\", omics_df_name=\"transcriptomics\")\n",
    "clin_and_tran.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Joining only specific columns:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Age</th>\n",
       "      <th>Histologic_type</th>\n",
       "      <th>ZZZ3_transcriptomics</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sample_ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>S001</th>\n",
       "      <td>64.0</td>\n",
       "      <td>Endometrioid</td>\n",
       "      <td>11.47</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S002</th>\n",
       "      <td>58.0</td>\n",
       "      <td>Endometrioid</td>\n",
       "      <td>11.57</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S003</th>\n",
       "      <td>50.0</td>\n",
       "      <td>Endometrioid</td>\n",
       "      <td>11.09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S005</th>\n",
       "      <td>75.0</td>\n",
       "      <td>Endometrioid</td>\n",
       "      <td>11.51</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S006</th>\n",
       "      <td>63.0</td>\n",
       "      <td>Serous</td>\n",
       "      <td>12.35</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            Age Histologic_type  ZZZ3_transcriptomics\n",
       "Sample_ID                                            \n",
       "S001       64.0    Endometrioid                 11.47\n",
       "S002       58.0    Endometrioid                 11.57\n",
       "S003       50.0    Endometrioid                 11.09\n",
       "S005       75.0    Endometrioid                 11.51\n",
       "S006       63.0          Serous                 12.35"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clin_and_tran = en.join_metadata_to_omics(metadata_df_name=\"clinical\", omics_df_name=\"transcriptomics\", metadata_cols = [\"Age\", \"Histologic_type\"], omics_genes=\"ZZZ3\")\n",
    "clin_and_tran.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## join_omics_to_mutations\n",
    "\n",
    "The <code>join_omics_to_mutations</code> function joins an -omics dataframe with the mutation data for a specified gene or genes. Because there may be multiple mutations for one gene in a single sample, the mutation type and location data are returned in lists. If there is no mutation for the gene in a particular sample, the list contains either \"Wildtype_Tumor\" or \"Wildtype_Normal\", depending on whether it's a tumor or normal sample. The mutation status column contains either \"Single_mutation\", \"Multiple_mutation\", \"Wildtype_Tumor\", or \"Wildtype_Normal\", for help with parsing."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>A2M-K1168_acetylproteomics</th>\n",
       "      <th>A2M-K1176_acetylproteomics</th>\n",
       "      <th>A2M-K135_acetylproteomics</th>\n",
       "      <th>A2M-K145_acetylproteomics</th>\n",
       "      <th>A2M-K516_acetylproteomics</th>\n",
       "      <th>A2M-K664_acetylproteomics</th>\n",
       "      <th>A2M-K682_acetylproteomics</th>\n",
       "      <th>AACS-K391_acetylproteomics</th>\n",
       "      <th>AAGAB-K290_acetylproteomics</th>\n",
       "      <th>AAK1-K201_acetylproteomics</th>\n",
       "      <th>...</th>\n",
       "      <th>ZYX-K25_acetylproteomics</th>\n",
       "      <th>ZYX-K265_acetylproteomics</th>\n",
       "      <th>ZYX-K272_acetylproteomics</th>\n",
       "      <th>ZYX-K279_acetylproteomics</th>\n",
       "      <th>ZYX-K533_acetylproteomics</th>\n",
       "      <th>ZZZ3-K117_acetylproteomics</th>\n",
       "      <th>TP53_Mutation</th>\n",
       "      <th>TP53_Location</th>\n",
       "      <th>TP53_Mutation_Status</th>\n",
       "      <th>Sample_Status</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sample_ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>S001</th>\n",
       "      <td>NaN</td>\n",
       "      <td>1.080</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.461</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[Missense_Mutation]</td>\n",
       "      <td>[p.R248W]</td>\n",
       "      <td>Single_mutation</td>\n",
       "      <td>Tumor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S002</th>\n",
       "      <td>NaN</td>\n",
       "      <td>0.477</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.770</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.343</td>\n",
       "      <td>-0.307</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.0955</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[Wildtype_Tumor]</td>\n",
       "      <td>[No_mutation]</td>\n",
       "      <td>Wildtype_Tumor</td>\n",
       "      <td>Tumor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S003</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.815</td>\n",
       "      <td>-0.00573</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.170</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.7050</td>\n",
       "      <td>0.089</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[Wildtype_Tumor]</td>\n",
       "      <td>[No_mutation]</td>\n",
       "      <td>Wildtype_Tumor</td>\n",
       "      <td>Tumor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S005</th>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.608</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.919</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.537</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.3700</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[Wildtype_Tumor]</td>\n",
       "      <td>[No_mutation]</td>\n",
       "      <td>Wildtype_Tumor</td>\n",
       "      <td>Tumor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S006</th>\n",
       "      <td>NaN</td>\n",
       "      <td>1.630</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.26</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.205</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.358</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.9700</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[Missense_Mutation]</td>\n",
       "      <td>[p.S241C]</td>\n",
       "      <td>Single_mutation</td>\n",
       "      <td>Tumor</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 10866 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           A2M-K1168_acetylproteomics  A2M-K1176_acetylproteomics  \\\n",
       "Sample_ID                                                           \n",
       "S001                              NaN                       1.080   \n",
       "S002                              NaN                       0.477   \n",
       "S003                              NaN                         NaN   \n",
       "S005                              NaN                      -0.608   \n",
       "S006                              NaN                       1.630   \n",
       "\n",
       "           A2M-K135_acetylproteomics  A2M-K145_acetylproteomics  \\\n",
       "Sample_ID                                                         \n",
       "S001                             NaN                        NaN   \n",
       "S002                             NaN                        NaN   \n",
       "S003                             NaN                        NaN   \n",
       "S005                             NaN                        NaN   \n",
       "S006                             NaN                        2.4   \n",
       "\n",
       "           A2M-K516_acetylproteomics  A2M-K664_acetylproteomics  \\\n",
       "Sample_ID                                                         \n",
       "S001                             NaN                        NaN   \n",
       "S002                             NaN                        NaN   \n",
       "S003                             NaN                        NaN   \n",
       "S005                          -0.919                        NaN   \n",
       "S006                             NaN                        NaN   \n",
       "\n",
       "           A2M-K682_acetylproteomics  AACS-K391_acetylproteomics  \\\n",
       "Sample_ID                                                          \n",
       "S001                             NaN                         NaN   \n",
       "S002                             NaN                         NaN   \n",
       "S003                             NaN                         NaN   \n",
       "S005                             NaN                         NaN   \n",
       "S006                            1.26                         NaN   \n",
       "\n",
       "           AAGAB-K290_acetylproteomics  AAK1-K201_acetylproteomics  ...  \\\n",
       "Sample_ID                                                           ...   \n",
       "S001                             0.461                         NaN  ...   \n",
       "S002                             1.770                         NaN  ...   \n",
       "S003                            -0.815                    -0.00573  ...   \n",
       "S005                               NaN                         NaN  ...   \n",
       "S006                             0.205                         NaN  ...   \n",
       "\n",
       "           ZYX-K25_acetylproteomics  ZYX-K265_acetylproteomics  \\\n",
       "Sample_ID                                                        \n",
       "S001                            NaN                        NaN   \n",
       "S002                         -0.343                     -0.307   \n",
       "S003                         -1.170                        NaN   \n",
       "S005                         -0.537                        NaN   \n",
       "S006                         -0.358                        NaN   \n",
       "\n",
       "           ZYX-K272_acetylproteomics  ZYX-K279_acetylproteomics  \\\n",
       "Sample_ID                                                         \n",
       "S001                             NaN                        NaN   \n",
       "S002                             NaN                    -0.0955   \n",
       "S003                             NaN                    -0.7050   \n",
       "S005                             NaN                    -0.3700   \n",
       "S006                             NaN                    -0.9700   \n",
       "\n",
       "           ZYX-K533_acetylproteomics  ZZZ3-K117_acetylproteomics  \\\n",
       "Sample_ID                                                          \n",
       "S001                             NaN                         NaN   \n",
       "S002                             NaN                         NaN   \n",
       "S003                           0.089                         NaN   \n",
       "S005                             NaN                         NaN   \n",
       "S006                             NaN                         NaN   \n",
       "\n",
       "                 TP53_Mutation  TP53_Location  TP53_Mutation_Status  \\\n",
       "Sample_ID                                                             \n",
       "S001       [Missense_Mutation]      [p.R248W]       Single_mutation   \n",
       "S002          [Wildtype_Tumor]  [No_mutation]        Wildtype_Tumor   \n",
       "S003          [Wildtype_Tumor]  [No_mutation]        Wildtype_Tumor   \n",
       "S005          [Wildtype_Tumor]  [No_mutation]        Wildtype_Tumor   \n",
       "S006       [Missense_Mutation]      [p.S241C]       Single_mutation   \n",
       "\n",
       "           Sample_Status  \n",
       "Sample_ID                 \n",
       "S001               Tumor  \n",
       "S002               Tumor  \n",
       "S003               Tumor  \n",
       "S005               Tumor  \n",
       "S006               Tumor  \n",
       "\n",
       "[5 rows x 10866 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "acet_and_TP53_mut = en.join_omics_to_mutations(omics_df_name=\"acetylproteomics\", mutations_genes=\"TP53\")\n",
    "acet_and_TP53_mut.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Selecting only specified columns from the -omics dataframe:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AAGAB-K290_acetylproteomics</th>\n",
       "      <th>AACS-K391_acetylproteomics</th>\n",
       "      <th>TP53_Mutation</th>\n",
       "      <th>TP53_Location</th>\n",
       "      <th>TP53_Mutation_Status</th>\n",
       "      <th>Sample_Status</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sample_ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>S001</th>\n",
       "      <td>0.461</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[Missense_Mutation]</td>\n",
       "      <td>[p.R248W]</td>\n",
       "      <td>Single_mutation</td>\n",
       "      <td>Tumor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S002</th>\n",
       "      <td>1.770</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[Wildtype_Tumor]</td>\n",
       "      <td>[No_mutation]</td>\n",
       "      <td>Wildtype_Tumor</td>\n",
       "      <td>Tumor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S003</th>\n",
       "      <td>-0.815</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[Wildtype_Tumor]</td>\n",
       "      <td>[No_mutation]</td>\n",
       "      <td>Wildtype_Tumor</td>\n",
       "      <td>Tumor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S005</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[Wildtype_Tumor]</td>\n",
       "      <td>[No_mutation]</td>\n",
       "      <td>Wildtype_Tumor</td>\n",
       "      <td>Tumor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S006</th>\n",
       "      <td>0.205</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[Missense_Mutation]</td>\n",
       "      <td>[p.S241C]</td>\n",
       "      <td>Single_mutation</td>\n",
       "      <td>Tumor</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           AAGAB-K290_acetylproteomics  AACS-K391_acetylproteomics  \\\n",
       "Sample_ID                                                            \n",
       "S001                             0.461                         NaN   \n",
       "S002                             1.770                         NaN   \n",
       "S003                            -0.815                         NaN   \n",
       "S005                               NaN                         NaN   \n",
       "S006                             0.205                         NaN   \n",
       "\n",
       "                 TP53_Mutation  TP53_Location TP53_Mutation_Status  \\\n",
       "Sample_ID                                                            \n",
       "S001       [Missense_Mutation]      [p.R248W]      Single_mutation   \n",
       "S002          [Wildtype_Tumor]  [No_mutation]       Wildtype_Tumor   \n",
       "S003          [Wildtype_Tumor]  [No_mutation]       Wildtype_Tumor   \n",
       "S005          [Wildtype_Tumor]  [No_mutation]       Wildtype_Tumor   \n",
       "S006       [Missense_Mutation]      [p.S241C]      Single_mutation   \n",
       "\n",
       "          Sample_Status  \n",
       "Sample_ID                \n",
       "S001              Tumor  \n",
       "S002              Tumor  \n",
       "S003              Tumor  \n",
       "S005              Tumor  \n",
       "S006              Tumor  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "selected_acet_and_TP53_mut = en.join_omics_to_mutations(omics_df_name=\"acetylproteomics\", mutations_genes=\"TP53\", omics_genes=[\"AAGAB\", \"AACS\"])\n",
    "selected_acet_and_TP53_mut.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To not hide the mutation location column, pass <code>False</code> to the optional <code>show_location</code> parameter:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AAGAB-K290_acetylproteomics</th>\n",
       "      <th>AACS-K391_acetylproteomics</th>\n",
       "      <th>PIK3CA_Mutation</th>\n",
       "      <th>PIK3CA_Mutation_Status</th>\n",
       "      <th>TP53_Mutation</th>\n",
       "      <th>TP53_Mutation_Status</th>\n",
       "      <th>Sample_Status</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sample_ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>S001</th>\n",
       "      <td>0.461</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[Missense_Mutation]</td>\n",
       "      <td>Single_mutation</td>\n",
       "      <td>[Missense_Mutation]</td>\n",
       "      <td>Single_mutation</td>\n",
       "      <td>Tumor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S002</th>\n",
       "      <td>1.770</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[Wildtype_Tumor]</td>\n",
       "      <td>Wildtype_Tumor</td>\n",
       "      <td>[Wildtype_Tumor]</td>\n",
       "      <td>Wildtype_Tumor</td>\n",
       "      <td>Tumor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S003</th>\n",
       "      <td>-0.815</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[Missense_Mutation]</td>\n",
       "      <td>Single_mutation</td>\n",
       "      <td>[Wildtype_Tumor]</td>\n",
       "      <td>Wildtype_Tumor</td>\n",
       "      <td>Tumor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S005</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[Wildtype_Tumor]</td>\n",
       "      <td>Wildtype_Tumor</td>\n",
       "      <td>[Wildtype_Tumor]</td>\n",
       "      <td>Wildtype_Tumor</td>\n",
       "      <td>Tumor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S006</th>\n",
       "      <td>0.205</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[Wildtype_Tumor]</td>\n",
       "      <td>Wildtype_Tumor</td>\n",
       "      <td>[Missense_Mutation]</td>\n",
       "      <td>Single_mutation</td>\n",
       "      <td>Tumor</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           AAGAB-K290_acetylproteomics  AACS-K391_acetylproteomics  \\\n",
       "Sample_ID                                                            \n",
       "S001                             0.461                         NaN   \n",
       "S002                             1.770                         NaN   \n",
       "S003                            -0.815                         NaN   \n",
       "S005                               NaN                         NaN   \n",
       "S006                             0.205                         NaN   \n",
       "\n",
       "               PIK3CA_Mutation PIK3CA_Mutation_Status        TP53_Mutation  \\\n",
       "Sample_ID                                                                    \n",
       "S001       [Missense_Mutation]        Single_mutation  [Missense_Mutation]   \n",
       "S002          [Wildtype_Tumor]         Wildtype_Tumor     [Wildtype_Tumor]   \n",
       "S003       [Missense_Mutation]        Single_mutation     [Wildtype_Tumor]   \n",
       "S005          [Wildtype_Tumor]         Wildtype_Tumor     [Wildtype_Tumor]   \n",
       "S006          [Wildtype_Tumor]         Wildtype_Tumor  [Missense_Mutation]   \n",
       "\n",
       "          TP53_Mutation_Status Sample_Status  \n",
       "Sample_ID                                     \n",
       "S001           Single_mutation         Tumor  \n",
       "S002            Wildtype_Tumor         Tumor  \n",
       "S003            Wildtype_Tumor         Tumor  \n",
       "S005            Wildtype_Tumor         Tumor  \n",
       "S006           Single_mutation         Tumor  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "more_selected_acet_and_mut = en.join_omics_to_mutations(omics_df_name=\"acetylproteomics\", mutations_genes=[\"PIK3CA\", \"TP53\"], omics_genes=[\"AAGAB\", \"AACS\"], show_location=False)\n",
    "more_selected_acet_and_mut.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Exporting dataframes\n",
    "\n",
    "If you wish to export a dataframe to a file, simply call the dataframe's to_csv method, passing the path you wish to save the file to, and the value separator you want:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "more_selected_acet_and_mut.to_csv(path_or_buf=\"more_selected_acet_and_mut.tsv\", sep='\\t')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
