{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Tutorial 1: CPTAC Data Introduction\n",
    "\n",
    "The National Cancer Institute’s Clinical Proteomic Tumor Analysis Consortium (CPTAC) is a national effort to accelerate the understanding of the molecular basis of cancer through the application of large-scale proteome and genome analysis, or proteogenomics. CPTAC generates comprehensive proteomics and genomics data from clinical cohorts, typically with ~100 samples per tumor type. The graphic below summarizes the structure of each CPTAC dataset. For more information, visit the [NIH website](https://proteomics.cancer.gov/programs/cptac). \n",
    "\n",
    "<img src=\"img/Graphical_Abstract.png\" alt=\"CPTAC cohort\" width=\"700\"/>\n",
    "\n",
    "This Python package makes accessing CPTAC data easy with Python code and Jupyter notebooks. The package contains several tutorials which demonstrate data access and usage. This first tutorial serves as an introduction to the data to help users become familiar with what is included and how it is presented."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Data Overview\n",
    "\n",
    "Our package provides data access in a Python programming environment. If you have not installed Python or have not installed the package, see our installation documentation [here](https://paynelab.github.io/cptac/#installation).\n",
    "\n",
    "Once we have the package installed and we're in our Python environment, we begin by importing the package with a standard Python import statement:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import cptac"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "cptac data is broken down into datasets by cancer type. To view the available datasets, call the `cptac.get_cancer_options()` function:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['brca', 'ccrcc', 'coad', 'gbm', 'hnscc', 'lscc', 'luad', 'ov', 'pdac', 'ucec']"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cptac.get_cancer_options()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Data Availability\n",
    "The goals of CPTAC as a consortium include the broad and open dissemination of cancer proteogenomic data. The timing of a dataset's public release generally follows three stages: internal release to CPTAC investigators, public release with a publication embargo, and full public release. Each of the cancer types may be at a different data availability stage, depending on the date of data creation. In the Python `cptac` package, these three stages are dealt with as follows:\n",
    "\n",
    "**Internally released data** requires a password to download.\n",
    "\n",
    "**Embargoed release data** is publicly available, but prints an embargo statement every time you interact with the data.\n",
    "\n",
    "**Public data** is fully released without restrictions."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Exploring the data\n",
    "\n",
    "`cptac` allows you to load the dataset into a Python variable, and you can use that variable to access and work with the data. To load a particular dataset into a variable, type the name you want to give the variable, followed by `=`, and then type `cptac.` and the name of the dataset in [UpperCamelCase](https://en.wikipedia.org/wiki/Camel_case) followed by two parentheses, e.g. `cptac.Ucec()` or `cptac.Ccrcc()`:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "en = cptac.Ucec()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To see what data is available, use the `en.list_data_sources()` function. This displays the different types of data included in the dataset for this particular cancer type, each stored in a [pandas dataframe](https://pandas.pydata.org/pandas-docs/stable/getting_started/dsintro.html#dataframe)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Data type</th>\n",
       "      <th>Available sources</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CNV</td>\n",
       "      <td>washu</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>circular_RNA</td>\n",
       "      <td>bcm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>clinical</td>\n",
       "      <td>mssm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>deconvolution_cibersort</td>\n",
       "      <td>washu</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>deconvolution_xcell</td>\n",
       "      <td>washu</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>miRNA</td>\n",
       "      <td>washu</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>phosphoproteomics</td>\n",
       "      <td>umich</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>proteomics</td>\n",
       "      <td>umich</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>somatic_mutation</td>\n",
       "      <td>harmonized, washu</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>transcriptomics</td>\n",
       "      <td>bcm, broad, washu</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>tumor_purity</td>\n",
       "      <td>washu</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  Data type  Available sources\n",
       "0                       CNV              washu\n",
       "1              circular_RNA                bcm\n",
       "2                  clinical               mssm\n",
       "3   deconvolution_cibersort              washu\n",
       "4       deconvolution_xcell              washu\n",
       "5                     miRNA              washu\n",
       "6         phosphoproteomics              umich\n",
       "7                proteomics              umich\n",
       "8          somatic_mutation  harmonized, washu\n",
       "9           transcriptomics  bcm, broad, washu\n",
       "10             tumor_purity              washu"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "en.list_data_sources()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Molecular Omics\n",
    "\n",
    "You'll notice that some datatypes have more than one source, usually named after the organization which generated the data. Since not all of this data has been publicly released yet, we will use 'awg' data in these tutorials. AWG stands for \"All Working Groups\" and was generated collaboratively by many different groups, albeit with some inconsistencies between datasets.\n",
    "\n",
    "Data can be accessed through the `get_dataframe` function, or through one of several helper \"get\" functions. For example, we can look at the proteomics data by using `en.get_proteomics()`. This returns a [pandas dataframe](https://pandas.pydata.org/pandas-docs/stable/getting_started/dsintro.html#dataframe) containing the proteomic data. Each column in the proteomics dataframe is the quantitiative measurement for a particular protein. Each row in the proteomics dataframe is a sample of either a tumor or non-tumor from a cancer patient."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                                  \r",
      "Samples: ['C3L-00006', 'C3L-00008', 'C3L-00032', 'C3L-00084', 'C3L-00090', 'C3L-00098', 'C3L-00136', 'C3L-00137', 'C3L-00139', 'C3L-00143', 'C3L-00145', 'C3L-00156', 'C3L-00157', 'C3L-00161', 'C3L-00356', 'C3L-00358', 'C3L-00361', 'C3L-00362', 'C3L-00413', 'C3L-00449']\n",
      "Proteins: [('ARF5', 'ENSP00000000233.5'), ('M6PR', 'ENSP00000000412.3'), ('ESRRA', 'ENSP00000000442.6'), ('FKBP4', 'ENSP00000001008.4'), ('NDUFAF7', 'ENSP00000002125.4'), ('FUCA2', 'ENSP00000002165.5'), ('DBNDD1', 'ENSP00000002501.6'), ('SEMA3F', 'ENSP00000002829.3'), ('CFTR', 'ENSP00000003084.6'), ('CYP51A1', 'ENSP00000003100.8'), ('USP28', 'ENSP00000003302.4'), ('SLC7A2', 'ENSP00000004531.10'), ('HSPB6', 'ENSP00000004982.3'), ('PDK4', 'ENSP00000005178.5'), ('USH1C', 'ENSP00000005226.7'), ('RALA', 'ENSP00000005257.2'), ('BAIAP2L1', 'ENSP00000005260.8'), ('TMEM132A', 'ENSP00000005286.4'), ('DVL2', 'ENSP00000005340.4'), ('RPAP3', 'ENSP00000005386.3')]\n"
     ]
    }
   ],
   "source": [
    "# These two methods of getting proteomics data are functionally equivalent:\n",
    "# get_dataframe needs at least two arguments: datatype, and source\n",
    "proteomics = en.get_dataframe('proteomics', 'umich')\n",
    "# there is also a get function for each datatype, this is the same as get_dataframe but does not need the datatype argument\n",
    "proteomics = en.get_proteomics('umich')\n",
    "\n",
    "samples = proteomics.index\n",
    "proteins = proteomics.columns\n",
    "print(\"Samples:\",samples[0:20].tolist()) #the first twenty samples\n",
    "print(\"Proteins:\",proteins[0:20].tolist()) #the first twenty proteins"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Dataframe values\n",
    "\n",
    "Values in the dataframe are protein abundance values. Values that read \"NaN\" mean that particular sample from that patient had no data for that particular protein. For the awg endometrial CPTAC proteomics data, a TMT-reference channel strategy was used. A detailed description of this strategy can be found at [Nature Protocols](https://www.nature.com/articles/s41596-018-0006-9) and also at [PubMed Central](https://www.ncbi.nlm.nih.gov/pubmed/?term=29988108). This strategy ratios each sample's abundance to a pooled reference. The ratio is then log transformed. Therefore positive values indicate a measurement higher than the pooled reference; negative values are lower than the pooled reference."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr:last-of-type th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th>Name</th>\n",
       "      <th>ARF5</th>\n",
       "      <th>M6PR</th>\n",
       "      <th>ESRRA</th>\n",
       "      <th>FKBP4</th>\n",
       "      <th>NDUFAF7</th>\n",
       "      <th>FUCA2</th>\n",
       "      <th>DBNDD1</th>\n",
       "      <th>SEMA3F</th>\n",
       "      <th>CFTR</th>\n",
       "      <th>CYP51A1</th>\n",
       "      <th>...</th>\n",
       "      <th>SCRIB</th>\n",
       "      <th>WIZ</th>\n",
       "      <th>BPIFB4</th>\n",
       "      <th>LDB1</th>\n",
       "      <th>WIZ</th>\n",
       "      <th>TSGA10</th>\n",
       "      <th>RFX7</th>\n",
       "      <th>SWSAP1</th>\n",
       "      <th>MSANTD2</th>\n",
       "      <th>SVIL</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Database_ID</th>\n",
       "      <th>ENSP00000000233.5</th>\n",
       "      <th>ENSP00000000412.3</th>\n",
       "      <th>ENSP00000000442.6</th>\n",
       "      <th>ENSP00000001008.4</th>\n",
       "      <th>ENSP00000002125.4</th>\n",
       "      <th>ENSP00000002165.5</th>\n",
       "      <th>ENSP00000002501.6</th>\n",
       "      <th>ENSP00000002829.3</th>\n",
       "      <th>ENSP00000003084.6</th>\n",
       "      <th>ENSP00000003100.8</th>\n",
       "      <th>...</th>\n",
       "      <th>ENSP00000501177.1</th>\n",
       "      <th>ENSP00000501256.3</th>\n",
       "      <th>ENSP00000501266.1</th>\n",
       "      <th>ENSP00000501277.1</th>\n",
       "      <th>ENSP00000501300.1</th>\n",
       "      <th>ENSP00000501312.1</th>\n",
       "      <th>ENSP00000501317.1</th>\n",
       "      <th>ENSP00000501355.1</th>\n",
       "      <th>ENSP00000501466.1</th>\n",
       "      <th>ENSP00000501521.1</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient_ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>C3L-00006</th>\n",
       "      <td>-0.056513</td>\n",
       "      <td>0.016557</td>\n",
       "      <td>0.002569</td>\n",
       "      <td>0.389819</td>\n",
       "      <td>0.603610</td>\n",
       "      <td>-0.332543</td>\n",
       "      <td>-0.790426</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.822732</td>\n",
       "      <td>0.039134</td>\n",
       "      <td>...</td>\n",
       "      <td>0.161720</td>\n",
       "      <td>-0.884807</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.268247</td>\n",
       "      <td>0.125392</td>\n",
       "      <td>-0.880833</td>\n",
       "      <td>0.108554</td>\n",
       "      <td>0.107413</td>\n",
       "      <td>-0.085833</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C3L-00008</th>\n",
       "      <td>0.549959</td>\n",
       "      <td>-0.206129</td>\n",
       "      <td>0.905784</td>\n",
       "      <td>-0.303631</td>\n",
       "      <td>0.018767</td>\n",
       "      <td>0.503513</td>\n",
       "      <td>0.950955</td>\n",
       "      <td>0.080142</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.063213</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.054284</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.106450</td>\n",
       "      <td>0.380557</td>\n",
       "      <td>-0.756099</td>\n",
       "      <td>0.264611</td>\n",
       "      <td>0.044423</td>\n",
       "      <td>-0.248319</td>\n",
       "      <td>-1.206596</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C3L-00032</th>\n",
       "      <td>0.088681</td>\n",
       "      <td>-0.154447</td>\n",
       "      <td>-0.190515</td>\n",
       "      <td>0.170753</td>\n",
       "      <td>0.196356</td>\n",
       "      <td>0.544194</td>\n",
       "      <td>-0.179078</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.377405</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.086905</td>\n",
       "      <td>0.055991</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.021986</td>\n",
       "      <td>-0.229645</td>\n",
       "      <td>1.923986</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.176694</td>\n",
       "      <td>-0.332384</td>\n",
       "      <td>-1.330653</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C3L-00084</th>\n",
       "      <td>-0.846555</td>\n",
       "      <td>0.027740</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.178700</td>\n",
       "      <td>0.264054</td>\n",
       "      <td>-0.183548</td>\n",
       "      <td>0.077215</td>\n",
       "      <td>-0.247164</td>\n",
       "      <td>0.152277</td>\n",
       "      <td>-0.279549</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.125796</td>\n",
       "      <td>0.944212</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.917409</td>\n",
       "      <td>0.026862</td>\n",
       "      <td>-0.885976</td>\n",
       "      <td>-0.006510</td>\n",
       "      <td>-0.014162</td>\n",
       "      <td>0.365158</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C3L-00090</th>\n",
       "      <td>0.539019</td>\n",
       "      <td>0.956619</td>\n",
       "      <td>-0.039516</td>\n",
       "      <td>0.323656</td>\n",
       "      <td>0.064605</td>\n",
       "      <td>0.173433</td>\n",
       "      <td>-0.524325</td>\n",
       "      <td>-0.038590</td>\n",
       "      <td>-0.311486</td>\n",
       "      <td>0.309905</td>\n",
       "      <td>...</td>\n",
       "      <td>0.853362</td>\n",
       "      <td>-0.716947</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.286277</td>\n",
       "      <td>-0.046076</td>\n",
       "      <td>0.089645</td>\n",
       "      <td>-0.444506</td>\n",
       "      <td>-0.072531</td>\n",
       "      <td>-0.463495</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 12662 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Name                     ARF5              M6PR             ESRRA  \\\n",
       "Database_ID ENSP00000000233.5 ENSP00000000412.3 ENSP00000000442.6   \n",
       "Patient_ID                                                          \n",
       "C3L-00006           -0.056513          0.016557          0.002569   \n",
       "C3L-00008            0.549959         -0.206129          0.905784   \n",
       "C3L-00032            0.088681         -0.154447         -0.190515   \n",
       "C3L-00084           -0.846555          0.027740               NaN   \n",
       "C3L-00090            0.539019          0.956619         -0.039516   \n",
       "\n",
       "Name                    FKBP4           NDUFAF7             FUCA2  \\\n",
       "Database_ID ENSP00000001008.4 ENSP00000002125.4 ENSP00000002165.5   \n",
       "Patient_ID                                                          \n",
       "C3L-00006            0.389819          0.603610         -0.332543   \n",
       "C3L-00008           -0.303631          0.018767          0.503513   \n",
       "C3L-00032            0.170753          0.196356          0.544194   \n",
       "C3L-00084            0.178700          0.264054         -0.183548   \n",
       "C3L-00090            0.323656          0.064605          0.173433   \n",
       "\n",
       "Name                   DBNDD1            SEMA3F              CFTR  \\\n",
       "Database_ID ENSP00000002501.6 ENSP00000002829.3 ENSP00000003084.6   \n",
       "Patient_ID                                                          \n",
       "C3L-00006           -0.790426               NaN          0.822732   \n",
       "C3L-00008            0.950955          0.080142               NaN   \n",
       "C3L-00032           -0.179078               NaN               NaN   \n",
       "C3L-00084            0.077215         -0.247164          0.152277   \n",
       "C3L-00090           -0.524325         -0.038590         -0.311486   \n",
       "\n",
       "Name                  CYP51A1  ...             SCRIB               WIZ  \\\n",
       "Database_ID ENSP00000003100.8  ... ENSP00000501177.1 ENSP00000501256.3   \n",
       "Patient_ID                     ...                                       \n",
       "C3L-00006            0.039134  ...          0.161720         -0.884807   \n",
       "C3L-00008           -0.063213  ...               NaN          0.054284   \n",
       "C3L-00032            0.377405  ...         -1.086905          0.055991   \n",
       "C3L-00084           -0.279549  ...         -0.125796          0.944212   \n",
       "C3L-00090            0.309905  ...          0.853362         -0.716947   \n",
       "\n",
       "Name                   BPIFB4              LDB1               WIZ  \\\n",
       "Database_ID ENSP00000501266.1 ENSP00000501277.1 ENSP00000501300.1   \n",
       "Patient_ID                                                          \n",
       "C3L-00006                 NaN          0.268247          0.125392   \n",
       "C3L-00008                 NaN         -0.106450          0.380557   \n",
       "C3L-00032                 NaN         -0.021986         -0.229645   \n",
       "C3L-00084                 NaN          0.917409          0.026862   \n",
       "C3L-00090                 NaN         -0.286277         -0.046076   \n",
       "\n",
       "Name                   TSGA10              RFX7            SWSAP1  \\\n",
       "Database_ID ENSP00000501312.1 ENSP00000501317.1 ENSP00000501355.1   \n",
       "Patient_ID                                                          \n",
       "C3L-00006           -0.880833          0.108554          0.107413   \n",
       "C3L-00008           -0.756099          0.264611          0.044423   \n",
       "C3L-00032            1.923986               NaN         -0.176694   \n",
       "C3L-00084           -0.885976         -0.006510         -0.014162   \n",
       "C3L-00090            0.089645         -0.444506         -0.072531   \n",
       "\n",
       "Name                  MSANTD2              SVIL  \n",
       "Database_ID ENSP00000501466.1 ENSP00000501521.1  \n",
       "Patient_ID                                       \n",
       "C3L-00006           -0.085833               NaN  \n",
       "C3L-00008           -0.248319         -1.206596  \n",
       "C3L-00032           -0.332384         -1.330653  \n",
       "C3L-00084            0.365158               NaN  \n",
       "C3L-00090           -0.463495               NaN  \n",
       "\n",
       "[5 rows x 12662 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "proteomics.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "As seen in `en.list_data_sources()`, other omics data are also available (e.g. transcriptomics, copy number variation, phoshoproteomics).\n",
    "\n",
    "The transcriptomics looks almost identical to the proteomics data, available in a pandas dataframe with the same convention. Each set of samples is consistent, meaning samples found in the endometrial proteomics data will be the same samples in all other endometrial dataframes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                                     \r"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr:last-of-type th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th>Name</th>\n",
       "      <th>A1BG</th>\n",
       "      <th>A1BG-AS1</th>\n",
       "      <th>A1CF</th>\n",
       "      <th>A2M</th>\n",
       "      <th>A2M-AS1</th>\n",
       "      <th>A2ML1</th>\n",
       "      <th>A2ML1-AS1</th>\n",
       "      <th>A2ML1-AS2</th>\n",
       "      <th>A2MP1</th>\n",
       "      <th>A3GALT2</th>\n",
       "      <th>...</th>\n",
       "      <th>ZXDB</th>\n",
       "      <th>ZXDC</th>\n",
       "      <th>ZYG11A</th>\n",
       "      <th>ZYG11AP1</th>\n",
       "      <th>ZYG11B</th>\n",
       "      <th>ZYX</th>\n",
       "      <th>ZYXP1</th>\n",
       "      <th>ZZEF1</th>\n",
       "      <th>hsa-mir-1253</th>\n",
       "      <th>hsa-mir-423</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Database_ID</th>\n",
       "      <th>ENSG00000121410.12</th>\n",
       "      <th>ENSG00000268895.6</th>\n",
       "      <th>ENSG00000148584.15</th>\n",
       "      <th>ENSG00000175899.15</th>\n",
       "      <th>ENSG00000245105.4</th>\n",
       "      <th>ENSG00000166535.20</th>\n",
       "      <th>ENSG00000256661.1</th>\n",
       "      <th>ENSG00000256904.1</th>\n",
       "      <th>ENSG00000256069.7</th>\n",
       "      <th>ENSG00000184389.9</th>\n",
       "      <th>...</th>\n",
       "      <th>ENSG00000198455.4</th>\n",
       "      <th>ENSG00000070476.15</th>\n",
       "      <th>ENSG00000203995.10</th>\n",
       "      <th>ENSG00000232242.2</th>\n",
       "      <th>ENSG00000162378.13</th>\n",
       "      <th>ENSG00000159840.16</th>\n",
       "      <th>ENSG00000274572.1</th>\n",
       "      <th>ENSG00000074755.15</th>\n",
       "      <th>ENSG00000272920.1</th>\n",
       "      <th>ENSG00000266919.3</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient_ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>C3L-00006</th>\n",
       "      <td>2.54</td>\n",
       "      <td>5.11</td>\n",
       "      <td>3.60</td>\n",
       "      <td>13.75</td>\n",
       "      <td>6.45</td>\n",
       "      <td>7.08</td>\n",
       "      <td>1.80</td>\n",
       "      <td>0.00</td>\n",
       "      <td>2.60</td>\n",
       "      <td>1.16</td>\n",
       "      <td>...</td>\n",
       "      <td>10.17</td>\n",
       "      <td>10.61</td>\n",
       "      <td>5.54</td>\n",
       "      <td>0.0</td>\n",
       "      <td>11.85</td>\n",
       "      <td>10.60</td>\n",
       "      <td>0.0</td>\n",
       "      <td>11.87</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C3L-00008</th>\n",
       "      <td>4.40</td>\n",
       "      <td>4.63</td>\n",
       "      <td>5.49</td>\n",
       "      <td>13.89</td>\n",
       "      <td>6.61</td>\n",
       "      <td>6.97</td>\n",
       "      <td>0.00</td>\n",
       "      <td>2.74</td>\n",
       "      <td>3.25</td>\n",
       "      <td>0.00</td>\n",
       "      <td>...</td>\n",
       "      <td>9.79</td>\n",
       "      <td>10.48</td>\n",
       "      <td>7.79</td>\n",
       "      <td>0.0</td>\n",
       "      <td>12.28</td>\n",
       "      <td>11.28</td>\n",
       "      <td>0.0</td>\n",
       "      <td>11.93</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C3L-00032</th>\n",
       "      <td>4.83</td>\n",
       "      <td>7.26</td>\n",
       "      <td>3.73</td>\n",
       "      <td>14.48</td>\n",
       "      <td>6.91</td>\n",
       "      <td>9.56</td>\n",
       "      <td>0.98</td>\n",
       "      <td>0.00</td>\n",
       "      <td>3.26</td>\n",
       "      <td>0.00</td>\n",
       "      <td>...</td>\n",
       "      <td>9.43</td>\n",
       "      <td>9.97</td>\n",
       "      <td>6.48</td>\n",
       "      <td>0.0</td>\n",
       "      <td>11.72</td>\n",
       "      <td>10.37</td>\n",
       "      <td>0.0</td>\n",
       "      <td>11.70</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C3L-00084</th>\n",
       "      <td>4.73</td>\n",
       "      <td>6.01</td>\n",
       "      <td>5.37</td>\n",
       "      <td>15.17</td>\n",
       "      <td>7.93</td>\n",
       "      <td>3.86</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>3.73</td>\n",
       "      <td>1.15</td>\n",
       "      <td>...</td>\n",
       "      <td>9.23</td>\n",
       "      <td>10.37</td>\n",
       "      <td>7.47</td>\n",
       "      <td>0.0</td>\n",
       "      <td>11.86</td>\n",
       "      <td>10.13</td>\n",
       "      <td>0.0</td>\n",
       "      <td>11.19</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C3L-00090</th>\n",
       "      <td>4.14</td>\n",
       "      <td>6.24</td>\n",
       "      <td>5.69</td>\n",
       "      <td>13.87</td>\n",
       "      <td>6.79</td>\n",
       "      <td>4.32</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>3.23</td>\n",
       "      <td>0.00</td>\n",
       "      <td>...</td>\n",
       "      <td>9.69</td>\n",
       "      <td>9.64</td>\n",
       "      <td>7.60</td>\n",
       "      <td>0.0</td>\n",
       "      <td>11.98</td>\n",
       "      <td>10.31</td>\n",
       "      <td>0.0</td>\n",
       "      <td>11.45</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 59286 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Name                      A1BG          A1BG-AS1               A1CF  \\\n",
       "Database_ID ENSG00000121410.12 ENSG00000268895.6 ENSG00000148584.15   \n",
       "Patient_ID                                                            \n",
       "C3L-00006                 2.54              5.11               3.60   \n",
       "C3L-00008                 4.40              4.63               5.49   \n",
       "C3L-00032                 4.83              7.26               3.73   \n",
       "C3L-00084                 4.73              6.01               5.37   \n",
       "C3L-00090                 4.14              6.24               5.69   \n",
       "\n",
       "Name                       A2M           A2M-AS1              A2ML1  \\\n",
       "Database_ID ENSG00000175899.15 ENSG00000245105.4 ENSG00000166535.20   \n",
       "Patient_ID                                                            \n",
       "C3L-00006                13.75              6.45               7.08   \n",
       "C3L-00008                13.89              6.61               6.97   \n",
       "C3L-00032                14.48              6.91               9.56   \n",
       "C3L-00084                15.17              7.93               3.86   \n",
       "C3L-00090                13.87              6.79               4.32   \n",
       "\n",
       "Name                A2ML1-AS1         A2ML1-AS2             A2MP1  \\\n",
       "Database_ID ENSG00000256661.1 ENSG00000256904.1 ENSG00000256069.7   \n",
       "Patient_ID                                                          \n",
       "C3L-00006                1.80              0.00              2.60   \n",
       "C3L-00008                0.00              2.74              3.25   \n",
       "C3L-00032                0.98              0.00              3.26   \n",
       "C3L-00084                0.00              0.00              3.73   \n",
       "C3L-00090                0.00              0.00              3.23   \n",
       "\n",
       "Name                  A3GALT2  ...              ZXDB               ZXDC  \\\n",
       "Database_ID ENSG00000184389.9  ... ENSG00000198455.4 ENSG00000070476.15   \n",
       "Patient_ID                     ...                                        \n",
       "C3L-00006                1.16  ...             10.17              10.61   \n",
       "C3L-00008                0.00  ...              9.79              10.48   \n",
       "C3L-00032                0.00  ...              9.43               9.97   \n",
       "C3L-00084                1.15  ...              9.23              10.37   \n",
       "C3L-00090                0.00  ...              9.69               9.64   \n",
       "\n",
       "Name                    ZYG11A          ZYG11AP1             ZYG11B  \\\n",
       "Database_ID ENSG00000203995.10 ENSG00000232242.2 ENSG00000162378.13   \n",
       "Patient_ID                                                            \n",
       "C3L-00006                 5.54               0.0              11.85   \n",
       "C3L-00008                 7.79               0.0              12.28   \n",
       "C3L-00032                 6.48               0.0              11.72   \n",
       "C3L-00084                 7.47               0.0              11.86   \n",
       "C3L-00090                 7.60               0.0              11.98   \n",
       "\n",
       "Name                       ZYX             ZYXP1              ZZEF1  \\\n",
       "Database_ID ENSG00000159840.16 ENSG00000274572.1 ENSG00000074755.15   \n",
       "Patient_ID                                                            \n",
       "C3L-00006                10.60               0.0              11.87   \n",
       "C3L-00008                11.28               0.0              11.93   \n",
       "C3L-00032                10.37               0.0              11.70   \n",
       "C3L-00084                10.13               0.0              11.19   \n",
       "C3L-00090                10.31               0.0              11.45   \n",
       "\n",
       "Name             hsa-mir-1253       hsa-mir-423  \n",
       "Database_ID ENSG00000272920.1 ENSG00000266919.3  \n",
       "Patient_ID                                       \n",
       "C3L-00006                 0.0               0.0  \n",
       "C3L-00008                 0.0               0.0  \n",
       "C3L-00032                 0.0               0.0  \n",
       "C3L-00084                 0.0               0.0  \n",
       "C3L-00090                 0.0               0.0  \n",
       "\n",
       "[5 rows x 59286 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "transcriptomics = en.get_transcriptomics('bcm')\n",
    "transcriptomics.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Clinical Data\n",
    "\n",
    "The clinical dataframe lists clinical information for the patient associated with each sample (e.g. age, race, diabetes status, tumor size). "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                               \r"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "cptac warning: Using source mssm for clinical data as no other sources provide this data. To remove this warning, pass mssm as the source parameter. (/var/folders/n2/95tlbmw5173fm8yjh4ffxggr0000gn/T/ipykernel_82825/241189121.py, line 1)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Name</th>\n",
       "      <th>tumor_code</th>\n",
       "      <th>discovery_study</th>\n",
       "      <th>discovery_study/type_of_analyzed_samples</th>\n",
       "      <th>consent/age</th>\n",
       "      <th>consent/sex</th>\n",
       "      <th>consent/race</th>\n",
       "      <th>consent/ethnicity</th>\n",
       "      <th>consent/ethnicity_race_ancestry_identified</th>\n",
       "      <th>consent/collection_in_us</th>\n",
       "      <th>consent/participant_country</th>\n",
       "      <th>...</th>\n",
       "      <th>follow-up/additional_treatment_for_new_tumor_radiation</th>\n",
       "      <th>follow-up/additional_treatment_for_new_tumor_pharmaceutical</th>\n",
       "      <th>follow-up/additional_treatment_for_new_tumor_immunological</th>\n",
       "      <th>follow-up/days_from_date_of_initial_pathologic_diagnosis_to_date_of_additional_surgery_for_new_tumor</th>\n",
       "      <th>follow-up/cause_of_death</th>\n",
       "      <th>follow-up/days_from_date_of_initial_pathologic_diagnosis_to_date_of_death</th>\n",
       "      <th>Recurrence-free survival</th>\n",
       "      <th>Overall survial</th>\n",
       "      <th>Recurrence status (1, yes; 0, no)</th>\n",
       "      <th>Survial status (1, dead; 0, alive)</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient_ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>C3L-00006</th>\n",
       "      <td>UCEC</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Tumor_and_Normal</td>\n",
       "      <td>64</td>\n",
       "      <td>Female</td>\n",
       "      <td>White</td>\n",
       "      <td>Not-Hispanic or Latino</td>\n",
       "      <td>White</td>\n",
       "      <td>Yes</td>\n",
       "      <td>United States</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>737.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C3L-00008</th>\n",
       "      <td>UCEC</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>58</td>\n",
       "      <td>Female</td>\n",
       "      <td>White</td>\n",
       "      <td>Not-Hispanic or Latino</td>\n",
       "      <td>White</td>\n",
       "      <td>Yes</td>\n",
       "      <td>United States</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>898.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C3L-00032</th>\n",
       "      <td>UCEC</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>50</td>\n",
       "      <td>Female</td>\n",
       "      <td>White</td>\n",
       "      <td>Not-Hispanic or Latino</td>\n",
       "      <td>White</td>\n",
       "      <td>Yes</td>\n",
       "      <td>United States</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1275.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C3L-00084</th>\n",
       "      <td>UCEC</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>74</td>\n",
       "      <td>Female</td>\n",
       "      <td>White</td>\n",
       "      <td>Not-Hispanic or Latino</td>\n",
       "      <td>White</td>\n",
       "      <td>Yes</td>\n",
       "      <td>United States</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>335.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C3L-00090</th>\n",
       "      <td>UCEC</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>75</td>\n",
       "      <td>Female</td>\n",
       "      <td>White</td>\n",
       "      <td>Not-Hispanic or Latino</td>\n",
       "      <td>White</td>\n",
       "      <td>Yes</td>\n",
       "      <td>United States</td>\n",
       "      <td>...</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Yes</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Other reasons</td>\n",
       "      <td>1281.0</td>\n",
       "      <td>50.0</td>\n",
       "      <td>1281.0</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 113 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Name       tumor_code discovery_study  \\\n",
       "Patient_ID                              \n",
       "C3L-00006        UCEC             Yes   \n",
       "C3L-00008        UCEC             Yes   \n",
       "C3L-00032        UCEC             Yes   \n",
       "C3L-00084        UCEC             Yes   \n",
       "C3L-00090        UCEC             Yes   \n",
       "\n",
       "Name       discovery_study/type_of_analyzed_samples consent/age consent/sex  \\\n",
       "Patient_ID                                                                    \n",
       "C3L-00006                          Tumor_and_Normal          64      Female   \n",
       "C3L-00008                                     Tumor          58      Female   \n",
       "C3L-00032                                     Tumor          50      Female   \n",
       "C3L-00084                                     Tumor          74      Female   \n",
       "C3L-00090                                     Tumor          75      Female   \n",
       "\n",
       "Name       consent/race       consent/ethnicity  \\\n",
       "Patient_ID                                        \n",
       "C3L-00006         White  Not-Hispanic or Latino   \n",
       "C3L-00008         White  Not-Hispanic or Latino   \n",
       "C3L-00032         White  Not-Hispanic or Latino   \n",
       "C3L-00084         White  Not-Hispanic or Latino   \n",
       "C3L-00090         White  Not-Hispanic or Latino   \n",
       "\n",
       "Name       consent/ethnicity_race_ancestry_identified  \\\n",
       "Patient_ID                                              \n",
       "C3L-00006                                       White   \n",
       "C3L-00008                                       White   \n",
       "C3L-00032                                       White   \n",
       "C3L-00084                                       White   \n",
       "C3L-00090                                       White   \n",
       "\n",
       "Name       consent/collection_in_us consent/participant_country  ...  \\\n",
       "Patient_ID                                                       ...   \n",
       "C3L-00006                       Yes               United States  ...   \n",
       "C3L-00008                       Yes               United States  ...   \n",
       "C3L-00032                       Yes               United States  ...   \n",
       "C3L-00084                       Yes               United States  ...   \n",
       "C3L-00090                       Yes               United States  ...   \n",
       "\n",
       "Name       follow-up/additional_treatment_for_new_tumor_radiation  \\\n",
       "Patient_ID                                                          \n",
       "C3L-00006                                                 NaN       \n",
       "C3L-00008                                                 NaN       \n",
       "C3L-00032                                                 NaN       \n",
       "C3L-00084                                                 NaN       \n",
       "C3L-00090                                                 Yes       \n",
       "\n",
       "Name       follow-up/additional_treatment_for_new_tumor_pharmaceutical  \\\n",
       "Patient_ID                                                               \n",
       "C3L-00006                                                 NaN            \n",
       "C3L-00008                                                 NaN            \n",
       "C3L-00032                                                 NaN            \n",
       "C3L-00084                                                 NaN            \n",
       "C3L-00090                                                 Yes            \n",
       "\n",
       "Name       follow-up/additional_treatment_for_new_tumor_immunological  \\\n",
       "Patient_ID                                                              \n",
       "C3L-00006                                                 NaN           \n",
       "C3L-00008                                                 NaN           \n",
       "C3L-00032                                                 NaN           \n",
       "C3L-00084                                                 NaN           \n",
       "C3L-00090                                                  No           \n",
       "\n",
       "Name       follow-up/days_from_date_of_initial_pathologic_diagnosis_to_date_of_additional_surgery_for_new_tumor  \\\n",
       "Patient_ID                                                                                                        \n",
       "C3L-00006                                                 NaN                                                     \n",
       "C3L-00008                                                 NaN                                                     \n",
       "C3L-00032                                                 NaN                                                     \n",
       "C3L-00084                                                 NaN                                                     \n",
       "C3L-00090                                                 NaN                                                     \n",
       "\n",
       "Name       follow-up/cause_of_death  \\\n",
       "Patient_ID                            \n",
       "C3L-00006                       NaN   \n",
       "C3L-00008                       NaN   \n",
       "C3L-00032                       NaN   \n",
       "C3L-00084                       NaN   \n",
       "C3L-00090             Other reasons   \n",
       "\n",
       "Name       follow-up/days_from_date_of_initial_pathologic_diagnosis_to_date_of_death  \\\n",
       "Patient_ID                                                                             \n",
       "C3L-00006                                                 NaN                          \n",
       "C3L-00008                                                 NaN                          \n",
       "C3L-00032                                                 NaN                          \n",
       "C3L-00084                                                 NaN                          \n",
       "C3L-00090                                              1281.0                          \n",
       "\n",
       "Name       Recurrence-free survival Overall survial  \\\n",
       "Patient_ID                                            \n",
       "C3L-00006                       NaN           737.0   \n",
       "C3L-00008                       NaN           898.0   \n",
       "C3L-00032                       NaN          1275.0   \n",
       "C3L-00084                       NaN           335.0   \n",
       "C3L-00090                      50.0          1281.0   \n",
       "\n",
       "Name       Recurrence status (1, yes; 0, no)  \\\n",
       "Patient_ID                                     \n",
       "C3L-00006                                  0   \n",
       "C3L-00008                                  0   \n",
       "C3L-00032                                  0   \n",
       "C3L-00084                                  0   \n",
       "C3L-00090                                  1   \n",
       "\n",
       "Name       Survial status (1, dead; 0, alive)  \n",
       "Patient_ID                                     \n",
       "C3L-00006                                 0.0  \n",
       "C3L-00008                                 0.0  \n",
       "C3L-00032                                 0.0  \n",
       "C3L-00084                                 0.0  \n",
       "C3L-00090                                 1.0  \n",
       "\n",
       "[5 rows x 113 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clinical = en.get_clinical()\n",
    "clinical.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In addition to donating a tumor sample, some patients also had a normal sample taken for control and comparison. We can identify these samples by looking for samples marked \"Tumor_and_Normal\" in the \"type_of_analyzed_samples\" column. For example, patient C3L-00006 provided both a tumor sample and a normal sample. Since the data in these cases is for both tumor and normal samples, where a column would have identical values for both samples from the same patient (e.g., patient age and gender), the information is recorded only for the tumor sample."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Name</th>\n",
       "      <th>tumor_code</th>\n",
       "      <th>discovery_study</th>\n",
       "      <th>discovery_study/type_of_analyzed_samples</th>\n",
       "      <th>consent/age</th>\n",
       "      <th>consent/sex</th>\n",
       "      <th>consent/race</th>\n",
       "      <th>consent/ethnicity</th>\n",
       "      <th>consent/ethnicity_race_ancestry_identified</th>\n",
       "      <th>consent/collection_in_us</th>\n",
       "      <th>consent/participant_country</th>\n",
       "      <th>...</th>\n",
       "      <th>follow-up/additional_treatment_for_new_tumor_radiation</th>\n",
       "      <th>follow-up/additional_treatment_for_new_tumor_pharmaceutical</th>\n",
       "      <th>follow-up/additional_treatment_for_new_tumor_immunological</th>\n",
       "      <th>follow-up/days_from_date_of_initial_pathologic_diagnosis_to_date_of_additional_surgery_for_new_tumor</th>\n",
       "      <th>follow-up/cause_of_death</th>\n",
       "      <th>follow-up/days_from_date_of_initial_pathologic_diagnosis_to_date_of_death</th>\n",
       "      <th>Recurrence-free survival</th>\n",
       "      <th>Overall survial</th>\n",
       "      <th>Recurrence status (1, yes; 0, no)</th>\n",
       "      <th>Survial status (1, dead; 0, alive)</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient_ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>C3L-00006</th>\n",
       "      <td>UCEC</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Tumor_and_Normal</td>\n",
       "      <td>64</td>\n",
       "      <td>Female</td>\n",
       "      <td>White</td>\n",
       "      <td>Not-Hispanic or Latino</td>\n",
       "      <td>White</td>\n",
       "      <td>Yes</td>\n",
       "      <td>United States</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>737.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C3L-00361</th>\n",
       "      <td>UCEC</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Tumor_and_Normal</td>\n",
       "      <td>64</td>\n",
       "      <td>Female</td>\n",
       "      <td>White</td>\n",
       "      <td>Not-Hispanic or Latino</td>\n",
       "      <td>WHITE</td>\n",
       "      <td>Yes</td>\n",
       "      <td>United States</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1436.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C3L-01246</th>\n",
       "      <td>UCEC</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Tumor_and_Normal</td>\n",
       "      <td>62</td>\n",
       "      <td>Female</td>\n",
       "      <td>White</td>\n",
       "      <td>Not reported</td>\n",
       "      <td>Medical record does not state.</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Other: Medical record does not state.</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1132.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 113 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Name       tumor_code discovery_study  \\\n",
       "Patient_ID                              \n",
       "C3L-00006        UCEC             Yes   \n",
       "C3L-00361        UCEC             Yes   \n",
       "C3L-01246        UCEC             Yes   \n",
       "\n",
       "Name       discovery_study/type_of_analyzed_samples consent/age consent/sex  \\\n",
       "Patient_ID                                                                    \n",
       "C3L-00006                          Tumor_and_Normal          64      Female   \n",
       "C3L-00361                          Tumor_and_Normal          64      Female   \n",
       "C3L-01246                          Tumor_and_Normal          62      Female   \n",
       "\n",
       "Name       consent/race       consent/ethnicity  \\\n",
       "Patient_ID                                        \n",
       "C3L-00006         White  Not-Hispanic or Latino   \n",
       "C3L-00361         White  Not-Hispanic or Latino   \n",
       "C3L-01246         White            Not reported   \n",
       "\n",
       "Name       consent/ethnicity_race_ancestry_identified  \\\n",
       "Patient_ID                                              \n",
       "C3L-00006                                       White   \n",
       "C3L-00361                                       WHITE   \n",
       "C3L-01246              Medical record does not state.   \n",
       "\n",
       "Name       consent/collection_in_us            consent/participant_country  \\\n",
       "Patient_ID                                                                   \n",
       "C3L-00006                       Yes                          United States   \n",
       "C3L-00361                       Yes                          United States   \n",
       "C3L-01246                       Yes  Other: Medical record does not state.   \n",
       "\n",
       "Name        ... follow-up/additional_treatment_for_new_tumor_radiation  \\\n",
       "Patient_ID  ...                                                          \n",
       "C3L-00006   ...                                                NaN       \n",
       "C3L-00361   ...                                                NaN       \n",
       "C3L-01246   ...                                                NaN       \n",
       "\n",
       "Name       follow-up/additional_treatment_for_new_tumor_pharmaceutical  \\\n",
       "Patient_ID                                                               \n",
       "C3L-00006                                                 NaN            \n",
       "C3L-00361                                                 NaN            \n",
       "C3L-01246                                                 NaN            \n",
       "\n",
       "Name       follow-up/additional_treatment_for_new_tumor_immunological  \\\n",
       "Patient_ID                                                              \n",
       "C3L-00006                                                 NaN           \n",
       "C3L-00361                                                 NaN           \n",
       "C3L-01246                                                 NaN           \n",
       "\n",
       "Name       follow-up/days_from_date_of_initial_pathologic_diagnosis_to_date_of_additional_surgery_for_new_tumor  \\\n",
       "Patient_ID                                                                                                        \n",
       "C3L-00006                                                 NaN                                                     \n",
       "C3L-00361                                                 NaN                                                     \n",
       "C3L-01246                                                 NaN                                                     \n",
       "\n",
       "Name       follow-up/cause_of_death  \\\n",
       "Patient_ID                            \n",
       "C3L-00006                       NaN   \n",
       "C3L-00361                       NaN   \n",
       "C3L-01246                       NaN   \n",
       "\n",
       "Name       follow-up/days_from_date_of_initial_pathologic_diagnosis_to_date_of_death  \\\n",
       "Patient_ID                                                                             \n",
       "C3L-00006                                                 NaN                          \n",
       "C3L-00361                                                 NaN                          \n",
       "C3L-01246                                                 NaN                          \n",
       "\n",
       "Name       Recurrence-free survival Overall survial  \\\n",
       "Patient_ID                                            \n",
       "C3L-00006                       NaN           737.0   \n",
       "C3L-00361                       NaN          1436.0   \n",
       "C3L-01246                       NaN          1132.0   \n",
       "\n",
       "Name       Recurrence status (1, yes; 0, no)  \\\n",
       "Patient_ID                                     \n",
       "C3L-00006                                  0   \n",
       "C3L-00361                                  0   \n",
       "C3L-01246                                  0   \n",
       "\n",
       "Name       Survial status (1, dead; 0, alive)  \n",
       "Patient_ID                                     \n",
       "C3L-00006                                 0.0  \n",
       "C3L-00361                                 0.0  \n",
       "C3L-01246                                 0.0  \n",
       "\n",
       "[3 rows x 113 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clinical.loc[[\"C3L-00006\",\"C3L-00361\",\"C3L-01246\"]]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Mutation data\n",
    "\n",
    "Each cancer dataset contains mutation data for the cohort. The data consists of all somatic mutations found for each sample (meaning there will be many lines for each sample). Each row lists the specific gene that was mutated, the type of mutation, and the location of the mutation. This data is a direct import of a MAF file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "cptac warning: Your version of cptac (1.5.0rc1) is out-of-date. Latest is 1.1.2. Please run 'pip install --upgrade cptac' to update it. (/Users/blakemcgee/opt/anaconda3/lib/python3.9/threading.py, line 917)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                                             \r"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Name</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Mutation</th>\n",
       "      <th>Location</th>\n",
       "      <th>Entrez_Gene_Id</th>\n",
       "      <th>NCBI_Build</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start_Position</th>\n",
       "      <th>End_Position</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Variant_Type</th>\n",
       "      <th>...</th>\n",
       "      <th>HGNC_UniProt_ID(supplied_by_UniProt)</th>\n",
       "      <th>HGNC_Ensembl_ID(supplied_by_Ensembl)</th>\n",
       "      <th>HGNC_UCSC_ID(supplied_by_UCSC)</th>\n",
       "      <th>Oreganno_Build</th>\n",
       "      <th>Simple_Uniprot_alt_uniprot_accessions</th>\n",
       "      <th>dbSNP_TOPMED</th>\n",
       "      <th>HGNC_Entrez_Gene_ID(supplied_by_NCBI)</th>\n",
       "      <th>COHORT</th>\n",
       "      <th>getz</th>\n",
       "      <th>washu</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient_ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>C3L-00006</th>\n",
       "      <td>CPEB3</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>p.R595C</td>\n",
       "      <td>22849.0</td>\n",
       "      <td>hg38</td>\n",
       "      <td>chr10</td>\n",
       "      <td>92081406</td>\n",
       "      <td>92081406</td>\n",
       "      <td>+</td>\n",
       "      <td>SNP</td>\n",
       "      <td>...</td>\n",
       "      <td>Q8NE35</td>\n",
       "      <td>ENSG00000107864</td>\n",
       "      <td>uc001khw.3</td>\n",
       "      <td>hg38</td>\n",
       "      <td>Q5T389|Q9NQJ7|Q9Y2E9</td>\n",
       "      <td>NaN</td>\n",
       "      <td>22849.0</td>\n",
       "      <td>UCEC</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C3L-00006</th>\n",
       "      <td>SHANK2</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>p.S1692R</td>\n",
       "      <td>22941.0</td>\n",
       "      <td>hg38</td>\n",
       "      <td>chr11</td>\n",
       "      <td>70473343</td>\n",
       "      <td>70473343</td>\n",
       "      <td>+</td>\n",
       "      <td>SNP</td>\n",
       "      <td>...</td>\n",
       "      <td>Q9UPX8</td>\n",
       "      <td>ENSG00000162105</td>\n",
       "      <td>uc058etp.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C0SPG8|C0SPG9|Q3Y8G9|Q52LK2|Q9UKP1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>22941.0</td>\n",
       "      <td>UCEC</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C3L-00006</th>\n",
       "      <td>CTTN</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>p.R119C</td>\n",
       "      <td>2017.0</td>\n",
       "      <td>hg38</td>\n",
       "      <td>chr11</td>\n",
       "      <td>70414605</td>\n",
       "      <td>70414605</td>\n",
       "      <td>+</td>\n",
       "      <td>SNP</td>\n",
       "      <td>...</td>\n",
       "      <td>Q14247</td>\n",
       "      <td>ENSG00000085733</td>\n",
       "      <td>uc001opw.5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Q8N707|Q96H99</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2017.0</td>\n",
       "      <td>UCEC</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C3L-00006</th>\n",
       "      <td>TPCN2</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>p.V572I</td>\n",
       "      <td>219931.0</td>\n",
       "      <td>hg38</td>\n",
       "      <td>chr11</td>\n",
       "      <td>69083969</td>\n",
       "      <td>69083969</td>\n",
       "      <td>+</td>\n",
       "      <td>SNP</td>\n",
       "      <td>...</td>\n",
       "      <td>Q8NHX9</td>\n",
       "      <td>ENSG00000162341</td>\n",
       "      <td>uc001oos.3</td>\n",
       "      <td>hg38</td>\n",
       "      <td>Q9NT82</td>\n",
       "      <td>NaN</td>\n",
       "      <td>219931.0</td>\n",
       "      <td>UCEC</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C3L-00006</th>\n",
       "      <td>PITPNM1</td>\n",
       "      <td>Silent</td>\n",
       "      <td>p.R478R</td>\n",
       "      <td>9600.0</td>\n",
       "      <td>hg38</td>\n",
       "      <td>chr11</td>\n",
       "      <td>67498646</td>\n",
       "      <td>67498646</td>\n",
       "      <td>+</td>\n",
       "      <td>SNP</td>\n",
       "      <td>...</td>\n",
       "      <td>O00562</td>\n",
       "      <td>ENSG00000110697</td>\n",
       "      <td>uc001oly.4</td>\n",
       "      <td>hg38</td>\n",
       "      <td>A6NME4|Q6T7X3|Q8TBN3|Q9BZ73</td>\n",
       "      <td>NaN</td>\n",
       "      <td>9600.0</td>\n",
       "      <td>UCEC</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 107 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Name           Gene           Mutation  Location  Entrez_Gene_Id NCBI_Build  \\\n",
       "Patient_ID                                                                    \n",
       "C3L-00006     CPEB3  Missense_Mutation   p.R595C         22849.0       hg38   \n",
       "C3L-00006    SHANK2  Missense_Mutation  p.S1692R         22941.0       hg38   \n",
       "C3L-00006      CTTN  Missense_Mutation   p.R119C          2017.0       hg38   \n",
       "C3L-00006     TPCN2  Missense_Mutation   p.V572I        219931.0       hg38   \n",
       "C3L-00006   PITPNM1             Silent   p.R478R          9600.0       hg38   \n",
       "\n",
       "Name       Chromosome  Start_Position  End_Position Strand Variant_Type  ...  \\\n",
       "Patient_ID                                                               ...   \n",
       "C3L-00006       chr10        92081406      92081406      +          SNP  ...   \n",
       "C3L-00006       chr11        70473343      70473343      +          SNP  ...   \n",
       "C3L-00006       chr11        70414605      70414605      +          SNP  ...   \n",
       "C3L-00006       chr11        69083969      69083969      +          SNP  ...   \n",
       "C3L-00006       chr11        67498646      67498646      +          SNP  ...   \n",
       "\n",
       "Name       HGNC_UniProt_ID(supplied_by_UniProt)  \\\n",
       "Patient_ID                                        \n",
       "C3L-00006                                Q8NE35   \n",
       "C3L-00006                                Q9UPX8   \n",
       "C3L-00006                                Q14247   \n",
       "C3L-00006                                Q8NHX9   \n",
       "C3L-00006                                O00562   \n",
       "\n",
       "Name       HGNC_Ensembl_ID(supplied_by_Ensembl)  \\\n",
       "Patient_ID                                        \n",
       "C3L-00006                       ENSG00000107864   \n",
       "C3L-00006                       ENSG00000162105   \n",
       "C3L-00006                       ENSG00000085733   \n",
       "C3L-00006                       ENSG00000162341   \n",
       "C3L-00006                       ENSG00000110697   \n",
       "\n",
       "Name       HGNC_UCSC_ID(supplied_by_UCSC) Oreganno_Build  \\\n",
       "Patient_ID                                                 \n",
       "C3L-00006                      uc001khw.3           hg38   \n",
       "C3L-00006                      uc058etp.1            NaN   \n",
       "C3L-00006                      uc001opw.5            NaN   \n",
       "C3L-00006                      uc001oos.3           hg38   \n",
       "C3L-00006                      uc001oly.4           hg38   \n",
       "\n",
       "Name       Simple_Uniprot_alt_uniprot_accessions dbSNP_TOPMED  \\\n",
       "Patient_ID                                                      \n",
       "C3L-00006                   Q5T389|Q9NQJ7|Q9Y2E9          NaN   \n",
       "C3L-00006     C0SPG8|C0SPG9|Q3Y8G9|Q52LK2|Q9UKP1          NaN   \n",
       "C3L-00006                          Q8N707|Q96H99          NaN   \n",
       "C3L-00006                                 Q9NT82          NaN   \n",
       "C3L-00006            A6NME4|Q6T7X3|Q8TBN3|Q9BZ73          NaN   \n",
       "\n",
       "Name        HGNC_Entrez_Gene_ID(supplied_by_NCBI)  COHORT  getz washu  \n",
       "Patient_ID                                                             \n",
       "C3L-00006                                 22849.0    UCEC  True   NaN  \n",
       "C3L-00006                                 22941.0    UCEC  True  True  \n",
       "C3L-00006                                  2017.0    UCEC  True   NaN  \n",
       "C3L-00006                                219931.0    UCEC  True  True  \n",
       "C3L-00006                                  9600.0    UCEC  True  True  \n",
       "\n",
       "[5 rows x 107 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "somatic_mutations = en.get_somatic_mutation('harmonized')\n",
    "somatic_mutations.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Exporting dataframes\n",
    "\n",
    "If you wish to export a dataframe to a file, simply call the dataframe's `to_csv` method, passing the path you wish to save the file to, and the value separator you want:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "clinical = en.get_clinical('mssm')\n",
    "clinical.to_csv(path_or_buf=\"clinical_dataframe.tsv\", sep='\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Downloading data\n",
    "\n",
    "The cptac package stores the data files for each dataset on a remote server. When you first install cptac, you will have no data files. Data files will be automatically downloaded the first time you try to use them. If you won't have internet access and need to download the files beforehand, that can be done with the `cptac.download` function:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Downloading Report_abundance_groupby=protein_protNorm=MD_gu=2.tsv: 100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████| 4/4 [01:37<00:00, 24.34s/chunk]\n",
      "Downloading Report_abundance_groupby=protein_protNorm=MD_gu=2.tsv: 100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████| 4/4 [01:39<00:00, 25.00s/chunk]\n",
      "Downloading Report_abundance_groupby=protein_protNorm=MD_gu=2.tsv: 100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████| 4/4 [00:42<00:00, 10.75s/chunk]\n",
      "Downloading Report_abundance_groupby=protein_protNorm=MD_gu=2.tsv: 100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████| 4/4 [00:45<00:00, 11.47s/chunk]\n",
      "Downloading Report_abundance_groupby=protein_protNorm=MD_gu=2.tsv: 100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████| 4/4 [00:58<00:00, 14.65s/chunk]\n",
      "Downloading Report_abundance_groupby=protein_protNorm=MD_gu=2.tsv: 100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████| 4/4 [00:58<00:00, 14.72s/chunk]\n",
      "Downloading Report_abundance_groupby=protein_protNorm=MD_gu=2.tsv: 100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████| 4/4 [00:52<00:00, 13.22s/chunk]\n",
      "Downloading Report_abundance_groupby=protein_protNorm=MD_gu=2.tsv: 100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████| 4/4 [00:31<00:00,  7.75s/chunk]\n",
      "Downloading Report_abundance_groupby=protein_protNorm=MD_gu=2.tsv: 100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████| 4/4 [01:08<00:00, 17.18s/chunk]\n",
      "Downloading Report_abundance_groupby=protein_protNorm=MD_gu=2.tsv: 100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████| 4/4 [00:56<00:00, 14.03s/chunk]\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cptac.download({\"umich\": [\"proteomics\"]})"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Getting help with a dataset or function\n",
    "\n",
    "To view the documentation for a dataset, pass it to the Python `help` function, e.g. `help(en)`. You can also view the documentation for just a specific function: `help(en.join_omics_to_omics)`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Help on method join_omics_to_omics in module cptac.cancers.cancer:\n",
      "\n",
      "join_omics_to_omics(df1_name: str, df2_name: str, df1_source: str = None, df2_source: str = None, genes1: str = None, genes2: str = None, how: str = 'outer', quiet: bool = False, tissue_type: str = 'both') -> pandas.core.frame.DataFrame method of cptac.cancers.ucec.Ucec instance\n",
      "    Take specified column(s) from one omics dataframe, and join to specified columns(s) from another omics dataframe. Intersection (inner join) of indices is used.\n",
      "    \n",
      "    Parameters:\n",
      "    df1_name (str): Name of first omics dataframe to select columns from.\n",
      "    df2_name (str): Name of second omics dataframe to select columns from.\n",
      "    df1_source (str, optional): Name of source for the first omics dataframe.\n",
      "    df2_source (str, optional): Name of source for the second omics dataframe.\n",
      "    genes1 (str, or list or array-like of str, optional): Gene(s) for column(s) to select from df1_name. str if one key, list or array-like of str if multiple.\n",
      "        Default of None will select entire dataframe.\n",
      "    genes2 (str, or list or array-like of str, optional): Gene(s) for Column(s) to select from df2_name. str if one key, list or array-like of str if multiple.\n",
      "        Default of None will select entire dataframe.\n",
      "    how (str, optional): How to perform the join, acceptable values are from ['outer', 'inner', 'left', 'right']. Defaults to 'outer'.\n",
      "    quiet (bool, optional): Whether to warn when inserting NaNs. Defaults to False.\n",
      "    tissue_type (str): Acceptable values in [\"tumor\",\"normal\",\"both\"]. Specifies the desired tissue type for the dataframe. Defaults to \"both\".\n",
      "    \n",
      "    Returns:\n",
      "    pandas.DataFrame: The selected columns from the two omics dataframes, joined into one dataframe.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "help(en.join_omics_to_omics)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
