---
title: "Using `reticulate` to access CPTAC data in R"
output:
  html_document:
    df_print: paged
urlcolor: blue
---
`reticulate` (https://rstudio.github.io/reticulate/) is an R package that allows you to use R create Python objects and convert them into R objects. You can then work with them like any other object in your R environment. Thus, even though `cptac` is a Python package, with the help of `reticulate` you can use it to load and work with the CPTAC datasets in R. All `cptac` features are available through `reticulate` just like they would be in a Python environment, including the `cptac.utils` module and the join functions.

## Environment setup
First, you need to load the `reticulate` package, and then tell it which Python environment you want to use to access the `cptac` package.

```{r, message=FALSE}
# Install reticulate if necessary
if (!require(reticulate)) install.packages("reticulate")

# Load the package
library(reticulate)

# Specify to use the environment where the cptac package is installed. Replace 
# "dev" with the name of your environment. If you use an environment manager 
# besides conda, you'll want this command: use_virtualenv("myenv")
use_condaenv("dev", required = TRUE)
```

## Import `cptac` and load data
Now we will import `cptac` and load the dataset we want.

```{r}
# Import the package. We pass convert = FALSE so that objects won't be 
# converted from Python to R until we explicitly ask for it. This is necessary 
# to properly prepare a multiindex dataframe for conversion to R (see below).
cptac <- import("cptac", convert = FALSE) 

# Load the dataset
cc <- cptac$Ccrcc()
```
## Get a dataframe and convert it to an R object
This dataframe just has a regular column index (not a multiindex), so we can directly convert it into an R object after accessing it.

When `reticulate` converts a `pandas.DataFrame` into R, it converts it to the R `data.frame` type. For a complete list of the type conversion conventions `reticulate` uses between Python and R, see [this section](https://rstudio.github.io/reticulate/index.html#type-conversions) of the `reticulate` documentation.

```{r}
# Load the table
transcript_py <- cc$get_transcriptomics("washu") 

# Convert into R
transcript <- py_to_r(transcript_py) 

print(transcript[1:10, 1:6])
```

## Load a multiindex dataframe
If you're loading a dataframe that has a column multiindex, you need to flatten the multiindex before converting it to an R object. We will use the provided `cptac.utils` function for this. See [tutorial 4](https://paynelab.github.io/cptac/tutorial04_multiindex.html) for more information on column multiindexes.

```{r}
# Load the table
prot_py_multiindex <- cc$get_proteomics("umich")

# Load cptac.utils so we can access a helper function to convert the multiindex 
# to a single level index.
utils <- import("cptac.utils", convert = FALSE)

# Convert the multiindex to a single level index. This works because we passed 
# convert = FALSE when we imported the cptac package, so the dataframe is still
# a Python object at this point (it  wasn't automatically converted into an R 
# object on loading) and can be passed to this Python function.
prot_py_single_index <- utils$reduce_multiindex(prot_py_multiindex, flatten = TRUE) 

# Convert to R
prot <- py_to_r(prot_py_single_index) 

print(prot[1:10, 1:3])
```

## Use a join function
Using `reticulate`, you can also access all the table joining functions provided by `cptac`.

```{r}
# Call the join function
transcript_and_mut_py <- cc$join_omics_to_mutations(
  omics_name = "transcriptomics",
  omics_source = "washu",
  mutations_genes = "VHL",
  mutations_source = "harmonized",
  omics_genes = "PBRM1",
  quiet = TRUE
)

# Convert to R
transcript_and_mut <- py_to_r(transcript_and_mut_py)

print(transcript_and_mut[1:10,])
```
Note that if we wanted to pass an empty list to the `mutations_filter` parameter in order to use the default mutation filter, we can't just do something like `mutations_filter = r_to_py(c())`. The reason is that the `r_to_py` function converts and empty R vector into a `None` in Python, not an empty list. However, we can solve our problem by first creating a non-empty vector, converting it to a Python list, and then calling the list's `clear` method:  

```{r, results="hide"}
empty_filter <- r_to_py(c(0, 0)) # Vector must have more than one element
empty_filter$clear()
```

`empty_filter` is now an empty Python list, and can be passed to `join_omics_to_mutations` to get the default mutations filter. 

If you're using R Markdown, another solution is to just execute the command in a Python chunk in your R Markdown document, passing an empty list directly using Python's syntax, and then access the output from the next R code chunk. This is demonstrated in the next section.

## `reticulate`'s Python engine for R Markdown

If you're using R Markdown, `reticulate` provides a Python engine that allows you to run Python code chunks, and then access objects within them from R through an object called `py`. Similarly, R objects are accessible from Python chunks through an object named `r`.

Here is a Python code chunk that performs the same join function as the previous example, but with an empty list as the mutations filter:

```{python, warning=FALSE, results="hide"}
import cptac
cc = cptac.Ccrcc()
transcript_and_mut = cc.join_omics_to_mutations(
  omics_name="transcriptomics",
  omics_source = "washu",
  mutations_genes="VHL",
  mutations_source = "harmonized",
  omics_genes="PBRM1",
  mutations_filter=[],
  quiet=True
)
```

And here we access the output in an R chunk, where we could then perform further analysis.
```{r, message=FALSE}
transcript_and_mut_2 <- py$transcript_and_mut
print(transcript_and_mut_2[1:10,])
```