{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning: Your version of cptac (0.6.5) is out-of-date. Latest is 0.7.0. Please run 'pip install --upgrade cptac' to update it. (/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/ipykernel_launcher.py, line 1)\n"
     ]
    }
   ],
   "source": [
    "import cptac"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Available datasets:\n",
      "\n",
      "Dataset name                               Description     Data reuse status  \\\n",
      "                                                                               \n",
      "Brca                                     breast cancer  password access only   \n",
      "Colon                                colorectal cancer       no restrictions   \n",
      "Endometrial            endometrial carcinoma (uterine)       no restrictions   \n",
      "Gbm                                       glioblastoma  password access only   \n",
      "Hnscc                                    head and neck  password access only   \n",
      "Luad                               lung adenocarcinoma  password access only   \n",
      "Ovarian               high grade serous ovarian cancer       no restrictions   \n",
      "Ccrcc         clear cell renal cell carcinoma (kidney)       no restrictions   \n",
      "\n",
      "Dataset name                              Publication link  \n",
      "                                                            \n",
      "Brca                                           unpublished  \n",
      "Colon         https://www.ncbi.nlm.nih.gov/pubmed/31031003  \n",
      "Endometrial                                    unpublished  \n",
      "Gbm                                            unpublished  \n",
      "Hnscc                                          unpublished  \n",
      "Luad                                           unpublished  \n",
      "Ovarian                                        unpublished  \n",
      "Ccrcc                                          unpublished  \n"
     ]
    }
   ],
   "source": [
    "#list datasets\n",
    "cptac.list_datasets()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                    \r"
     ]
    },
    {
     "ename": "PackageCannotHandleDataVersionError",
     "evalue": "You tried to load data version 2.0, but your version of cptac can only handle these versions: ['0.1']. Update your package to be able to load the new data. Or, if you cannot currently update, manually specify the old data version using the 'version' parameter when you load the dataset.",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mPackageCannotHandleDataVersionError\u001b[0m       Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-8-3081936ec397>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      4\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m \u001b[0;31m#k = cptac.Ccrcc()\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 6\u001b[0;31m \u001b[0mg\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mcptac\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mHnscc\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/cptac/hnscc.py\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self, version)\u001b[0m\n\u001b[1;32m     38\u001b[0m         }\n\u001b[1;32m     39\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 40\u001b[0;31m         \u001b[0msuper\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__init__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcancer_type\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m\"hnscc\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mversion\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mversion\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mvalid_versions\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mvalid_versions\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdata_files\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mdata_files\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     41\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     42\u001b[0m         \u001b[0;31m# Load the data into dataframes in the self._data dict\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/cptac/dataset.py\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self, cancer_type, version, valid_versions, data_files)\u001b[0m\n\u001b[1;32m     43\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     44\u001b[0m         \u001b[0;31m# Validate the version\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 45\u001b[0;31m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_version\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mvalidate_version\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mversion\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_cancer_type\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0muse_context\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m\"init\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mvalid_versions\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mvalid_versions\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     46\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     47\u001b[0m         \u001b[0;31m# Get the paths to the data files\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/cptac/file_tools.py\u001b[0m in \u001b[0;36mvalidate_version\u001b[0;34m(version, dataset, use_context, valid_versions)\u001b[0m\n\u001b[1;32m     73\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0mvalid_versions\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     74\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mreturn_version\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mvalid_versions\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 75\u001b[0;31m             \u001b[0;32mraise\u001b[0m \u001b[0mPackageCannotHandleDataVersionError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34mf\"You tried to load data version {return_version}, but your version of cptac can only handle these versions: {valid_versions}. Update your package to be able to load the new data. Or, if you cannot currently update, manually specify the old data version using the 'version' parameter when you load the dataset.\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     76\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     77\u001b[0m     \u001b[0;32mreturn\u001b[0m \u001b[0mreturn_version\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mPackageCannotHandleDataVersionError\u001b[0m: You tried to load data version 2.0, but your version of cptac can only handle these versions: ['0.1']. Update your package to be able to load the new data. Or, if you cannot currently update, manually specify the old data version using the 'version' parameter when you load the dataset."
     ]
    }
   ],
   "source": [
    "#Download lastest dataset provided\n",
    "#cptac.download(dataset='ccrcc', version='0.0')\n",
    "cptac.download(dataset='hnscc', version='latest')\n",
    "\n",
    "#k = cptac.Ccrcc()\n",
    "g = cptac.Hnscc()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Below are the dataframes contained in this dataset:\n",
      "\tclinical\n",
      "\t\tDimensions: (131, 86)\n",
      "\tCNV\n",
      "\t\tDimensions: (100, 17156)\n",
      "\tphosphoproteomics\n",
      "\t\tDimensions: (103, 110953)\n",
      "\tproteomics\n",
      "\t\tDimensions: (103, 10686)\n",
      "\tsomatic_mutation\n",
      "\t\tDimensions: (22399, 3)\n",
      "\ttranscriptomics\n",
      "\t\tDimensions: (101, 21991)\n",
      "\ttreatment\n",
      "\t\tDimensions: (199, 9)\n"
     ]
    }
   ],
   "source": [
    "#k.list_data()  #see what data is available\n",
    "g.list_data()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Name       A1BG-AS1      A1CF       A2M  A2M-AS1     A2ML1     A2MP1  \\\n",
      "Sample_ID                                                              \n",
      "S001       0.364000  0.006166   47.7384      0.0  0.272554  0.027215   \n",
      "S002       0.728112  0.000000   84.2226      0.0  0.031258  0.000000   \n",
      "S003       0.161232  0.000000   39.4228      0.0  0.039369  0.000000   \n",
      "S004       2.362500  0.000000   89.5759      0.0  0.915590  0.000000   \n",
      "S005       0.529878  0.000000   65.2893      0.0  0.005661  0.138317   \n",
      "S006       1.061360  0.000000   43.2028      0.0  0.129552  0.086932   \n",
      "S007       1.630710  0.000000   77.4961      0.0  0.058105  0.000000   \n",
      "S008       0.653510  0.000000   21.7834      0.0  0.092119  0.000000   \n",
      "S009       0.945673  0.003329    0.0000      0.0  0.326060  0.201690   \n",
      "S011       7.890010  0.000000  152.0830      0.0  0.035705  0.092975   \n",
      "S013       2.985270  0.000000   17.2690      0.0  0.111993  0.000000   \n",
      "S015       2.236830  0.000000   43.6967      0.0  0.096975  0.000000   \n",
      "S016       1.050050  0.000000  157.4640      0.0  0.008024  0.223129   \n",
      "S017       0.505459  0.000000   30.3616      0.0  0.064189  0.000000   \n",
      "S018       0.852912  0.000000   64.5888      0.0  1.925800  0.000000   \n",
      "S019       1.673820  0.000000   20.2543      0.0  0.698810  0.057793   \n",
      "S020       0.000000  0.000000   18.6749      0.0  0.009476  0.000000   \n",
      "S022       0.000000  0.000000   61.8057      0.0  0.211141  0.000000   \n",
      "S023       1.856580  0.000000    0.0000      0.0  0.188390  0.029225   \n",
      "S024       0.509986  0.000000  104.7680      0.0  1.837770  0.028599   \n",
      "S025       0.650860  0.000000   50.7993      0.0  0.063839  0.000000   \n",
      "\n",
      "Name        A3GALT2    A4GALT  A4GNT  AA06  ...    ZWILCH     ZWINT      ZXDA  \\\n",
      "Sample_ID                                   ...                                 \n",
      "S001       0.065777  5.005570    0.0   0.0  ...  12.72020  60.83850  1.734470   \n",
      "S002       0.000000  4.883460    0.0   0.0  ...   9.09380  18.89510  0.572400   \n",
      "S003       0.000000  4.002660    0.0   0.0  ...   0.00000  18.87590  1.425120   \n",
      "S004       0.000000  2.439680    0.0   0.0  ...  10.03360  50.46290  0.835355   \n",
      "S005       0.000000  2.669980    0.0   0.0  ...  16.05770  27.10560  0.638152   \n",
      "S006       0.000000  2.294190    0.0   0.0  ...  19.43110  79.59150  1.337140   \n",
      "S007       0.000000  3.427330    0.0   0.0  ...  15.42620  39.56730  1.194730   \n",
      "S008       0.000000  1.530590    0.0   0.0  ...   9.52869  77.10660  0.840225   \n",
      "S009       0.000000  1.438310    0.0   0.0  ...  12.80430  15.64000  1.201960   \n",
      "S011       0.000000  8.277170    0.0   0.0  ...   6.80181  28.47270  1.948970   \n",
      "S013       0.000000  6.217420    0.0   0.0  ...   4.71058   4.24464  1.852650   \n",
      "S015       0.000000  0.811302    0.0   0.0  ...   7.30360  44.47160  1.091900   \n",
      "S016       0.000000  1.981480    0.0   0.0  ...   7.63648  18.34270  1.755640   \n",
      "S017       0.000000  2.895880    0.0   0.0  ...   7.73802  34.78190  0.608341   \n",
      "S018       0.000000  1.390220    0.0   0.0  ...  12.12650  14.09660  3.496890   \n",
      "S019       0.000000  2.488420    0.0   0.0  ...   9.92233  26.00680  0.975760   \n",
      "S020       0.000000  2.300570    0.0   0.0  ...   9.87972  59.63580  0.579566   \n",
      "S022       0.000000  0.744234    0.0   0.0  ...   7.10841  39.29740  1.538910   \n",
      "S023       0.000000  2.741250    0.0   0.0  ...  20.62220  34.76280  1.101830   \n",
      "S024       0.000000  5.862730    0.0   0.0  ...   4.91671  16.65520  1.268050   \n",
      "S025       0.069345  7.191760    0.0   0.0  ...   7.69742  31.82520  1.517310   \n",
      "\n",
      "Name           ZXDB      ZXDC    ZYG11A   ZYG11B       ZYX     ZZEF1      ZZZ3  \n",
      "Sample_ID                                                                       \n",
      "S001       2.816760   5.89275  1.384150  8.26588  11.75730  2.264400  21.88050  \n",
      "S002       0.701954   4.01544  0.111984  4.11415  19.84100  1.384810  18.56750  \n",
      "S003       1.822010   2.97880  0.153857  2.91195   0.00000  2.359430  17.20760  \n",
      "S004       0.642220   1.96876  0.014043  6.17068   1.77563  0.821267  34.48840  \n",
      "S005       0.732492   3.43952  0.034667  3.73611   9.84522  1.920880  15.30810  \n",
      "S006       1.592940   9.81789  0.072338  7.03253  12.94460  1.767600  29.91750  \n",
      "S007       1.643500   3.80123  1.458800  4.16472  15.39000  2.572850  14.05160  \n",
      "S008       0.990311   4.54592  0.675662  3.77929   8.91699  1.801410  25.49100  \n",
      "S009       1.587360   6.21851  1.721800  5.89249  22.69950  2.818250  17.24740  \n",
      "S011       3.297630  11.13780  0.896745  5.23134  44.44080  3.715270  16.70650  \n",
      "S013       2.748940   8.18013  0.490881  5.50541  30.04200  4.517830  19.23570  \n",
      "S015       2.114840  10.59750  0.194627  8.13019  20.91450  5.995480  12.01880  \n",
      "S016       3.034330  17.26860  0.063760  5.02684  56.88150  4.752450  13.06160  \n",
      "S017       0.830715   6.88740  0.275040  3.63430   8.77731  0.902103  16.04340  \n",
      "S018       5.003600  14.55360  0.192423  9.43435  54.46640  2.554360   9.99008  \n",
      "S019       1.178800   5.11260  1.001290  3.89017  10.01740  0.871825  16.69370  \n",
      "S020       0.702658   8.08673  3.175120  4.36180  25.03410  0.779903  25.49370  \n",
      "S022       2.602820   7.46087  0.082061  6.85874  33.39950  1.832400  14.87700  \n",
      "S023       1.656370   7.24506  0.200741  4.32149   6.32237  1.043140  17.69250  \n",
      "S024       2.075380  11.85920  0.293322  5.47253   0.00000  4.864790  11.58680  \n",
      "S025       2.967230  12.20400  0.000000  5.37060  64.02920  4.627130  13.13530  \n",
      "\n",
      "[21 rows x 21991 columns]\n"
     ]
    }
   ],
   "source": [
    "\"\"\" Transcriptomics\n",
    "This strategy ratios each sample's abundance to a pooled reference in TMT. The ratio is \n",
    "then log transformed. Therefore positive values indicate a measurement \n",
    "higher than the pooled reference; negative values are lower than the pooled reference.\n",
    "\"\"\"\n",
    "dataf = g.get_transcriptomics()\n",
    "print(dataf[0:21])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Name</th>\n",
       "      <th>A1BG</th>\n",
       "      <th>PTEN</th>\n",
       "      <th>TP53</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sample_ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>S001</th>\n",
       "      <td>NaN</td>\n",
       "      <td>5.02901</td>\n",
       "      <td>3.70537</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S003</th>\n",
       "      <td>NaN</td>\n",
       "      <td>13.93630</td>\n",
       "      <td>5.75074</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S016</th>\n",
       "      <td>NaN</td>\n",
       "      <td>21.10090</td>\n",
       "      <td>7.93562</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Name       A1BG      PTEN     TP53\n",
       "Sample_ID                         \n",
       "S001        NaN   5.02901  3.70537\n",
       "S003        NaN  13.93630  5.75074\n",
       "S016        NaN  21.10090  7.93562"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#How to choose certain rows and columns\n",
    "samples = [\"S001\",\"S003\",\"S016\"]\n",
    "proteins = [\"A1BG\",\"PTEN\",\"TP53\"]\n",
    "dataf.loc[samples, proteins]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Name</th>\n",
       "      <th>A1BG-AS1</th>\n",
       "      <th>A1CF</th>\n",
       "      <th>A2M</th>\n",
       "      <th>A2M-AS1</th>\n",
       "      <th>A2ML1</th>\n",
       "      <th>A2MP1</th>\n",
       "      <th>A3GALT2</th>\n",
       "      <th>A4GALT</th>\n",
       "      <th>A4GNT</th>\n",
       "      <th>AA06</th>\n",
       "      <th>...</th>\n",
       "      <th>ZWILCH</th>\n",
       "      <th>ZWINT</th>\n",
       "      <th>ZXDA</th>\n",
       "      <th>ZXDB</th>\n",
       "      <th>ZXDC</th>\n",
       "      <th>ZYG11A</th>\n",
       "      <th>ZYG11B</th>\n",
       "      <th>ZYX</th>\n",
       "      <th>ZZEF1</th>\n",
       "      <th>ZZZ3</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sample_ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>S001</th>\n",
       "      <td>0.364000</td>\n",
       "      <td>0.006166</td>\n",
       "      <td>47.7384</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.272554</td>\n",
       "      <td>0.027215</td>\n",
       "      <td>0.065777</td>\n",
       "      <td>5.00557</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>12.7202</td>\n",
       "      <td>60.8385</td>\n",
       "      <td>1.734470</td>\n",
       "      <td>2.816760</td>\n",
       "      <td>5.89275</td>\n",
       "      <td>1.384150</td>\n",
       "      <td>8.26588</td>\n",
       "      <td>11.75730</td>\n",
       "      <td>2.264400</td>\n",
       "      <td>21.8805</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S002</th>\n",
       "      <td>0.728112</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>84.2226</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.031258</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>4.88346</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>9.0938</td>\n",
       "      <td>18.8951</td>\n",
       "      <td>0.572400</td>\n",
       "      <td>0.701954</td>\n",
       "      <td>4.01544</td>\n",
       "      <td>0.111984</td>\n",
       "      <td>4.11415</td>\n",
       "      <td>19.84100</td>\n",
       "      <td>1.384810</td>\n",
       "      <td>18.5675</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S003</th>\n",
       "      <td>0.161232</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>39.4228</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.039369</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>4.00266</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>18.8759</td>\n",
       "      <td>1.425120</td>\n",
       "      <td>1.822010</td>\n",
       "      <td>2.97880</td>\n",
       "      <td>0.153857</td>\n",
       "      <td>2.91195</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>2.359430</td>\n",
       "      <td>17.2076</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S004</th>\n",
       "      <td>2.362500</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>89.5759</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.915590</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.43968</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>10.0336</td>\n",
       "      <td>50.4629</td>\n",
       "      <td>0.835355</td>\n",
       "      <td>0.642220</td>\n",
       "      <td>1.96876</td>\n",
       "      <td>0.014043</td>\n",
       "      <td>6.17068</td>\n",
       "      <td>1.77563</td>\n",
       "      <td>0.821267</td>\n",
       "      <td>34.4884</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S005</th>\n",
       "      <td>0.529878</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>65.2893</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.005661</td>\n",
       "      <td>0.138317</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.66998</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>16.0577</td>\n",
       "      <td>27.1056</td>\n",
       "      <td>0.638152</td>\n",
       "      <td>0.732492</td>\n",
       "      <td>3.43952</td>\n",
       "      <td>0.034667</td>\n",
       "      <td>3.73611</td>\n",
       "      <td>9.84522</td>\n",
       "      <td>1.920880</td>\n",
       "      <td>15.3081</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 21991 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Name       A1BG-AS1      A1CF      A2M  A2M-AS1     A2ML1     A2MP1   A3GALT2  \\\n",
       "Sample_ID                                                                       \n",
       "S001       0.364000  0.006166  47.7384      0.0  0.272554  0.027215  0.065777   \n",
       "S002       0.728112  0.000000  84.2226      0.0  0.031258  0.000000  0.000000   \n",
       "S003       0.161232  0.000000  39.4228      0.0  0.039369  0.000000  0.000000   \n",
       "S004       2.362500  0.000000  89.5759      0.0  0.915590  0.000000  0.000000   \n",
       "S005       0.529878  0.000000  65.2893      0.0  0.005661  0.138317  0.000000   \n",
       "\n",
       "Name        A4GALT  A4GNT  AA06  ...   ZWILCH    ZWINT      ZXDA      ZXDB  \\\n",
       "Sample_ID                        ...                                         \n",
       "S001       5.00557    0.0   0.0  ...  12.7202  60.8385  1.734470  2.816760   \n",
       "S002       4.88346    0.0   0.0  ...   9.0938  18.8951  0.572400  0.701954   \n",
       "S003       4.00266    0.0   0.0  ...   0.0000  18.8759  1.425120  1.822010   \n",
       "S004       2.43968    0.0   0.0  ...  10.0336  50.4629  0.835355  0.642220   \n",
       "S005       2.66998    0.0   0.0  ...  16.0577  27.1056  0.638152  0.732492   \n",
       "\n",
       "Name          ZXDC    ZYG11A   ZYG11B       ZYX     ZZEF1     ZZZ3  \n",
       "Sample_ID                                                           \n",
       "S001       5.89275  1.384150  8.26588  11.75730  2.264400  21.8805  \n",
       "S002       4.01544  0.111984  4.11415  19.84100  1.384810  18.5675  \n",
       "S003       2.97880  0.153857  2.91195   0.00000  2.359430  17.2076  \n",
       "S004       1.96876  0.014043  6.17068   1.77563  0.821267  34.4884  \n",
       "S005       3.43952  0.034667  3.73611   9.84522  1.920880  15.3081  \n",
       "\n",
       "[5 rows x 21991 columns]"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Search for expression greater than 0 for a specific gene\n",
    "brca_positive = dataf.loc[dataf[\"BRCA1\"] > 0]\n",
    "brca_positive.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning: transcriptomics data was not found for the following samples, so transcriptomics data columns were filled with NaN for these samples: S012, S112, S113, S114, S115, S116, S117, S118, S119, S120, S121, S122, S123, S124, S125, S126, S127, S128, S129, S130, S131 (/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/ipykernel_launcher.py, line 2)\n",
      "Warning: proteomics data was not found for the following samples, so proteomics data columns were filled with NaN for these samples: S001, S003, S004, S005, S008, S017, S018, S020, S030, S032, S034, S036, S058, S072, S075, S092, S099, S105, S109 (/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/ipykernel_launcher.py, line 2)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr:last-of-type th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th>Name</th>\n",
       "      <th>A1BG_proteomics</th>\n",
       "      <th>A2M_proteomics</th>\n",
       "      <th>A2ML1_proteomics</th>\n",
       "      <th>AAAS_proteomics</th>\n",
       "      <th>AACS_proteomics</th>\n",
       "      <th>AAGAB_proteomics</th>\n",
       "      <th>AAK1_proteomics</th>\n",
       "      <th>AAMDC_proteomics</th>\n",
       "      <th>AAMP_proteomics</th>\n",
       "      <th>AAR2_proteomics</th>\n",
       "      <th>...</th>\n",
       "      <th>ZWILCH_transcriptomics</th>\n",
       "      <th>ZWINT_transcriptomics</th>\n",
       "      <th>ZXDA_transcriptomics</th>\n",
       "      <th>ZXDB_transcriptomics</th>\n",
       "      <th>ZXDC_transcriptomics</th>\n",
       "      <th>ZYG11A_transcriptomics</th>\n",
       "      <th>ZYG11B_transcriptomics</th>\n",
       "      <th>ZYX_transcriptomics</th>\n",
       "      <th>ZZEF1_transcriptomics</th>\n",
       "      <th>ZZZ3_transcriptomics</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Database_ID</th>\n",
       "      <th>NP_570602</th>\n",
       "      <th>NP_000005</th>\n",
       "      <th>NP_653271</th>\n",
       "      <th>NP_056480</th>\n",
       "      <th>NP_076417</th>\n",
       "      <th>NP_078942</th>\n",
       "      <th>NP_055726</th>\n",
       "      <th>NP_078960</th>\n",
       "      <th>NP_001078</th>\n",
       "      <th>NP_001258803</th>\n",
       "      <th>...</th>\n",
       "      <th>NaN</th>\n",
       "      <th>NaN</th>\n",
       "      <th>NaN</th>\n",
       "      <th>NaN</th>\n",
       "      <th>NaN</th>\n",
       "      <th>NaN</th>\n",
       "      <th>NaN</th>\n",
       "      <th>NaN</th>\n",
       "      <th>NaN</th>\n",
       "      <th>NaN</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sample_ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>S001</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>12.7202</td>\n",
       "      <td>60.8385</td>\n",
       "      <td>1.734470</td>\n",
       "      <td>2.816760</td>\n",
       "      <td>5.89275</td>\n",
       "      <td>1.384150</td>\n",
       "      <td>8.26588</td>\n",
       "      <td>11.75730</td>\n",
       "      <td>2.264400</td>\n",
       "      <td>21.8805</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S002</th>\n",
       "      <td>0.133634</td>\n",
       "      <td>0.76512</td>\n",
       "      <td>-0.778849</td>\n",
       "      <td>0.118236</td>\n",
       "      <td>0.11274</td>\n",
       "      <td>0.054942</td>\n",
       "      <td>0.313055</td>\n",
       "      <td>-0.78627</td>\n",
       "      <td>-0.38468</td>\n",
       "      <td>0.064017</td>\n",
       "      <td>...</td>\n",
       "      <td>9.0938</td>\n",
       "      <td>18.8951</td>\n",
       "      <td>0.572400</td>\n",
       "      <td>0.701954</td>\n",
       "      <td>4.01544</td>\n",
       "      <td>0.111984</td>\n",
       "      <td>4.11415</td>\n",
       "      <td>19.84100</td>\n",
       "      <td>1.384810</td>\n",
       "      <td>18.5675</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S003</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>18.8759</td>\n",
       "      <td>1.425120</td>\n",
       "      <td>1.822010</td>\n",
       "      <td>2.97880</td>\n",
       "      <td>0.153857</td>\n",
       "      <td>2.91195</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>2.359430</td>\n",
       "      <td>17.2076</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S004</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>10.0336</td>\n",
       "      <td>50.4629</td>\n",
       "      <td>0.835355</td>\n",
       "      <td>0.642220</td>\n",
       "      <td>1.96876</td>\n",
       "      <td>0.014043</td>\n",
       "      <td>6.17068</td>\n",
       "      <td>1.77563</td>\n",
       "      <td>0.821267</td>\n",
       "      <td>34.4884</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S005</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>16.0577</td>\n",
       "      <td>27.1056</td>\n",
       "      <td>0.638152</td>\n",
       "      <td>0.732492</td>\n",
       "      <td>3.43952</td>\n",
       "      <td>0.034667</td>\n",
       "      <td>3.73611</td>\n",
       "      <td>9.84522</td>\n",
       "      <td>1.920880</td>\n",
       "      <td>15.3081</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 32677 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Name        A1BG_proteomics A2M_proteomics A2ML1_proteomics AAAS_proteomics  \\\n",
       "Database_ID       NP_570602      NP_000005        NP_653271       NP_056480   \n",
       "Sample_ID                                                                     \n",
       "S001                    NaN            NaN              NaN             NaN   \n",
       "S002               0.133634        0.76512        -0.778849        0.118236   \n",
       "S003                    NaN            NaN              NaN             NaN   \n",
       "S004                    NaN            NaN              NaN             NaN   \n",
       "S005                    NaN            NaN              NaN             NaN   \n",
       "\n",
       "Name        AACS_proteomics AAGAB_proteomics AAK1_proteomics AAMDC_proteomics  \\\n",
       "Database_ID       NP_076417        NP_078942       NP_055726        NP_078960   \n",
       "Sample_ID                                                                       \n",
       "S001                    NaN              NaN             NaN              NaN   \n",
       "S002                0.11274         0.054942        0.313055         -0.78627   \n",
       "S003                    NaN              NaN             NaN              NaN   \n",
       "S004                    NaN              NaN             NaN              NaN   \n",
       "S005                    NaN              NaN             NaN              NaN   \n",
       "\n",
       "Name        AAMP_proteomics AAR2_proteomics  ... ZWILCH_transcriptomics  \\\n",
       "Database_ID       NP_001078    NP_001258803  ...                    NaN   \n",
       "Sample_ID                                    ...                          \n",
       "S001                    NaN             NaN  ...                12.7202   \n",
       "S002               -0.38468        0.064017  ...                 9.0938   \n",
       "S003                    NaN             NaN  ...                 0.0000   \n",
       "S004                    NaN             NaN  ...                10.0336   \n",
       "S005                    NaN             NaN  ...                16.0577   \n",
       "\n",
       "Name        ZWINT_transcriptomics ZXDA_transcriptomics ZXDB_transcriptomics  \\\n",
       "Database_ID                   NaN                  NaN                  NaN   \n",
       "Sample_ID                                                                     \n",
       "S001                      60.8385             1.734470             2.816760   \n",
       "S002                      18.8951             0.572400             0.701954   \n",
       "S003                      18.8759             1.425120             1.822010   \n",
       "S004                      50.4629             0.835355             0.642220   \n",
       "S005                      27.1056             0.638152             0.732492   \n",
       "\n",
       "Name        ZXDC_transcriptomics ZYG11A_transcriptomics  \\\n",
       "Database_ID                  NaN                    NaN   \n",
       "Sample_ID                                                 \n",
       "S001                     5.89275               1.384150   \n",
       "S002                     4.01544               0.111984   \n",
       "S003                     2.97880               0.153857   \n",
       "S004                     1.96876               0.014043   \n",
       "S005                     3.43952               0.034667   \n",
       "\n",
       "Name        ZYG11B_transcriptomics ZYX_transcriptomics ZZEF1_transcriptomics  \\\n",
       "Database_ID                    NaN                 NaN                   NaN   \n",
       "Sample_ID                                                                      \n",
       "S001                       8.26588            11.75730              2.264400   \n",
       "S002                       4.11415            19.84100              1.384810   \n",
       "S003                       2.91195             0.00000              2.359430   \n",
       "S004                       6.17068             1.77563              0.821267   \n",
       "S005                       3.73611             9.84522              1.920880   \n",
       "\n",
       "Name        ZZZ3_transcriptomics  \n",
       "Database_ID                  NaN  \n",
       "Sample_ID                         \n",
       "S001                     21.8805  \n",
       "S002                     18.5675  \n",
       "S003                     17.2076  \n",
       "S004                     34.4884  \n",
       "S005                     15.3081  \n",
       "\n",
       "[5 rows x 32677 columns]"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Join the omics\n",
    "prot_and_trans = g.join_omics_to_omics(df1_name=\"proteomics\", df2_name=\"transcriptomics\")\n",
    "prot_and_trans.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Name           Gene           Mutation  Location\n",
      "Sample_ID                                       \n",
      "S002          ALPK1  Missense_Mutation    p.C39F\n",
      "S002          AP4B1  Missense_Mutation   p.C420Y\n",
      "S002        ARHGAP6  Missense_Mutation   p.E274V\n",
      "S002          ARMC6  Missense_Mutation   p.L319V\n",
      "S002           BAI2  Missense_Mutation   p.A229P\n",
      "...             ...                ...       ...\n",
      "S111       TXNRD3NB  Missense_Mutation     p.E6K\n",
      "S111           VNN2  Missense_Mutation   p.A281S\n",
      "S111            VWF             Silent  p.A2569A\n",
      "S111           ZFR2  Missense_Mutation   p.R583W\n",
      "S111         ZNF831  Missense_Mutation  p.R1310C\n",
      "\n",
      "[22399 rows x 3 columns]\n"
     ]
    }
   ],
   "source": [
    "somatic_mutations = g.get_somatic_mutation()\n",
    "print(somatic_mutations)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Name                AAAS                                 AAGAB            \\\n",
      "Site                S495                                  S311             \n",
      "Peptide     R.FS*PVLGR.A           K.AFWMAIGGDRDEIEGLSS*DEEH.-             \n",
      "Database_ID NP_001166937 NP_056480                NP_001258815 NP_078942   \n",
      "Sample_ID                                                                  \n",
      "S002                 NaN       NaN                         NaN       NaN   \n",
      "S006                 NaN       NaN                         NaN       NaN   \n",
      "S007                 NaN       NaN                         NaN       NaN   \n",
      "S009              -1.137    -1.137                         NaN       NaN   \n",
      "S011              -0.987    -0.987                         NaN       NaN   \n",
      "...                  ...       ...                         ...       ...   \n",
      "S127              -2.667    -2.667                      -2.351    -2.351   \n",
      "S128              -2.706    -2.706                         NaN       NaN   \n",
      "S129              -2.959    -2.959                         NaN       NaN   \n",
      "S130                 NaN       NaN                         NaN       NaN   \n",
      "S131              -2.906    -2.906                         NaN       NaN   \n",
      "\n",
      "Name                                      AAK1  \\\n",
      "Site                                       S18   \n",
      "Peptide     R.REQGGSGLGS*GSSGGGGSTSGLGSGYIGR.V   \n",
      "Database_ID                          NP_055726   \n",
      "Sample_ID                                        \n",
      "S002                                       NaN   \n",
      "S006                                       NaN   \n",
      "S007                                       NaN   \n",
      "S009                                       NaN   \n",
      "S011                                       NaN   \n",
      "...                                        ...   \n",
      "S127                                       NaN   \n",
      "S128                                       NaN   \n",
      "S129                                       NaN   \n",
      "S130                                       NaN   \n",
      "S131                                       NaN   \n",
      "\n",
      "Name                                           \\\n",
      "Site                                      S20   \n",
      "Peptide     R.EQGGSGLGSGS*SGGGGSTSGLGSGYIGR.V   \n",
      "Database_ID                         NP_055726   \n",
      "Sample_ID                                       \n",
      "S002                                      NaN   \n",
      "S006                                      NaN   \n",
      "S007                                      NaN   \n",
      "S009                                      NaN   \n",
      "S011                                      NaN   \n",
      "...                                       ...   \n",
      "S127                                      NaN   \n",
      "S128                                      NaN   \n",
      "S129                                      NaN   \n",
      "S130                                      NaN   \n",
      "S131                                      NaN   \n",
      "\n",
      "Name                                            \\\n",
      "Site                                             \n",
      "Peptide     R.REQGGSGLGSGS*SGGGGSTSGLGSGYIGR.V   \n",
      "Database_ID                          NP_055726   \n",
      "Sample_ID                                        \n",
      "S002                                       NaN   \n",
      "S006                                       NaN   \n",
      "S007                                       NaN   \n",
      "S009                                       NaN   \n",
      "S011                                       NaN   \n",
      "...                                        ...   \n",
      "S127                                       NaN   \n",
      "S128                                       NaN   \n",
      "S129                                       NaN   \n",
      "S130                                       NaN   \n",
      "S131                                       NaN   \n",
      "\n",
      "Name                                                            \\\n",
      "Site                                      S21             S624   \n",
      "Peptide     R.EQGGSGLGSGSS*GGGGSTSGLGSGYIGR.V K.VGSLTPPSS*PK.T   \n",
      "Database_ID                         NP_055726        NP_055726   \n",
      "Sample_ID                                                        \n",
      "S002                                      NaN              NaN   \n",
      "S006                                      NaN              NaN   \n",
      "S007                                      NaN              NaN   \n",
      "S009                                      NaN              NaN   \n",
      "S011                                   -2.269           -3.363   \n",
      "...                                       ...              ...   \n",
      "S127                                   -2.694              NaN   \n",
      "S128                                      NaN              NaN   \n",
      "S129                                      NaN           -2.644   \n",
      "S130                                      NaN              NaN   \n",
      "S131                                      NaN           -3.114   \n",
      "\n",
      "Name                         ...            ZZZ3                   \\\n",
      "Site                   S637  ...            S113                    \n",
      "Peptide     R.ILS*DVTHSAV.F  ... R.QTEPVS*PVLK.R R.QTEPVS*PVLKR.I   \n",
      "Database_ID       NP_055726  ...       NP_056349        NP_056349   \n",
      "Sample_ID                    ...                                    \n",
      "S002                    NaN  ...             NaN              NaN   \n",
      "S006                    NaN  ...             NaN              NaN   \n",
      "S007                    NaN  ...             NaN              NaN   \n",
      "S009                    NaN  ...             NaN           -0.336   \n",
      "S011                    NaN  ...             NaN           -0.770   \n",
      "...                     ...  ...             ...              ...   \n",
      "S127                    NaN  ...             NaN              NaN   \n",
      "S128                    NaN  ...             NaN           -1.477   \n",
      "S129                    NaN  ...             NaN              NaN   \n",
      "S130                    NaN  ...             NaN           -2.837   \n",
      "S131                    NaN  ...             NaN              NaN   \n",
      "\n",
      "Name                                                                   \\\n",
      "Site                                                             S135   \n",
      "Peptide     R.RQTEPVS*PVLK.R R.RQTEPVS*PVLKR.I R.SEAPNSSEEDS*PIKSDK.E   \n",
      "Database_ID        NP_056349         NP_056349              NP_056349   \n",
      "Sample_ID                                                               \n",
      "S002                     NaN               NaN                    NaN   \n",
      "S006                     NaN               NaN                    NaN   \n",
      "S007                     NaN               NaN                    NaN   \n",
      "S009                     NaN               NaN                    NaN   \n",
      "S011                     NaN               NaN                    NaN   \n",
      "...                      ...               ...                    ...   \n",
      "S127                     NaN               NaN                    NaN   \n",
      "S128                     NaN               NaN                    NaN   \n",
      "S129                  -3.345               NaN                    NaN   \n",
      "S130                     NaN               NaN                    NaN   \n",
      "S131                  -2.754               NaN                    NaN   \n",
      "\n",
      "Name                                                                         \\\n",
      "Site                                 S314            S391               S89   \n",
      "Peptide     F.SETQSSLRDS*EEEVDVVGDSSASK.E R.AAPTRGS*PTK.N R.GLS*SSEKDNIER.Q   \n",
      "Database_ID                     NP_056349       NP_056349         NP_056349   \n",
      "Sample_ID                                                                     \n",
      "S002                               -2.533             NaN               NaN   \n",
      "S006                               -3.276             NaN               NaN   \n",
      "S007                                  NaN          -0.847               NaN   \n",
      "S009                                  NaN          -0.455               NaN   \n",
      "S011                                  NaN             NaN               NaN   \n",
      "...                                   ...             ...               ...   \n",
      "S127                                  NaN             NaN               NaN   \n",
      "S128                                  NaN          -2.759               NaN   \n",
      "S129                                  NaN          -1.466            -1.821   \n",
      "S130                                  NaN             NaN               NaN   \n",
      "S131                                  NaN          -2.477            -0.109   \n",
      "\n",
      "Name                                             \n",
      "Site                      S90               S91  \n",
      "Peptide     R.GLSS*SEKDNIER.Q R.GLSSS*EKDNIER.Q  \n",
      "Database_ID         NP_056349         NP_056349  \n",
      "Sample_ID                                        \n",
      "S002                      NaN            -1.320  \n",
      "S006                      NaN            -0.264  \n",
      "S007                      NaN               NaN  \n",
      "S009                      NaN            -1.129  \n",
      "S011                      NaN            -1.301  \n",
      "...                       ...               ...  \n",
      "S127                      NaN               NaN  \n",
      "S128                      NaN            -0.806  \n",
      "S129                      NaN               NaN  \n",
      "S130                      NaN            -1.720  \n",
      "S131                      NaN               NaN  \n",
      "\n",
      "[103 rows x 110953 columns]\n"
     ]
    }
   ],
   "source": [
    "p = g.get_phosphoproteomics()\n",
    "print(p)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Name      Patient_ID Sample_Tumor_Normal  Participant_Procurement_Age  \\\n",
      "Sample_ID                                                               \n",
      "S001         01OV002               Tumor                        540.0   \n",
      "S002         01OV007               Tumor                        820.0   \n",
      "S003         01OV008               Tumor                        665.0   \n",
      "S004         01OV010               Tumor                        625.0   \n",
      "S005         01OV013               Tumor                        828.0   \n",
      "...              ...                 ...                          ...   \n",
      "S127        N17OV001              Normal                          NaN   \n",
      "S128        N17OV002              Normal                          NaN   \n",
      "S129        N17OV003              Normal                          NaN   \n",
      "S130        N17OV004              Normal                          NaN   \n",
      "S131        N17OV005              Normal                          NaN   \n",
      "\n",
      "Name      Participant_Gender           Participant_Race  \\\n",
      "Sample_ID                                                 \n",
      "S001                  Female                      White   \n",
      "S002                  Female                      White   \n",
      "S003                  Female                      White   \n",
      "S004                  Female  Black or African American   \n",
      "S005                  Female                      White   \n",
      "...                      ...                        ...   \n",
      "S127                     NaN                        NaN   \n",
      "S128                     NaN                        NaN   \n",
      "S129                     NaN                        NaN   \n",
      "S130                     NaN                        NaN   \n",
      "S131                     NaN                        NaN   \n",
      "\n",
      "Name        Participant_Ethnicity Participant_Jewish_Heritage  \\\n",
      "Sample_ID                                                       \n",
      "S001       Not Hispanic or Latino                  Not Jewish   \n",
      "S002       Not Hispanic or Latino                     Unknown   \n",
      "S003       Not Hispanic or Latino                  Not Jewish   \n",
      "S004                      Unknown                     Unknown   \n",
      "S005       Not Hispanic or Latino                  Not Jewish   \n",
      "...                           ...                         ...   \n",
      "S127                          NaN                         NaN   \n",
      "S128                          NaN                         NaN   \n",
      "S129                          NaN                         NaN   \n",
      "S130                          NaN                         NaN   \n",
      "S131                          NaN                         NaN   \n",
      "\n",
      "Name      Participant_History_Malignancy Participant_History_Chemotherapy  \\\n",
      "Sample_ID                                                                   \n",
      "S001                                  No                               No   \n",
      "S002                                  No                               No   \n",
      "S003                                  No                               No   \n",
      "S004                                  No                               No   \n",
      "S005                                  No                               No   \n",
      "...                                  ...                              ...   \n",
      "S127                                 NaN                              NaN   \n",
      "S128                                 NaN                              NaN   \n",
      "S129                                 NaN                              NaN   \n",
      "S130                                 NaN                              NaN   \n",
      "S131                                 NaN                              NaN   \n",
      "\n",
      "Name      Participant_History_Neo-adjuvant_Treatment  ...  \\\n",
      "Sample_ID                                             ...   \n",
      "S001                                              No  ...   \n",
      "S002                                              No  ...   \n",
      "S003                                              No  ...   \n",
      "S004                                              No  ...   \n",
      "S005                                              No  ...   \n",
      "...                                              ...  ...   \n",
      "S127                                             NaN  ...   \n",
      "S128                                             NaN  ...   \n",
      "S129                                             NaN  ...   \n",
      "S130                                             NaN  ...   \n",
      "S131                                             NaN  ...   \n",
      "\n",
      "Name      New_Tumor_Event_Site Other_New_Tumor_Event_Site  \\\n",
      "Sample_ID                                                   \n",
      "S001                       NaN                        NaN   \n",
      "S002                       NaN                        NaN   \n",
      "S003                       NaN                        NaN   \n",
      "S004                       NaN                        NaN   \n",
      "S005                       NaN                        NaN   \n",
      "...                        ...                        ...   \n",
      "S127                       NaN                        NaN   \n",
      "S128                       NaN                        NaN   \n",
      "S129                       NaN                        NaN   \n",
      "S130                       NaN                        NaN   \n",
      "S131                       NaN                        NaN   \n",
      "\n",
      "Name       Days_Between_Collection_And_New_Tumor_Event  \\\n",
      "Sample_ID                                                \n",
      "S001                                               NaN   \n",
      "S002                                               NaN   \n",
      "S003                                               NaN   \n",
      "S004                                               NaN   \n",
      "S005                                               NaN   \n",
      "...                                                ...   \n",
      "S127                                               NaN   \n",
      "S128                                               NaN   \n",
      "S129                                               NaN   \n",
      "S130                                               NaN   \n",
      "S131                                               NaN   \n",
      "\n",
      "Name      New_Tumor_Event_Diagnosis New_Tumor_Event_Surgery  \\\n",
      "Sample_ID                                                     \n",
      "S001                            NaN                     NaN   \n",
      "S002                            NaN                     NaN   \n",
      "S003                            NaN                     NaN   \n",
      "S004                            NaN                     NaN   \n",
      "S005                            NaN                     NaN   \n",
      "...                             ...                     ...   \n",
      "S127                            NaN                     NaN   \n",
      "S128                            NaN                     NaN   \n",
      "S129                            NaN                     NaN   \n",
      "S130                            NaN                     NaN   \n",
      "S131                            NaN                     NaN   \n",
      "\n",
      "Name       Days_Between_Collection_And_New_Tumor_Event_Surgery  \\\n",
      "Sample_ID                                                        \n",
      "S001                                                     NaN     \n",
      "S002                                                     NaN     \n",
      "S003                                                     NaN     \n",
      "S004                                                     NaN     \n",
      "S005                                                     NaN     \n",
      "...                                                      ...     \n",
      "S127                                                     NaN     \n",
      "S128                                                     NaN     \n",
      "S129                                                     NaN     \n",
      "S130                                                     NaN     \n",
      "S131                                                     NaN     \n",
      "\n",
      "Name       New_Tumor_Event_Chemotherapy  New_Tumor_Event_Immunotherapy  \\\n",
      "Sample_ID                                                                \n",
      "S001                                NaN                            NaN   \n",
      "S002                                NaN                            NaN   \n",
      "S003                                NaN                            NaN   \n",
      "S004                                NaN                            NaN   \n",
      "S005                                NaN                            NaN   \n",
      "...                                 ...                            ...   \n",
      "S127                                NaN                            NaN   \n",
      "S128                                NaN                            NaN   \n",
      "S129                                NaN                            NaN   \n",
      "S130                                NaN                            NaN   \n",
      "S131                                NaN                            NaN   \n",
      "\n",
      "Name       New_Tumor_Event_Hormone_Therapy  \\\n",
      "Sample_ID                                    \n",
      "S001                                   NaN   \n",
      "S002                                   NaN   \n",
      "S003                                   NaN   \n",
      "S004                                   NaN   \n",
      "S005                                   NaN   \n",
      "...                                    ...   \n",
      "S127                                   NaN   \n",
      "S128                                   NaN   \n",
      "S129                                   NaN   \n",
      "S130                                   NaN   \n",
      "S131                                   NaN   \n",
      "\n",
      "Name       New_Tumor_Event_Targeted_Molecular_Therapy  \n",
      "Sample_ID                                              \n",
      "S001                                              NaN  \n",
      "S002                                              NaN  \n",
      "S003                                              NaN  \n",
      "S004                                              NaN  \n",
      "S005                                              NaN  \n",
      "...                                               ...  \n",
      "S127                                              NaN  \n",
      "S128                                              NaN  \n",
      "S129                                              NaN  \n",
      "S130                                              NaN  \n",
      "S131                                              NaN  \n",
      "\n",
      "[131 rows x 86 columns]\n",
      "['Patient_ID', 'Sample_Tumor_Normal', 'Participant_Procurement_Age', 'Participant_Gender', 'Participant_Race', 'Participant_Ethnicity', 'Participant_Jewish_Heritage', 'Participant_History_Malignancy', 'Participant_History_Chemotherapy', 'Participant_History_Neo-adjuvant_Treatment', 'Participant_History_Radiation_Therapy', 'Participant_History_Hormonal_Therapy', 'Aliquots_Plasma', 'Blood_Collection_Time', 'Blood_Collection_Method', 'Anesthesia_Time', 'Tumor_Surgical_Devascularized_Time', 'Tumor_Sample_Number', 'Tumor_Sample_1_Weight', 'Tumor_Sample_1_LN2_Time', 'Tumor_Sample_1_Ischemia_Time', 'Tumor_Sample_2_Weight', 'Tumor_Sample_2_LN2_Time', 'Tumor_Sample_2_Ischemia_Time', 'Tumor_Sample_3_Weight', 'Tumor_Sample_3_LN2_Time', 'Tumor_Sample_3_Ischemia_Time', 'Tumor_Sample_4_Weight', 'Tumor_Sample_4_LN2_Time', 'Tumor_Sample_4_Ischemia_Time', 'Tumor_Sample_5_Weight', 'Tumor_Sample_5_LN2_Time', 'Tumor_Sample_5_Ischemia_Time', 'Normal_Sample_Number', 'Normal_Sample_1_Surgical_Devascularized_Time', 'Normal_Sample_1_Weight', 'Normal_Sample_1_LN2_Time', 'Normal_Sample_1_Ischemia_Time', 'Normal_Sample_2_Surgical_Devascularized_Time', 'Normal_Sample_2_Weight', 'Normal_Sample_2_LN2_Time', 'Normal_Sample_2_Ischemia_Time', 'Normal_Sample_3_Surgical_Devascularized_Time', 'Normal_Sample_3_Weight', 'Normal_Sample_3_LN2_Time', 'Normal_Sample_3_Ischemia_Time', 'Normal_Sample_4_Surgical_Devascularized_Time', 'Normal_Sample_4_Weight', 'Normal_Sample_4_LN2_Time', 'Normal_Sample_4_Ischemia_Time', 'Normal_Sample_5_Surgical_Devascularized_Time', 'Normal_Sample_5_Weight', 'Normal_Sample_5_LN2_Time', 'Normal_Sample_5_Ischemia_Time', 'Origin_Site_Disease', 'Anatomic_Site_Tumor', 'Anatomic_Lateral_Position_Tumor', 'Histological_Subtype', 'Method_of_Pathologic_Diagnosis', 'Tumor_Stage_Ovary_FIGO', 'Tumor_Grade', 'Tumor_Residual_Disease_Max_Diameter', 'Days_Between_Collection_And_Last_Contact', 'Vital_Status', 'Days_Between_Collection_And_Death', 'Tumor_Status', 'Review_Of_Initial_Pathological_Findings', 'Pathology_Review_Consistent_With_Diagnosis', 'Adjuvant_Radiation_Therapy', 'Adjuvant_Pharmaceutical_Therapy', 'Adjuvant_Immunotherapy', 'Adjuvant_Hormone_Therapy', 'Adjuvant_Targeted_Molecular_Therapy', 'Response_After_Surgery_And_Adjuvant_Therapies', 'New_Tumor_Event_After_Initial_Treatment', 'New_Tumor_Event_Type', 'New_Tumor_Event_Site', 'Other_New_Tumor_Event_Site', 'Days_Between_Collection_And_New_Tumor_Event', 'New_Tumor_Event_Diagnosis', 'New_Tumor_Event_Surgery', 'Days_Between_Collection_And_New_Tumor_Event_Surgery', 'New_Tumor_Event_Chemotherapy', 'New_Tumor_Event_Immunotherapy', 'New_Tumor_Event_Hormone_Therapy', 'New_Tumor_Event_Targeted_Molecular_Therapy']\n"
     ]
    }
   ],
   "source": [
    "#CLINICAL DATA\n",
    "clinical = g.get_clinical()\n",
    "print(clinical)\n",
    "print(clinical.columns.tolist())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "#clinical.to_excel('../../Desktop/TESTING_ccrcc.xlsx')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Help on Ccrcc in module cptac.ccrcc object:\n",
      "\n",
      "class Ccrcc(cptac.dataset.DataSet)\n",
      " |  Ccrcc(version='latest')\n",
      " |  \n",
      " |  Note that all cancer datasets are class objects that inherit from cptac.dataset. Therefore\n",
      " |  the same function calls exist for cptac.Endometrial, cptac.Colon, etc.\n",
      " |  \n",
      " |  Method resolution order:\n",
      " |      Ccrcc\n",
      " |      cptac.dataset.DataSet\n",
      " |      builtins.object\n",
      " |  \n",
      " |  Methods defined here:\n",
      " |  \n",
      " |  __init__(self, version='latest')\n",
      " |      Load all of the ccrcc dataframes as values in the self._data dict variable, with names as keys, and format them properly.\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Methods inherited from cptac.dataset.DataSet:\n",
      " |  \n",
      " |  define(self, term)\n",
      " |      Print the definition a term, if it is in the dataset's list of definitions.\n",
      " |      \n",
      " |      Parameters:\n",
      " |      term (str): term to be defined\n",
      " |      \n",
      " |      Returns: None\n",
      " |  \n",
      " |  get_CNV(self)\n",
      " |      Get the CNV dataframe.\n",
      " |  \n",
      " |  get_acetylproteomics(self)\n",
      " |      Get the acetylproteomics dataframe.\n",
      " |  \n",
      " |  get_cancer_type(self)\n",
      " |      Return the cancer type for this dataset, as a string.\n",
      " |  \n",
      " |  get_circular_RNA(self)\n",
      " |      Get the circular_RNA dataframe.\n",
      " |  \n",
      " |  get_clinical(self)\n",
      " |      Get the clinical dataframe.\n",
      " |  \n",
      " |  get_derived_molecular(self)\n",
      " |      Get the derived_molecular dataframe.\n",
      " |  \n",
      " |  get_experimental_design(self)\n",
      " |      Get the experimental_design dataframe.\n",
      " |  \n",
      " |  get_gene_fusion(self)\n",
      " |      Get the gene_fusion dataframe.\n",
      " |  \n",
      " |  get_lipidomics(self)\n",
      " |      Get the lipidomics dataframe.\n",
      " |  \n",
      " |  get_medical_history(self)\n",
      " |      Get the medical_history dataframe.\n",
      " |  \n",
      " |  get_metabolomics(self)\n",
      " |      Get the metabolomics dataframe.\n",
      " |  \n",
      " |  get_methylation(self)\n",
      " |      Get the methylation dataframe.\n",
      " |  \n",
      " |  get_miRNA(self)\n",
      " |      Get the miRNA dataframe.\n",
      " |  \n",
      " |  get_phosphoproteomics(self)\n",
      " |      Get the phosphoproteomics dataframe.\n",
      " |  \n",
      " |  get_phosphoproteomics_gene(self)\n",
      " |      Get the phosphoproteomics_gene dataframe. The gene level phosphorylation measurement is an aggregate metric which potentially averages together individual measurements of different sites. Use get_phosphoproteomics() to view the data for individual sites.\n",
      " |  \n",
      " |  get_phosphosites(self, genes)\n",
      " |      Returns dataframe with all phosphosites of specified gene or list of genes.\n",
      " |      \n",
      " |      Parameters:\n",
      " |      genes (str, or list or array-like of str): gene or list of genes to use to select phosphosites. str if single, list or array-like of str if multiple.\n",
      " |      \n",
      " |      Returns:\n",
      " |      pandas DataFrame: The phosphoproteomics for the specified gene(s).\n",
      " |  \n",
      " |  get_proteomics(self)\n",
      " |      Get the proteomics dataframe.\n",
      " |  \n",
      " |  get_somatic_mutation(self)\n",
      " |      Get the somatic_mutation dataframe.\n",
      " |  \n",
      " |  get_somatic_mutation_binary(self)\n",
      " |      Get the somatic_mutation_binary dataframe, which has a binary value indicating, for each location on each gene, whether there was a mutation in that gene at that location, for each sample.\n",
      " |  \n",
      " |  get_transcriptomics(self)\n",
      " |      Get the transcriptomics dataframe.\n",
      " |  \n",
      " |  get_treatment(self)\n",
      " |      Get the treatment dataframe.\n",
      " |  \n",
      " |  how_to_cite(self)\n",
      " |      Print instructions for citing the data.\n",
      " |  \n",
      " |  join_metadata_to_metadata(self, df1_name, df2_name, cols1=None, cols2=None)\n",
      " |      Take specified column(s) from one metadata dataframe, and join to specified columns(s) from another metadata dataframe. Intersection (inner join) of indices is used.\n",
      " |      \n",
      " |      Parameters:\n",
      " |      df1_name (str): Name of first metadata dataframe to select columns from.\n",
      " |      df2_name (str): Name of second metadata dataframe to select columns from.\n",
      " |      cols1 (str, or list or array-like of str, optional): Column(s) to select from df1_name. str if one key, list or array-like of str if multiple. Default of None will select entire dataframe.\n",
      " |      cols2 (str, or list or array-like of str, optional): Column(s) to select from df2_name. str if one key, list or array-like of str if multiple. Default of None will select entire dataframe.\n",
      " |      \n",
      " |      Returns:\n",
      " |      pandas DataFrame: The selected columns from the two metadata dataframes, joined into one dataframe.\n",
      " |  \n",
      " |  join_metadata_to_mutations(self, metadata_df_name, mutations_genes, metadata_cols=None, mutations_filter=None, show_location=True)\n",
      " |      Select all mutations for specified gene(s), and joins them to all or part of the given metadata dataframe. Intersection (inner join) of indices is used. Each location or mutation cell contains a list, which contains the one or more location or mutation values corresponding to that sample for that gene, or a value indicating that the sample didn't have a mutation in that gene.\n",
      " |      \n",
      " |      Parameters:\n",
      " |      metadata_df_name (str): Name of metadata dataframe to join the mutation data to.\n",
      " |      mutations_genes (str, or list or array-like of str): The gene(s) to get mutation data for. str if one gene, list or array-like of str if multiple.\n",
      " |      metadata_cols (str, or list or array-like of str, optional): Gene(s) to select from the metadata dataframe. str if one gene, list or array-like of str if multiple. Default will select entire dataframe.\n",
      " |      mutations_filter (list, optional): List of mutations to prioritize when filtering out multiple mutations, in order of priority. If none of the multiple mutations in a sample are included in filter_prefer, the function will automatically prioritize truncation over missense mutations, and then mutations earlier in the sequence over later mutations. Passing an empty list will cause this default hierarchy to be applied to all samples. Default parameter of None will cause no filtering to be done, and all mutation data will be included, in a list.\n",
      " |      show_location (bool, optional): Whether to include the Location column from the mutation dataframe. Defaults to True.\n",
      " |      \n",
      " |      Returns:\n",
      " |      pandas DataFrame: The mutations for the specified gene, joined to all or part of the metadata dataframe. Each location or mutation cell contains a list, which contains the one or more location or mutation values corresponding to that sample for that gene, or a value indicating that the sample didn't have a mutation in that gene.\n",
      " |  \n",
      " |  join_metadata_to_omics(self, metadata_df_name, omics_df_name, metadata_cols=None, omics_genes=None)\n",
      " |      Joins columns from a metadata dataframe (clinical, derived_molecular, or experimental_design) to part or all of an omics dataframe. Intersection (inner join) of indices is used.\n",
      " |      \n",
      " |      Parameters:\n",
      " |      metadata_df_name (str): Name of metadata dataframe to select columns from.\n",
      " |      omics_df_name (str): Name of omics dataframe to join the metadata columns to.\n",
      " |      metadata_cols (str, or list or array-like of str, optional): Column(s) to select from the metadata dataframe. str if one gene, list or array-like of str if multiple. Default is None, which will select the entire metadata dataframe.\n",
      " |      omics_genes (str, or list or array-like of str, optional): Gene(s) to select data for from the omics dataframe. str if one gene, list or array-like of str if multiple. Default is None, which will select entire dataframe.\n",
      " |      \n",
      " |      Returns:\n",
      " |      pandas DataFrame: The selected metadata columns, joined with all or part of the omics dataframe.\n",
      " |  \n",
      " |  join_omics_to_mutations(self, omics_df_name, mutations_genes, omics_genes=None, mutations_filter=None, show_location=True)\n",
      " |      Select all mutations for specified gene(s), and joins them to all or part of the given omics dataframe. Intersection (inner join) of indices is used. Each location or mutation cell contains a list, which contains the one or more location or mutation values corresponding to that sample for that gene, or a value indicating that the sample didn't have a mutation in that gene.\n",
      " |      \n",
      " |      Parameters:\n",
      " |      omics_df (str): Name of omics dataframe to join the mutation data to.\n",
      " |      mutations_genes (str, or list or array-like of str): The gene(s) to get mutation data for. str if one gene, list or array-like of str if multiple.\n",
      " |      omics_genes (str, or list or array-like of str, optional): Gene(s) to select from the omics dataframe. str if one gene, list or array-like of str if multiple. Default will select entire dataframe.\n",
      " |      mutations_filter (list, optional): List of mutations to prioritize when filtering out multiple mutations, in order of priority. If none of the multiple mutations in a sample are included in filter_prefer, the function will automatically prioritize truncation over missense mutations, and then mutations earlier in the sequence over later mutations. Passing an empty list will cause this default hierarchy to be applied to all samples. Default parameter of None will cause no filtering to be done, and all mutation data will be included, in a list.\n",
      " |      show_location (bool, optional): Whether to include the Location column from the mutation dataframe. Defaults to True.\n",
      " |      \n",
      " |      Returns:\n",
      " |      pandas DataFrame: The mutations for the specified gene, joined to all or part of the omics dataframe. Each location or mutation cell contains a list, which contains the one or more location or mutation values corresponding to that sample for that gene, or a value indicating that the sample didn't have a mutation in that gene.\n",
      " |  \n",
      " |  join_omics_to_omics(self, df1_name, df2_name, genes1=None, genes2=None)\n",
      " |      Take specified column(s) from one omics dataframe, and join to specified columns(s) from another omics dataframe. Intersection (inner join) of indices is used.\n",
      " |      \n",
      " |      Parameters:\n",
      " |      df1_name (str): Name of first omics dataframe to select columns from.\n",
      " |      df2_name (str): Name of second omics dataframe to select columns from.\n",
      " |      genes1 (str, or list or array-like of str, optional): Gene(s) for column(s) to select from df1_name. str if one key, list or array-like of str if multiple. Default of None will select entire dataframe.\n",
      " |      genes2 (str, or list or array-like of str, optional): Gene(s) for Column(s) to select from df2_name. str if one key, list or array-like of str if multiple. Default of None will select entire dataframe.\n",
      " |      \n",
      " |      Returns:\n",
      " |      pandas DataFrame: The selected columns from the two omics dataframes, joined into one dataframe.\n",
      " |  \n",
      " |  list_data(self)\n",
      " |      Print list of loaded dataframes and dimensions.\n",
      " |  \n",
      " |  list_definitions(self)\n",
      " |      Print all terms defined in the dataset's list of definitions.\n",
      " |  \n",
      " |  reduce_multiindex(self, df, levels_to_drop=None, flatten=False, sep='_')\n",
      " |      Drop levels from and/or flatten the column axis of a dataframe with a column multiindex.\n",
      " |      \n",
      " |      Parameters:\n",
      " |      df (pandas DataFrame): The dataframe to make the changes to.\n",
      " |      levels_to_drop (str, int, or list or array-like of str or int, optional): Levels, or indices of levels, to drop from the dataframe's column multiindex. These must match the names or indices of actual levels of the multiindex. Must be either all strings, or all ints. Default of None will drop no levels.\n",
      " |      flatten (bool, optional): Whether or not to flatten the multiindex. Default of False will not flatten.\n",
      " |      sep (str, optional): String to use to separate index levels when flattening. Default is underscore.\n",
      " |      \n",
      " |      Returns:\n",
      " |      pandas DataFrame: The dataframe, with the desired column index changes made.\n",
      " |  \n",
      " |  search(self, term)\n",
      " |      Search for a term in a web browser.\n",
      " |      \n",
      " |      Parameters:\n",
      " |      term (str): term to be searched\n",
      " |      \n",
      " |      Returns: None\n",
      " |  \n",
      " |  version(self)\n",
      " |      Return the dataset version of this instance, as a string.\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Data descriptors inherited from cptac.dataset.DataSet:\n",
      " |  \n",
      " |  __dict__\n",
      " |      dictionary for instance variables (if defined)\n",
      " |  \n",
      " |  __weakref__\n",
      " |      list of weak references to the object (if defined)\n",
      "\n"
     ]
    }
   ],
   "source": [
    "help(k)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
