{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Dataframe join examples\n",
    "\n",
    "In this use case, we provide several examples of how to use the built-in <code>cptac</code> functions for joining different dataframes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                    \r"
     ]
    }
   ],
   "source": [
    "import cptac\n",
    "en = cptac.Endometrial()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## General format\n",
    "\n",
    "In all of the join functions, you specify the dataframes you want to join by passing their names to the appropriate parameters in the function call. The function will automatically check that the dataframes whose names you provided are valid for the join function, and print an error message if they aren't.\n",
    "\n",
    "Whenever a column from an -omics dataframe is included in a joined table, the name of the -omics dataframe it came from is joined to the column header, to avoid confusion.\n",
    "\n",
    "If you wish to only include particular columns in the join, pass them to the appropriate parameters in the join function. All such parameters will accept either a single column name as a string, or a list of column name strings. In this use case, we will usually only select specific columns for readability, but you could select the whole dataframe in all these cases, except for the mutations dataframe.\n",
    "\n",
    "The join functions use logic analogous to an SQL INNER JOIN."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## <code>join_omics_to_omics</code>\n",
    "\n",
    "The <code>join_omics_to_omics</code> function joins two -omics dataframes to each other:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>A1BG_proteomics</th>\n",
       "      <th>A2M_proteomics</th>\n",
       "      <th>A2ML1_proteomics</th>\n",
       "      <th>A4GALT_proteomics</th>\n",
       "      <th>AAAS_proteomics</th>\n",
       "      <th>AACS_proteomics</th>\n",
       "      <th>AADAT_proteomics</th>\n",
       "      <th>AAED1_proteomics</th>\n",
       "      <th>AAGAB_proteomics</th>\n",
       "      <th>AAK1_proteomics</th>\n",
       "      <th>...</th>\n",
       "      <th>ZZZ3-S397_phosphoproteomics</th>\n",
       "      <th>ZZZ3-S411_phosphoproteomics</th>\n",
       "      <th>ZZZ3-S420_phosphoproteomics</th>\n",
       "      <th>ZZZ3-S424_phosphoproteomics</th>\n",
       "      <th>ZZZ3-S426_phosphoproteomics</th>\n",
       "      <th>ZZZ3-S468_phosphoproteomics</th>\n",
       "      <th>ZZZ3-S89_phosphoproteomics</th>\n",
       "      <th>ZZZ3-T415_phosphoproteomics</th>\n",
       "      <th>ZZZ3-T418_phosphoproteomics</th>\n",
       "      <th>ZZZ3-Y399_phosphoproteomics</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sample_ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>S001</th>\n",
       "      <td>-1.180</td>\n",
       "      <td>-0.8630</td>\n",
       "      <td>-0.802</td>\n",
       "      <td>0.222</td>\n",
       "      <td>0.2560</td>\n",
       "      <td>0.6650</td>\n",
       "      <td>1.2800</td>\n",
       "      <td>-0.3390</td>\n",
       "      <td>0.412</td>\n",
       "      <td>-0.664</td>\n",
       "      <td>...</td>\n",
       "      <td>0.18400</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.20500</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S002</th>\n",
       "      <td>-0.685</td>\n",
       "      <td>-1.0700</td>\n",
       "      <td>-0.684</td>\n",
       "      <td>0.984</td>\n",
       "      <td>0.1350</td>\n",
       "      <td>0.3340</td>\n",
       "      <td>1.3000</td>\n",
       "      <td>0.1390</td>\n",
       "      <td>1.330</td>\n",
       "      <td>-0.367</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.17100</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.393</td>\n",
       "      <td>-0.17100</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.29</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.1605</td>\n",
       "      <td>-0.0635</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S003</th>\n",
       "      <td>-0.528</td>\n",
       "      <td>-1.3200</td>\n",
       "      <td>0.435</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.2400</td>\n",
       "      <td>1.0400</td>\n",
       "      <td>-0.0213</td>\n",
       "      <td>-0.0479</td>\n",
       "      <td>0.419</td>\n",
       "      <td>-0.500</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S005</th>\n",
       "      <td>-1.670</td>\n",
       "      <td>-1.1900</td>\n",
       "      <td>-0.443</td>\n",
       "      <td>0.243</td>\n",
       "      <td>-0.0993</td>\n",
       "      <td>0.7570</td>\n",
       "      <td>0.7400</td>\n",
       "      <td>-0.9290</td>\n",
       "      <td>0.229</td>\n",
       "      <td>-0.223</td>\n",
       "      <td>...</td>\n",
       "      <td>0.13970</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.55900</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.2980</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S006</th>\n",
       "      <td>-0.374</td>\n",
       "      <td>-0.0206</td>\n",
       "      <td>-0.537</td>\n",
       "      <td>0.311</td>\n",
       "      <td>0.3750</td>\n",
       "      <td>0.0131</td>\n",
       "      <td>-1.1000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.565</td>\n",
       "      <td>-0.101</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.15875</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.196</td>\n",
       "      <td>0.06175</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.2900</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 84211 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           A1BG_proteomics  A2M_proteomics  A2ML1_proteomics  \\\n",
       "Sample_ID                                                      \n",
       "S001                -1.180         -0.8630            -0.802   \n",
       "S002                -0.685         -1.0700            -0.684   \n",
       "S003                -0.528         -1.3200             0.435   \n",
       "S005                -1.670         -1.1900            -0.443   \n",
       "S006                -0.374         -0.0206            -0.537   \n",
       "\n",
       "           A4GALT_proteomics  AAAS_proteomics  AACS_proteomics  \\\n",
       "Sample_ID                                                        \n",
       "S001                   0.222           0.2560           0.6650   \n",
       "S002                   0.984           0.1350           0.3340   \n",
       "S003                     NaN          -0.2400           1.0400   \n",
       "S005                   0.243          -0.0993           0.7570   \n",
       "S006                   0.311           0.3750           0.0131   \n",
       "\n",
       "           AADAT_proteomics  AAED1_proteomics  AAGAB_proteomics  \\\n",
       "Sample_ID                                                         \n",
       "S001                 1.2800           -0.3390             0.412   \n",
       "S002                 1.3000            0.1390             1.330   \n",
       "S003                -0.0213           -0.0479             0.419   \n",
       "S005                 0.7400           -0.9290             0.229   \n",
       "S006                -1.1000               NaN             0.565   \n",
       "\n",
       "           AAK1_proteomics  ...  ZZZ3-S397_phosphoproteomics  \\\n",
       "Sample_ID                   ...                                \n",
       "S001                -0.664  ...                      0.18400   \n",
       "S002                -0.367  ...                     -0.17100   \n",
       "S003                -0.500  ...                          NaN   \n",
       "S005                -0.223  ...                      0.13970   \n",
       "S006                -0.101  ...                     -0.15875   \n",
       "\n",
       "           ZZZ3-S411_phosphoproteomics  ZZZ3-S420_phosphoproteomics  \\\n",
       "Sample_ID                                                             \n",
       "S001                               NaN                          NaN   \n",
       "S002                               NaN                          NaN   \n",
       "S003                               NaN                          NaN   \n",
       "S005                               NaN                          NaN   \n",
       "S006                               NaN                          NaN   \n",
       "\n",
       "           ZZZ3-S424_phosphoproteomics  ZZZ3-S426_phosphoproteomics  \\\n",
       "Sample_ID                                                             \n",
       "S001                               NaN                     -0.20500   \n",
       "S002                            -0.393                     -0.17100   \n",
       "S003                               NaN                          NaN   \n",
       "S005                               NaN                     -0.55900   \n",
       "S006                             0.196                      0.06175   \n",
       "\n",
       "           ZZZ3-S468_phosphoproteomics  ZZZ3-S89_phosphoproteomics  \\\n",
       "Sample_ID                                                            \n",
       "S001                               NaN                         NaN   \n",
       "S002                               NaN                        0.29   \n",
       "S003                               NaN                         NaN   \n",
       "S005                               NaN                         NaN   \n",
       "S006                               NaN                         NaN   \n",
       "\n",
       "           ZZZ3-T415_phosphoproteomics  ZZZ3-T418_phosphoproteomics  \\\n",
       "Sample_ID                                                             \n",
       "S001                               NaN                          NaN   \n",
       "S002                               NaN                       0.1605   \n",
       "S003                               NaN                          NaN   \n",
       "S005                               NaN                          NaN   \n",
       "S006                               NaN                          NaN   \n",
       "\n",
       "           ZZZ3-Y399_phosphoproteomics  \n",
       "Sample_ID                               \n",
       "S001                               NaN  \n",
       "S002                           -0.0635  \n",
       "S003                               NaN  \n",
       "S005                            0.2980  \n",
       "S006                           -0.2900  \n",
       "\n",
       "[5 rows x 84211 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "prot_and_phos = en.join_omics_to_omics(df1_name=\"proteomics\", df2_name=\"phosphoproteomics\")\n",
    "prot_and_phos.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Joining only specific columns.\n",
    "(Note that when a gene is selected from the phosphoproteomics dataframe, data for all sites of the gene are selected. The same is done for acetylproteomics data.)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>A1BG_proteomics</th>\n",
       "      <th>PIK3CA-S312_phosphoproteomics</th>\n",
       "      <th>PIK3CA-T313_phosphoproteomics</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sample_ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>S001</th>\n",
       "      <td>-1.180</td>\n",
       "      <td>-0.00615</td>\n",
       "      <td>0.0731</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S002</th>\n",
       "      <td>-0.685</td>\n",
       "      <td>-0.02220</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S003</th>\n",
       "      <td>-0.528</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0830</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S005</th>\n",
       "      <td>-1.670</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.8460</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S006</th>\n",
       "      <td>-0.374</td>\n",
       "      <td>0.43600</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           A1BG_proteomics  PIK3CA-S312_phosphoproteomics  \\\n",
       "Sample_ID                                                   \n",
       "S001                -1.180                       -0.00615   \n",
       "S002                -0.685                       -0.02220   \n",
       "S003                -0.528                            NaN   \n",
       "S005                -1.670                            NaN   \n",
       "S006                -0.374                        0.43600   \n",
       "\n",
       "           PIK3CA-T313_phosphoproteomics  \n",
       "Sample_ID                                 \n",
       "S001                              0.0731  \n",
       "S002                                 NaN  \n",
       "S003                              0.0830  \n",
       "S005                             -0.8460  \n",
       "S006                                 NaN  "
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "prot_and_phos_selected = en.join_omics_to_omics(\n",
    "    df1_name=\"proteomics\", \n",
    "    df2_name=\"phosphoproteomics\", \n",
    "    genes1=\"A1BG\", \n",
    "    genes2=\"PIK3CA\")\n",
    "\n",
    "prot_and_phos_selected.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## <code>join_metadata_to_omics</code>\n",
    "\n",
    "The <code>join_metadata_to_omics</code> function joins a metadata dataframe (e.g. clinical or derived_molecular) with an -omics dataframe:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Patient_ID</th>\n",
       "      <th>Proteomics_Tumor_Normal</th>\n",
       "      <th>Country</th>\n",
       "      <th>Histologic_Grade_FIGO</th>\n",
       "      <th>Myometrial_invasion_Specify</th>\n",
       "      <th>Histologic_type</th>\n",
       "      <th>Treatment_naive</th>\n",
       "      <th>Tumor_purity</th>\n",
       "      <th>Path_Stage_Primary_Tumor-pT</th>\n",
       "      <th>Path_Stage_Reg_Lymph_Nodes-pN</th>\n",
       "      <th>...</th>\n",
       "      <th>ZWILCH_transcriptomics</th>\n",
       "      <th>ZWINT_transcriptomics</th>\n",
       "      <th>ZXDA_transcriptomics</th>\n",
       "      <th>ZXDB_transcriptomics</th>\n",
       "      <th>ZXDC_transcriptomics</th>\n",
       "      <th>ZYG11A_transcriptomics</th>\n",
       "      <th>ZYG11B_transcriptomics</th>\n",
       "      <th>ZYX_transcriptomics</th>\n",
       "      <th>ZZEF1_transcriptomics</th>\n",
       "      <th>ZZZ3_transcriptomics</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sample_ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>S001</th>\n",
       "      <td>C3L-00006</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>United States</td>\n",
       "      <td>FIGO grade 1</td>\n",
       "      <td>under 50 %</td>\n",
       "      <td>Endometrioid</td>\n",
       "      <td>YES</td>\n",
       "      <td>Normal</td>\n",
       "      <td>pT1a (FIGO IA)</td>\n",
       "      <td>pN0</td>\n",
       "      <td>...</td>\n",
       "      <td>11.06</td>\n",
       "      <td>10.73</td>\n",
       "      <td>8.40</td>\n",
       "      <td>9.78</td>\n",
       "      <td>10.88</td>\n",
       "      <td>5.93</td>\n",
       "      <td>11.52</td>\n",
       "      <td>10.23</td>\n",
       "      <td>11.50</td>\n",
       "      <td>11.47</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S002</th>\n",
       "      <td>C3L-00008</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>United States</td>\n",
       "      <td>FIGO grade 1</td>\n",
       "      <td>under 50 %</td>\n",
       "      <td>Endometrioid</td>\n",
       "      <td>YES</td>\n",
       "      <td>Normal</td>\n",
       "      <td>pT1a (FIGO IA)</td>\n",
       "      <td>pNX</td>\n",
       "      <td>...</td>\n",
       "      <td>10.87</td>\n",
       "      <td>11.43</td>\n",
       "      <td>8.39</td>\n",
       "      <td>9.14</td>\n",
       "      <td>10.38</td>\n",
       "      <td>7.25</td>\n",
       "      <td>11.64</td>\n",
       "      <td>10.64</td>\n",
       "      <td>11.26</td>\n",
       "      <td>11.57</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S003</th>\n",
       "      <td>C3L-00032</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>United States</td>\n",
       "      <td>FIGO grade 2</td>\n",
       "      <td>under 50 %</td>\n",
       "      <td>Endometrioid</td>\n",
       "      <td>YES</td>\n",
       "      <td>Normal</td>\n",
       "      <td>pT1a (FIGO IA)</td>\n",
       "      <td>pN0</td>\n",
       "      <td>...</td>\n",
       "      <td>10.06</td>\n",
       "      <td>10.13</td>\n",
       "      <td>8.35</td>\n",
       "      <td>9.27</td>\n",
       "      <td>10.46</td>\n",
       "      <td>6.85</td>\n",
       "      <td>11.60</td>\n",
       "      <td>10.21</td>\n",
       "      <td>11.51</td>\n",
       "      <td>11.09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S005</th>\n",
       "      <td>C3L-00090</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>United States</td>\n",
       "      <td>FIGO grade 2</td>\n",
       "      <td>under 50 %</td>\n",
       "      <td>Endometrioid</td>\n",
       "      <td>YES</td>\n",
       "      <td>Normal</td>\n",
       "      <td>pT1a (FIGO IA)</td>\n",
       "      <td>pNX</td>\n",
       "      <td>...</td>\n",
       "      <td>10.29</td>\n",
       "      <td>10.41</td>\n",
       "      <td>9.10</td>\n",
       "      <td>9.59</td>\n",
       "      <td>10.15</td>\n",
       "      <td>7.89</td>\n",
       "      <td>11.90</td>\n",
       "      <td>10.21</td>\n",
       "      <td>11.34</td>\n",
       "      <td>11.51</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S006</th>\n",
       "      <td>C3L-00098</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>United States</td>\n",
       "      <td>NaN</td>\n",
       "      <td>under 50 %</td>\n",
       "      <td>Serous</td>\n",
       "      <td>YES</td>\n",
       "      <td>Normal</td>\n",
       "      <td>pT1a (FIGO IA)</td>\n",
       "      <td>pNX</td>\n",
       "      <td>...</td>\n",
       "      <td>10.36</td>\n",
       "      <td>11.24</td>\n",
       "      <td>8.60</td>\n",
       "      <td>9.44</td>\n",
       "      <td>11.80</td>\n",
       "      <td>9.32</td>\n",
       "      <td>11.97</td>\n",
       "      <td>9.77</td>\n",
       "      <td>11.37</td>\n",
       "      <td>12.35</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 28083 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          Patient_ID Proteomics_Tumor_Normal        Country  \\\n",
       "Sample_ID                                                     \n",
       "S001       C3L-00006                   Tumor  United States   \n",
       "S002       C3L-00008                   Tumor  United States   \n",
       "S003       C3L-00032                   Tumor  United States   \n",
       "S005       C3L-00090                   Tumor  United States   \n",
       "S006       C3L-00098                   Tumor  United States   \n",
       "\n",
       "          Histologic_Grade_FIGO Myometrial_invasion_Specify Histologic_type  \\\n",
       "Sample_ID                                                                     \n",
       "S001               FIGO grade 1                  under 50 %    Endometrioid   \n",
       "S002               FIGO grade 1                  under 50 %    Endometrioid   \n",
       "S003               FIGO grade 2                  under 50 %    Endometrioid   \n",
       "S005               FIGO grade 2                  under 50 %    Endometrioid   \n",
       "S006                        NaN                  under 50 %          Serous   \n",
       "\n",
       "          Treatment_naive Tumor_purity Path_Stage_Primary_Tumor-pT  \\\n",
       "Sample_ID                                                            \n",
       "S001                  YES       Normal              pT1a (FIGO IA)   \n",
       "S002                  YES       Normal              pT1a (FIGO IA)   \n",
       "S003                  YES       Normal              pT1a (FIGO IA)   \n",
       "S005                  YES       Normal              pT1a (FIGO IA)   \n",
       "S006                  YES       Normal              pT1a (FIGO IA)   \n",
       "\n",
       "          Path_Stage_Reg_Lymph_Nodes-pN  ... ZWILCH_transcriptomics  \\\n",
       "Sample_ID                                ...                          \n",
       "S001                                pN0  ...                  11.06   \n",
       "S002                                pNX  ...                  10.87   \n",
       "S003                                pN0  ...                  10.06   \n",
       "S005                                pNX  ...                  10.29   \n",
       "S006                                pNX  ...                  10.36   \n",
       "\n",
       "          ZWINT_transcriptomics ZXDA_transcriptomics ZXDB_transcriptomics  \\\n",
       "Sample_ID                                                                   \n",
       "S001                      10.73                 8.40                 9.78   \n",
       "S002                      11.43                 8.39                 9.14   \n",
       "S003                      10.13                 8.35                 9.27   \n",
       "S005                      10.41                 9.10                 9.59   \n",
       "S006                      11.24                 8.60                 9.44   \n",
       "\n",
       "           ZXDC_transcriptomics  ZYG11A_transcriptomics  \\\n",
       "Sample_ID                                                 \n",
       "S001                      10.88                    5.93   \n",
       "S002                      10.38                    7.25   \n",
       "S003                      10.46                    6.85   \n",
       "S005                      10.15                    7.89   \n",
       "S006                      11.80                    9.32   \n",
       "\n",
       "           ZYG11B_transcriptomics ZYX_transcriptomics ZZEF1_transcriptomics  \\\n",
       "Sample_ID                                                                     \n",
       "S001                        11.52               10.23                 11.50   \n",
       "S002                        11.64               10.64                 11.26   \n",
       "S003                        11.60               10.21                 11.51   \n",
       "S005                        11.90               10.21                 11.34   \n",
       "S006                        11.97                9.77                 11.37   \n",
       "\n",
       "          ZZZ3_transcriptomics  \n",
       "Sample_ID                       \n",
       "S001                     11.47  \n",
       "S002                     11.57  \n",
       "S003                     11.09  \n",
       "S005                     11.51  \n",
       "S006                     12.35  \n",
       "\n",
       "[5 rows x 28083 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clin_and_tran = en.join_metadata_to_omics(metadata_df_name=\"clinical\", omics_df_name=\"transcriptomics\")\n",
    "clin_and_tran.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Joining only specific columns:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Age</th>\n",
       "      <th>Histologic_type</th>\n",
       "      <th>ZZZ3_transcriptomics</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sample_ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>S001</th>\n",
       "      <td>64.0</td>\n",
       "      <td>Endometrioid</td>\n",
       "      <td>11.47</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S002</th>\n",
       "      <td>58.0</td>\n",
       "      <td>Endometrioid</td>\n",
       "      <td>11.57</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S003</th>\n",
       "      <td>50.0</td>\n",
       "      <td>Endometrioid</td>\n",
       "      <td>11.09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S005</th>\n",
       "      <td>75.0</td>\n",
       "      <td>Endometrioid</td>\n",
       "      <td>11.51</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S006</th>\n",
       "      <td>63.0</td>\n",
       "      <td>Serous</td>\n",
       "      <td>12.35</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            Age Histologic_type  ZZZ3_transcriptomics\n",
       "Sample_ID                                            \n",
       "S001       64.0    Endometrioid                 11.47\n",
       "S002       58.0    Endometrioid                 11.57\n",
       "S003       50.0    Endometrioid                 11.09\n",
       "S005       75.0    Endometrioid                 11.51\n",
       "S006       63.0          Serous                 12.35"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clin_and_tran = en.join_metadata_to_omics(\n",
    "    metadata_df_name=\"clinical\", \n",
    "    omics_df_name=\"transcriptomics\", \n",
    "    metadata_cols = [\"Age\", \"Histologic_type\"], \n",
    "    omics_genes=\"ZZZ3\")\n",
    "\n",
    "clin_and_tran.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## <code>join_metadata_to_metadata</code>\n",
    "\n",
    "The <code>join_metadata_to_metadata</code> function joins two metadata dataframes (e.g. clinical or derived_molecular) to each other. Note how we passed a column name to select from the clinical dataframe, but passing <code>None</code> for the column parameter for the derived_molecular dataframe caused the entire dataframe to be selected. We could have omitted the <code>cols2</code> parameter altogether, as it is assigned to None by default."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Histologic_type</th>\n",
       "      <th>Estrogen_Receptor</th>\n",
       "      <th>Estrogen_Receptor_%</th>\n",
       "      <th>Progesterone_Receptor</th>\n",
       "      <th>Progesterone_Receptor_%</th>\n",
       "      <th>MLH1</th>\n",
       "      <th>MLH2</th>\n",
       "      <th>MSH6</th>\n",
       "      <th>PMS2</th>\n",
       "      <th>p53</th>\n",
       "      <th>...</th>\n",
       "      <th>Log2_variant_total</th>\n",
       "      <th>Log2_SNP_total</th>\n",
       "      <th>Log2_INDEL_total</th>\n",
       "      <th>Genomics_subtype</th>\n",
       "      <th>Mutation_signature_C&gt;A</th>\n",
       "      <th>Mutation_signature_C&gt;G</th>\n",
       "      <th>Mutation_signature_C&gt;T</th>\n",
       "      <th>Mutation_signature_T&gt;C</th>\n",
       "      <th>Mutation_signature_T&gt;A</th>\n",
       "      <th>Mutation_signature_T&gt;G</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sample_ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>S001</th>\n",
       "      <td>Endometrioid</td>\n",
       "      <td>Cannot be determined</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Cannot be determined</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Intact nuclear expression</td>\n",
       "      <td>Intact nuclear expression</td>\n",
       "      <td>Loss of nuclear expression</td>\n",
       "      <td>Intact nuclear expression</td>\n",
       "      <td>Cannot be determined</td>\n",
       "      <td>...</td>\n",
       "      <td>10.062046</td>\n",
       "      <td>9.984418</td>\n",
       "      <td>5.832890</td>\n",
       "      <td>MSI-H</td>\n",
       "      <td>8.300395</td>\n",
       "      <td>1.482213</td>\n",
       "      <td>72.529644</td>\n",
       "      <td>14.426877</td>\n",
       "      <td>1.383399</td>\n",
       "      <td>1.877470</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S002</th>\n",
       "      <td>Endometrioid</td>\n",
       "      <td>Cannot be determined</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Cannot be determined</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Intact nuclear expression</td>\n",
       "      <td>Intact nuclear expression</td>\n",
       "      <td>Intact nuclear expression</td>\n",
       "      <td>Loss of nuclear expression</td>\n",
       "      <td>Cannot be determined</td>\n",
       "      <td>...</td>\n",
       "      <td>8.861087</td>\n",
       "      <td>8.330917</td>\n",
       "      <td>7.169925</td>\n",
       "      <td>MSI-H</td>\n",
       "      <td>14.641745</td>\n",
       "      <td>2.803738</td>\n",
       "      <td>64.485981</td>\n",
       "      <td>15.264798</td>\n",
       "      <td>0.934579</td>\n",
       "      <td>1.869159</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S003</th>\n",
       "      <td>Endometrioid</td>\n",
       "      <td>Cannot be determined</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Cannot be determined</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Intact nuclear expression</td>\n",
       "      <td>Intact nuclear expression</td>\n",
       "      <td>Intact nuclear expression</td>\n",
       "      <td>Intact nuclear expression</td>\n",
       "      <td>Cannot be determined</td>\n",
       "      <td>...</td>\n",
       "      <td>5.321928</td>\n",
       "      <td>5.000000</td>\n",
       "      <td>3.169925</td>\n",
       "      <td>CNV_low</td>\n",
       "      <td>16.129032</td>\n",
       "      <td>3.225806</td>\n",
       "      <td>70.967742</td>\n",
       "      <td>3.225806</td>\n",
       "      <td>3.225806</td>\n",
       "      <td>3.225806</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S005</th>\n",
       "      <td>Endometrioid</td>\n",
       "      <td>Cannot be determined</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Cannot be determined</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Intact nuclear expression</td>\n",
       "      <td>Intact nuclear expression</td>\n",
       "      <td>Intact nuclear expression</td>\n",
       "      <td>Intact nuclear expression</td>\n",
       "      <td>Cannot be determined</td>\n",
       "      <td>...</td>\n",
       "      <td>5.672425</td>\n",
       "      <td>5.523562</td>\n",
       "      <td>2.584963</td>\n",
       "      <td>CNV_low</td>\n",
       "      <td>17.777778</td>\n",
       "      <td>8.888889</td>\n",
       "      <td>62.222222</td>\n",
       "      <td>8.888889</td>\n",
       "      <td>2.222222</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S006</th>\n",
       "      <td>Serous</td>\n",
       "      <td>Cannot be determined</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Cannot be determined</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Intact nuclear expression</td>\n",
       "      <td>Intact nuclear expression</td>\n",
       "      <td>Intact nuclear expression</td>\n",
       "      <td>Intact nuclear expression</td>\n",
       "      <td>Normal</td>\n",
       "      <td>...</td>\n",
       "      <td>6.108524</td>\n",
       "      <td>5.954196</td>\n",
       "      <td>3.000000</td>\n",
       "      <td>CNV_high</td>\n",
       "      <td>9.836066</td>\n",
       "      <td>13.114754</td>\n",
       "      <td>62.295082</td>\n",
       "      <td>3.278689</td>\n",
       "      <td>8.196721</td>\n",
       "      <td>3.278689</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 126 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          Histologic_type     Estrogen_Receptor  Estrogen_Receptor_%  \\\n",
       "Sample_ID                                                              \n",
       "S001         Endometrioid  Cannot be determined                  NaN   \n",
       "S002         Endometrioid  Cannot be determined                  NaN   \n",
       "S003         Endometrioid  Cannot be determined                  NaN   \n",
       "S005         Endometrioid  Cannot be determined                  NaN   \n",
       "S006               Serous  Cannot be determined                  NaN   \n",
       "\n",
       "          Progesterone_Receptor  Progesterone_Receptor_%  \\\n",
       "Sample_ID                                                  \n",
       "S001       Cannot be determined                      NaN   \n",
       "S002       Cannot be determined                      NaN   \n",
       "S003       Cannot be determined                      NaN   \n",
       "S005       Cannot be determined                      NaN   \n",
       "S006       Cannot be determined                      NaN   \n",
       "\n",
       "                                MLH1                       MLH2  \\\n",
       "Sample_ID                                                         \n",
       "S001       Intact nuclear expression  Intact nuclear expression   \n",
       "S002       Intact nuclear expression  Intact nuclear expression   \n",
       "S003       Intact nuclear expression  Intact nuclear expression   \n",
       "S005       Intact nuclear expression  Intact nuclear expression   \n",
       "S006       Intact nuclear expression  Intact nuclear expression   \n",
       "\n",
       "                                 MSH6                        PMS2  \\\n",
       "Sample_ID                                                           \n",
       "S001       Loss of nuclear expression   Intact nuclear expression   \n",
       "S002        Intact nuclear expression  Loss of nuclear expression   \n",
       "S003        Intact nuclear expression   Intact nuclear expression   \n",
       "S005        Intact nuclear expression   Intact nuclear expression   \n",
       "S006        Intact nuclear expression   Intact nuclear expression   \n",
       "\n",
       "                            p53  ... Log2_variant_total Log2_SNP_total  \\\n",
       "Sample_ID                        ...                                     \n",
       "S001       Cannot be determined  ...          10.062046       9.984418   \n",
       "S002       Cannot be determined  ...           8.861087       8.330917   \n",
       "S003       Cannot be determined  ...           5.321928       5.000000   \n",
       "S005       Cannot be determined  ...           5.672425       5.523562   \n",
       "S006                     Normal  ...           6.108524       5.954196   \n",
       "\n",
       "           Log2_INDEL_total  Genomics_subtype  Mutation_signature_C>A  \\\n",
       "Sample_ID                                                               \n",
       "S001               5.832890             MSI-H                8.300395   \n",
       "S002               7.169925             MSI-H               14.641745   \n",
       "S003               3.169925           CNV_low               16.129032   \n",
       "S005               2.584963           CNV_low               17.777778   \n",
       "S006               3.000000          CNV_high                9.836066   \n",
       "\n",
       "           Mutation_signature_C>G  Mutation_signature_C>T  \\\n",
       "Sample_ID                                                   \n",
       "S001                     1.482213               72.529644   \n",
       "S002                     2.803738               64.485981   \n",
       "S003                     3.225806               70.967742   \n",
       "S005                     8.888889               62.222222   \n",
       "S006                    13.114754               62.295082   \n",
       "\n",
       "           Mutation_signature_T>C  Mutation_signature_T>A  \\\n",
       "Sample_ID                                                   \n",
       "S001                    14.426877                1.383399   \n",
       "S002                    15.264798                0.934579   \n",
       "S003                     3.225806                3.225806   \n",
       "S005                     8.888889                2.222222   \n",
       "S006                     3.278689                8.196721   \n",
       "\n",
       "           Mutation_signature_T>G  \n",
       "Sample_ID                          \n",
       "S001                     1.877470  \n",
       "S002                     1.869159  \n",
       "S003                     3.225806  \n",
       "S005                     0.000000  \n",
       "S006                     3.278689  \n",
       "\n",
       "[5 rows x 126 columns]"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hist_and_derived_molecular = en.join_metadata_to_metadata(\n",
    "    df1_name=\"clinical\",\n",
    "    df2_name=\"derived_molecular\",\n",
    "    cols1=\"Histologic_type\",\n",
    "    cols2=None) # Could have omitted cols2=None, as that's the default value\n",
    "\n",
    "hist_and_derived_molecular.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## <code>join_omics_to_mutations</code>\n",
    "\n",
    "The <code>join_omics_to_mutations</code> function joins an -omics dataframe with the mutation data for a specified gene or genes. Because there may be multiple mutations for one gene in a single sample, the mutation type and location data are returned in lists by default. If there is no mutation for the gene in a particular sample, the list contains either \"Wildtype_Tumor\" or \"Wildtype_Normal\", depending on whether it's a tumor or normal sample. The mutation status column contains either \"Single_mutation\", \"Multiple_mutation\", \"Wildtype_Tumor\", or \"Wildtype_Normal\", for help with parsing.\n",
    "\n",
    "The function has the ability to filter multiple mutations down to just one mutation. It first prioritizes mutations and locations (hotspots) that you pass to the optional <code>mutations_filter</code> parameter. Mutations and locations earlier in your filtering list are given priority over mutations and locations later in the list. When a sample has multiple mutations in a gene, but none of them were in your filter, the function chooses truncations over missense mutations, and silent mutations last of all. Between mutations of the same type, the function chooses mutations occurring earlier in the sequence. To filter all mutations based on this default hierarchy, simply pass an empty list to the <code>mutations_filter</code> parameter. Passing nothing to the <code>mutations_filter</code> parameter will cause no filtering to be done.\n",
    "\n",
    "Note that when multiple mutations are filtered, the Mutation_Status column is still included, so that if there were originally multiple mutations, you can easily know it.\n",
    "\n",
    "If you wish to drop the location columns, pass <code>False</code> to the optional show_location parameter.\n",
    "\n",
    "First, we demonstrate the function with no filtering of multiple mutations. Note that the Mutation and Location column values are in lists, even if there is only one mutation."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AAGAB-K290_acetylproteomics</th>\n",
       "      <th>AACS-K391_acetylproteomics</th>\n",
       "      <th>PTEN_Mutation</th>\n",
       "      <th>PTEN_Location</th>\n",
       "      <th>PTEN_Mutation_Status</th>\n",
       "      <th>Sample_Status</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sample_ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>S001</th>\n",
       "      <td>0.461</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[Missense_Mutation, Nonsense_Mutation]</td>\n",
       "      <td>[p.R130Q, p.R233*]</td>\n",
       "      <td>Multiple_mutation</td>\n",
       "      <td>Tumor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S002</th>\n",
       "      <td>1.770</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[Missense_Mutation]</td>\n",
       "      <td>[p.G127R]</td>\n",
       "      <td>Single_mutation</td>\n",
       "      <td>Tumor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S003</th>\n",
       "      <td>-0.815</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[Nonsense_Mutation]</td>\n",
       "      <td>[p.W111*]</td>\n",
       "      <td>Single_mutation</td>\n",
       "      <td>Tumor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S005</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[Missense_Mutation]</td>\n",
       "      <td>[p.R130G]</td>\n",
       "      <td>Single_mutation</td>\n",
       "      <td>Tumor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S006</th>\n",
       "      <td>0.205</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[Wildtype_Tumor]</td>\n",
       "      <td>[No_mutation]</td>\n",
       "      <td>Wildtype_Tumor</td>\n",
       "      <td>Tumor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S007</th>\n",
       "      <td>0.403</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[Missense_Mutation, Missense_Mutation]</td>\n",
       "      <td>[p.Y68C, p.R130G]</td>\n",
       "      <td>Multiple_mutation</td>\n",
       "      <td>Tumor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S008</th>\n",
       "      <td>0.792</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[Frame_Shift_Ins, Nonsense_Mutation]</td>\n",
       "      <td>[p.H118Qfs*8, p.Y180*]</td>\n",
       "      <td>Multiple_mutation</td>\n",
       "      <td>Tumor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S009</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[Wildtype_Tumor]</td>\n",
       "      <td>[No_mutation]</td>\n",
       "      <td>Wildtype_Tumor</td>\n",
       "      <td>Tumor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S010</th>\n",
       "      <td>-1.110</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[Missense_Mutation]</td>\n",
       "      <td>[p.R130G]</td>\n",
       "      <td>Single_mutation</td>\n",
       "      <td>Tumor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S011</th>\n",
       "      <td>0.400</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[Missense_Mutation, Frame_Shift_Ins]</td>\n",
       "      <td>[p.H93R, p.E242*]</td>\n",
       "      <td>Multiple_mutation</td>\n",
       "      <td>Tumor</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           AAGAB-K290_acetylproteomics  AACS-K391_acetylproteomics  \\\n",
       "Sample_ID                                                            \n",
       "S001                             0.461                         NaN   \n",
       "S002                             1.770                         NaN   \n",
       "S003                            -0.815                         NaN   \n",
       "S005                               NaN                         NaN   \n",
       "S006                             0.205                         NaN   \n",
       "S007                             0.403                         NaN   \n",
       "S008                             0.792                         NaN   \n",
       "S009                               NaN                         NaN   \n",
       "S010                            -1.110                         NaN   \n",
       "S011                             0.400                         NaN   \n",
       "\n",
       "                                    PTEN_Mutation           PTEN_Location  \\\n",
       "Sample_ID                                                                   \n",
       "S001       [Missense_Mutation, Nonsense_Mutation]      [p.R130Q, p.R233*]   \n",
       "S002                          [Missense_Mutation]               [p.G127R]   \n",
       "S003                          [Nonsense_Mutation]               [p.W111*]   \n",
       "S005                          [Missense_Mutation]               [p.R130G]   \n",
       "S006                             [Wildtype_Tumor]           [No_mutation]   \n",
       "S007       [Missense_Mutation, Missense_Mutation]       [p.Y68C, p.R130G]   \n",
       "S008         [Frame_Shift_Ins, Nonsense_Mutation]  [p.H118Qfs*8, p.Y180*]   \n",
       "S009                             [Wildtype_Tumor]           [No_mutation]   \n",
       "S010                          [Missense_Mutation]               [p.R130G]   \n",
       "S011         [Missense_Mutation, Frame_Shift_Ins]       [p.H93R, p.E242*]   \n",
       "\n",
       "          PTEN_Mutation_Status Sample_Status  \n",
       "Sample_ID                                     \n",
       "S001         Multiple_mutation         Tumor  \n",
       "S002           Single_mutation         Tumor  \n",
       "S003           Single_mutation         Tumor  \n",
       "S005           Single_mutation         Tumor  \n",
       "S006            Wildtype_Tumor         Tumor  \n",
       "S007         Multiple_mutation         Tumor  \n",
       "S008         Multiple_mutation         Tumor  \n",
       "S009            Wildtype_Tumor         Tumor  \n",
       "S010           Single_mutation         Tumor  \n",
       "S011         Multiple_mutation         Tumor  "
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "selected_acet_and_PTEN_mut = en.join_omics_to_mutations(\n",
    "    omics_df_name=\"acetylproteomics\",\n",
    "    mutations_genes=\"PTEN\", \n",
    "    omics_genes=[\"AAGAB\", \"AACS\"])\n",
    "\n",
    "selected_acet_and_PTEN_mut.head(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Using the mutations filtering functionality. Notice how mutations matching the mutation types and locations in the filter are prioritized, even if they occur later in the sequence than others, and how filter values earlier in the list are prioritized over values later in the list."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AAGAB-K290_acetylproteomics</th>\n",
       "      <th>AACS-K391_acetylproteomics</th>\n",
       "      <th>PTEN_Mutation</th>\n",
       "      <th>PTEN_Location</th>\n",
       "      <th>PTEN_Mutation_Status</th>\n",
       "      <th>Sample_Status</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sample_ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>S001</th>\n",
       "      <td>0.461</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Nonsense_Mutation</td>\n",
       "      <td>p.R233*</td>\n",
       "      <td>Multiple_mutation</td>\n",
       "      <td>Tumor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S002</th>\n",
       "      <td>1.770</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>p.G127R</td>\n",
       "      <td>Single_mutation</td>\n",
       "      <td>Tumor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S003</th>\n",
       "      <td>-0.815</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Nonsense_Mutation</td>\n",
       "      <td>p.W111*</td>\n",
       "      <td>Single_mutation</td>\n",
       "      <td>Tumor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S005</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>p.R130G</td>\n",
       "      <td>Single_mutation</td>\n",
       "      <td>Tumor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S006</th>\n",
       "      <td>0.205</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Wildtype_Tumor</td>\n",
       "      <td>No_mutation</td>\n",
       "      <td>Wildtype_Tumor</td>\n",
       "      <td>Tumor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S007</th>\n",
       "      <td>0.403</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>p.R130G</td>\n",
       "      <td>Multiple_mutation</td>\n",
       "      <td>Tumor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S008</th>\n",
       "      <td>0.792</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Nonsense_Mutation</td>\n",
       "      <td>p.Y180*</td>\n",
       "      <td>Multiple_mutation</td>\n",
       "      <td>Tumor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S009</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Wildtype_Tumor</td>\n",
       "      <td>No_mutation</td>\n",
       "      <td>Wildtype_Tumor</td>\n",
       "      <td>Tumor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S010</th>\n",
       "      <td>-1.110</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>p.R130G</td>\n",
       "      <td>Single_mutation</td>\n",
       "      <td>Tumor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S011</th>\n",
       "      <td>0.400</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>p.H93R</td>\n",
       "      <td>Multiple_mutation</td>\n",
       "      <td>Tumor</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           AAGAB-K290_acetylproteomics  AACS-K391_acetylproteomics  \\\n",
       "Sample_ID                                                            \n",
       "S001                             0.461                         NaN   \n",
       "S002                             1.770                         NaN   \n",
       "S003                            -0.815                         NaN   \n",
       "S005                               NaN                         NaN   \n",
       "S006                             0.205                         NaN   \n",
       "S007                             0.403                         NaN   \n",
       "S008                             0.792                         NaN   \n",
       "S009                               NaN                         NaN   \n",
       "S010                            -1.110                         NaN   \n",
       "S011                             0.400                         NaN   \n",
       "\n",
       "               PTEN_Mutation PTEN_Location PTEN_Mutation_Status Sample_Status  \n",
       "Sample_ID                                                                      \n",
       "S001       Nonsense_Mutation       p.R233*    Multiple_mutation         Tumor  \n",
       "S002       Missense_Mutation       p.G127R      Single_mutation         Tumor  \n",
       "S003       Nonsense_Mutation       p.W111*      Single_mutation         Tumor  \n",
       "S005       Missense_Mutation       p.R130G      Single_mutation         Tumor  \n",
       "S006          Wildtype_Tumor   No_mutation       Wildtype_Tumor         Tumor  \n",
       "S007       Missense_Mutation       p.R130G    Multiple_mutation         Tumor  \n",
       "S008       Nonsense_Mutation       p.Y180*    Multiple_mutation         Tumor  \n",
       "S009          Wildtype_Tumor   No_mutation       Wildtype_Tumor         Tumor  \n",
       "S010       Missense_Mutation       p.R130G      Single_mutation         Tumor  \n",
       "S011       Missense_Mutation        p.H93R    Multiple_mutation         Tumor  "
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "selected_acet_and_PTEN_mut = en.join_omics_to_mutations(\n",
    "    omics_df_name=\"acetylproteomics\",\n",
    "    mutations_genes=\"PTEN\", \n",
    "    omics_genes=[\"AAGAB\", \"AACS\"],\n",
    "    mutations_filter=[\"Nonsense_Mutation\", \"p.R130G\", \"p.Y68C\", \"p.H93R\"])\n",
    "\n",
    "selected_acet_and_PTEN_mut.head(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Using the default filtering hierarchy. Notice how without any values in the filter list, all truncations are chosen over missense mutations, and mutations earlier in the sequence are chosen over mutations later in it."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AAGAB-K290_acetylproteomics</th>\n",
       "      <th>AACS-K391_acetylproteomics</th>\n",
       "      <th>PTEN_Mutation</th>\n",
       "      <th>PTEN_Location</th>\n",
       "      <th>PTEN_Mutation_Status</th>\n",
       "      <th>Sample_Status</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sample_ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>S001</th>\n",
       "      <td>0.461</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Nonsense_Mutation</td>\n",
       "      <td>p.R233*</td>\n",
       "      <td>Multiple_mutation</td>\n",
       "      <td>Tumor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S002</th>\n",
       "      <td>1.770</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>p.G127R</td>\n",
       "      <td>Single_mutation</td>\n",
       "      <td>Tumor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S003</th>\n",
       "      <td>-0.815</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Nonsense_Mutation</td>\n",
       "      <td>p.W111*</td>\n",
       "      <td>Single_mutation</td>\n",
       "      <td>Tumor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S005</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>p.R130G</td>\n",
       "      <td>Single_mutation</td>\n",
       "      <td>Tumor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S006</th>\n",
       "      <td>0.205</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Wildtype_Tumor</td>\n",
       "      <td>No_mutation</td>\n",
       "      <td>Wildtype_Tumor</td>\n",
       "      <td>Tumor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S007</th>\n",
       "      <td>0.403</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>p.Y68C</td>\n",
       "      <td>Multiple_mutation</td>\n",
       "      <td>Tumor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S008</th>\n",
       "      <td>0.792</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Nonsense_Mutation</td>\n",
       "      <td>p.Y180*</td>\n",
       "      <td>Multiple_mutation</td>\n",
       "      <td>Tumor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S009</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Wildtype_Tumor</td>\n",
       "      <td>No_mutation</td>\n",
       "      <td>Wildtype_Tumor</td>\n",
       "      <td>Tumor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S010</th>\n",
       "      <td>-1.110</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>p.R130G</td>\n",
       "      <td>Single_mutation</td>\n",
       "      <td>Tumor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S011</th>\n",
       "      <td>0.400</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Frame_Shift_Ins</td>\n",
       "      <td>p.E242*</td>\n",
       "      <td>Multiple_mutation</td>\n",
       "      <td>Tumor</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           AAGAB-K290_acetylproteomics  AACS-K391_acetylproteomics  \\\n",
       "Sample_ID                                                            \n",
       "S001                             0.461                         NaN   \n",
       "S002                             1.770                         NaN   \n",
       "S003                            -0.815                         NaN   \n",
       "S005                               NaN                         NaN   \n",
       "S006                             0.205                         NaN   \n",
       "S007                             0.403                         NaN   \n",
       "S008                             0.792                         NaN   \n",
       "S009                               NaN                         NaN   \n",
       "S010                            -1.110                         NaN   \n",
       "S011                             0.400                         NaN   \n",
       "\n",
       "               PTEN_Mutation PTEN_Location PTEN_Mutation_Status Sample_Status  \n",
       "Sample_ID                                                                      \n",
       "S001       Nonsense_Mutation       p.R233*    Multiple_mutation         Tumor  \n",
       "S002       Missense_Mutation       p.G127R      Single_mutation         Tumor  \n",
       "S003       Nonsense_Mutation       p.W111*      Single_mutation         Tumor  \n",
       "S005       Missense_Mutation       p.R130G      Single_mutation         Tumor  \n",
       "S006          Wildtype_Tumor   No_mutation       Wildtype_Tumor         Tumor  \n",
       "S007       Missense_Mutation        p.Y68C    Multiple_mutation         Tumor  \n",
       "S008       Nonsense_Mutation       p.Y180*    Multiple_mutation         Tumor  \n",
       "S009          Wildtype_Tumor   No_mutation       Wildtype_Tumor         Tumor  \n",
       "S010       Missense_Mutation       p.R130G      Single_mutation         Tumor  \n",
       "S011         Frame_Shift_Ins       p.E242*    Multiple_mutation         Tumor  "
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "selected_acet_and_PTEN_mut = en.join_omics_to_mutations(\n",
    "    omics_df_name=\"acetylproteomics\",\n",
    "    mutations_genes=\"PTEN\", \n",
    "    omics_genes=[\"AAGAB\", \"AACS\"],\n",
    "    mutations_filter=[])\n",
    "\n",
    "selected_acet_and_PTEN_mut.head(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## <code>join_metadata_to_mutations</code>\n",
    "\n",
    "The <code>join_metadata_to_mutations</code> function works exactly like <code>join_omics_to_mutations</code>, except that it works with metadata dataframes (e.g. clinical and derived molecular) instead of omics dataframes. It also can filter multiple mutations, which you control through the <code>mutations_filter</code> parameter, and has the ability to hide the location colunms."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Histologic_type</th>\n",
       "      <th>PTEN_Mutation</th>\n",
       "      <th>PTEN_Location</th>\n",
       "      <th>PTEN_Mutation_Status</th>\n",
       "      <th>Sample_Status</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sample_ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>S001</th>\n",
       "      <td>Endometrioid</td>\n",
       "      <td>[Missense_Mutation, Nonsense_Mutation]</td>\n",
       "      <td>[p.R130Q, p.R233*]</td>\n",
       "      <td>Multiple_mutation</td>\n",
       "      <td>Tumor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S002</th>\n",
       "      <td>Endometrioid</td>\n",
       "      <td>[Missense_Mutation]</td>\n",
       "      <td>[p.G127R]</td>\n",
       "      <td>Single_mutation</td>\n",
       "      <td>Tumor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S003</th>\n",
       "      <td>Endometrioid</td>\n",
       "      <td>[Nonsense_Mutation]</td>\n",
       "      <td>[p.W111*]</td>\n",
       "      <td>Single_mutation</td>\n",
       "      <td>Tumor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S005</th>\n",
       "      <td>Endometrioid</td>\n",
       "      <td>[Missense_Mutation]</td>\n",
       "      <td>[p.R130G]</td>\n",
       "      <td>Single_mutation</td>\n",
       "      <td>Tumor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S006</th>\n",
       "      <td>Serous</td>\n",
       "      <td>[Wildtype_Tumor]</td>\n",
       "      <td>[No_mutation]</td>\n",
       "      <td>Wildtype_Tumor</td>\n",
       "      <td>Tumor</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          Histologic_type                           PTEN_Mutation  \\\n",
       "Sample_ID                                                           \n",
       "S001         Endometrioid  [Missense_Mutation, Nonsense_Mutation]   \n",
       "S002         Endometrioid                     [Missense_Mutation]   \n",
       "S003         Endometrioid                     [Nonsense_Mutation]   \n",
       "S005         Endometrioid                     [Missense_Mutation]   \n",
       "S006               Serous                        [Wildtype_Tumor]   \n",
       "\n",
       "                PTEN_Location PTEN_Mutation_Status Sample_Status  \n",
       "Sample_ID                                                         \n",
       "S001       [p.R130Q, p.R233*]    Multiple_mutation         Tumor  \n",
       "S002                [p.G127R]      Single_mutation         Tumor  \n",
       "S003                [p.W111*]      Single_mutation         Tumor  \n",
       "S005                [p.R130G]      Single_mutation         Tumor  \n",
       "S006            [No_mutation]       Wildtype_Tumor         Tumor  "
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hist_and_PTEN = en.join_metadata_to_mutations(\n",
    "    metadata_df_name=\"clinical\",\n",
    "    mutations_genes=\"PTEN\",\n",
    "    metadata_cols=\"Histologic_type\")\n",
    "\n",
    "hist_and_PTEN.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "With multiple mutations filtered:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Histologic_type</th>\n",
       "      <th>PTEN_Mutation</th>\n",
       "      <th>PTEN_Location</th>\n",
       "      <th>PTEN_Mutation_Status</th>\n",
       "      <th>Sample_Status</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sample_ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>S001</th>\n",
       "      <td>Endometrioid</td>\n",
       "      <td>Nonsense_Mutation</td>\n",
       "      <td>p.R233*</td>\n",
       "      <td>Multiple_mutation</td>\n",
       "      <td>Tumor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S002</th>\n",
       "      <td>Endometrioid</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>p.G127R</td>\n",
       "      <td>Single_mutation</td>\n",
       "      <td>Tumor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S003</th>\n",
       "      <td>Endometrioid</td>\n",
       "      <td>Nonsense_Mutation</td>\n",
       "      <td>p.W111*</td>\n",
       "      <td>Single_mutation</td>\n",
       "      <td>Tumor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S005</th>\n",
       "      <td>Endometrioid</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>p.R130G</td>\n",
       "      <td>Single_mutation</td>\n",
       "      <td>Tumor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S006</th>\n",
       "      <td>Serous</td>\n",
       "      <td>Wildtype_Tumor</td>\n",
       "      <td>No_mutation</td>\n",
       "      <td>Wildtype_Tumor</td>\n",
       "      <td>Tumor</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          Histologic_type      PTEN_Mutation PTEN_Location  \\\n",
       "Sample_ID                                                    \n",
       "S001         Endometrioid  Nonsense_Mutation       p.R233*   \n",
       "S002         Endometrioid  Missense_Mutation       p.G127R   \n",
       "S003         Endometrioid  Nonsense_Mutation       p.W111*   \n",
       "S005         Endometrioid  Missense_Mutation       p.R130G   \n",
       "S006               Serous     Wildtype_Tumor   No_mutation   \n",
       "\n",
       "          PTEN_Mutation_Status Sample_Status  \n",
       "Sample_ID                                     \n",
       "S001         Multiple_mutation         Tumor  \n",
       "S002           Single_mutation         Tumor  \n",
       "S003           Single_mutation         Tumor  \n",
       "S005           Single_mutation         Tumor  \n",
       "S006            Wildtype_Tumor         Tumor  "
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hist_and_PTEN = en.join_metadata_to_mutations(\n",
    "    metadata_df_name=\"clinical\",\n",
    "    mutations_genes=\"PTEN\",\n",
    "    metadata_cols=\"Histologic_type\",\n",
    "    mutations_filter=[\"Nonsense_Mutation\"])\n",
    "\n",
    "hist_and_PTEN.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Exporting dataframes\n",
    "\n",
    "If you wish to export a dataframe to a file, simply call the dataframe's to_csv method, passing the path you wish to save the file to, and the value separator you want:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "more_selected_acet_and_mut.to_csv(path_or_buf=\"more_selected_acet_and_mut.tsv\", sep='\\t')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
